// @razorpay/blade/build/lib/web/production/components/BaseMenu/BaseMenuItem/StyledMenuItemContainer.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import styled from 'styled-components';
import { getBaseMenuItemStyles } from './getBaseMenuItemStyles.js';
import { getItemPadding } from './tokens.js';
import '../../../utils/index.js';
import '../../../utils/getFocusRingStyles/index.js';
import '../../Box/BaseBox/index.js';
import { BaseBox } from '../../Box/BaseBox/BaseBox.web.js';
import { makeSize } from '../../../utils/makeSize/makeSize.js';
import { getMediaQuery } from '../../../utils/getMediaQuery/getMediaQuery.js';
import { getFocusRingStyles } from '../../../utils/getFocusRingStyles/getFocusRingStyles.web.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StyledMenuItemContainer = /*#__PURE__*/styled(BaseBox).withConfig({
  displayName: "StyledMenuItemContainerweb__StyledMenuItemContainer",
  componentId: "cjjm5b-0"
})(function (props) {
  return _objectSpread(_objectSpread({}, getBaseMenuItemStyles({
    theme: props.theme
  })), {}, _defineProperty(_defineProperty(_defineProperty(_defineProperty(_defineProperty(_defineProperty({
    padding: makeSize(getItemPadding(props.theme).itemPaddingMobile),
    display: props.isVisible ? 'flex' : 'none'
  }, "@media ".concat(getMediaQuery({
    min: props.theme.breakpoints.m
  })), {
    padding: makeSize(getItemPadding(props.theme).itemPaddingDesktop)
  }), '&:hover:not([aria-disabled=true]), &[aria-expanded="true"]', {
    backgroundColor: props.color === 'negative' ? props.theme.colors.interactive.background.negative.faded : props.theme.colors.interactive.background.gray["default"]
  }), '&:focus-visible', getFocusRingStyles({
    theme: props.theme
  })), '&.active-focus', props.isKeydownPressed ? getFocusRingStyles({
    theme: props.theme
  }) : undefined), '&[aria-selected=true]', {
    backgroundColor: props.selectionType === 'single' ? props.theme.colors.interactive.background.primary.faded : undefined
  }), '&[aria-selected=true]:hover', {
    backgroundColor: props.theme.colors.interactive.background.primary.fadedHighlighted
  }));
});

export { StyledMenuItemContainer };
