// @razorpay/blade/build/lib/web/production/components/Slide/Slide.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _typeof from '@babel/runtime/helpers/typeof';
import React__default from 'react';
import '../BaseMotion/index.js';
import { msToSeconds } from '../../utils/msToSeconds.js';
import { cssBezierToArray } from '../../utils/cssBezierToArray.js';
import '../../utils/index.js';
import { jsx } from 'react/jsx-runtime';
import useTheme from '../BladeProvider/useTheme.js';
import { castWebType } from '../../utils/platform/castUtils.js';
import { BaseMotionEntryExit } from '../BaseMotion/BaseMotion.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var getFromTransform = function getFromTransform(direction, fromOffset) {
  if (direction === 'top') {
    return "translateY(-".concat(fromOffset, ")");
  }
  if (direction === 'left') {
    return "translateX(-".concat(fromOffset, ")");
  }
  if (direction === 'right') {
    return "translateX(".concat(fromOffset, ")");
  }
  return "translateY(".concat(fromOffset, ")");
};

/**
 * ## Slide
 *
 * Slide is one of the motion presets that we expose from blade to help you make components slide in from outside of viewport
 *
 * If you're looking for subtle movement on enter in the viewport itself, checkout `Move` preset instead.
 *
 * ### Usage
 *
 * #### Slide in on mount
 *
 * ```jsx
 * <Slide direction="bottom">
 *   <Card />
 * </Slide>
 * ```
 *
 * #### Conditionally slide based on state
 *
 * ```jsx
 * <Slide isVisible={isVisibleState}>
 *   <Card />
 * </Slide>
 * ```
 */
var Slide = function Slide(_ref) {
  var children = _ref.children,
    _ref$type = _ref.type,
    type = _ref$type === void 0 ? 'inout' : _ref$type,
    _ref$direction = _ref.direction,
    direction = _ref$direction === void 0 ? 'bottom' : _ref$direction,
    isVisible = _ref.isVisible,
    motionTriggers = _ref.motionTriggers,
    shouldUnmountWhenHidden = _ref.shouldUnmountWhenHidden,
    fromOffset = _ref.fromOffset,
    delay = _ref.delay;
  var _useTheme = useTheme(),
    theme = _useTheme.theme;
  var enterDirection = _typeof(direction) === 'object' ? direction.enter : direction;
  var exitDirection = _typeof(direction) === 'object' ? direction.exit : direction;
  var isEnterDirectionHorizontal = ['left', 'right'].includes(enterDirection);
  var isExitDirectionHorizontal = ['left', 'right'].includes(exitDirection);
  var defaultOffset = isEnterDirectionHorizontal ? '100vw' : '100vh';
  var enterTransform = getFromTransform(enterDirection, fromOffset !== null && fromOffset !== void 0 ? fromOffset : defaultOffset);
  var exitTransform = getFromTransform(exitDirection, fromOffset !== null && fromOffset !== void 0 ? fromOffset : defaultOffset);
  var enterDelay = _typeof(delay) === 'object' ? delay.enter : delay;
  var exitDelay = _typeof(delay) === 'object' ? delay.exit : delay;
  var moveVariants = React__default.useMemo(function () {
    return {
      initial: {
        // We keep element in view with opacity 0 initially so that it works with `in-view` trigger as expected
        opacity: 0
      },
      animate: {
        transform: [enterTransform, 'translateY(0%)'],
        opacity: 1,
        transition: _objectSpread(_objectSpread({}, enterDelay ? {
          delay: msToSeconds(theme.motion.delay[enterDelay])
        } : {}), {}, {
          duration: msToSeconds(isEnterDirectionHorizontal ? theme.motion.duration.xmoderate : theme.motion.duration['2xgentle']),
          ease: cssBezierToArray(isEnterDirectionHorizontal ? castWebType(theme.motion.easing.entrance) : castWebType(theme.motion.easing.emphasized))
        })
      },
      exit: {
        opacity: 0,
        transform: exitTransform,
        transitionEnd: {
          transform: enterTransform
        },
        transition: _objectSpread(_objectSpread({}, exitDelay ? {
          delay: msToSeconds(theme.motion.delay[exitDelay])
        } : {}), {}, {
          duration: msToSeconds(isExitDirectionHorizontal ? theme.motion.duration.moderate : theme.motion.duration.xgentle),
          ease: cssBezierToArray(isExitDirectionHorizontal ? castWebType(theme.motion.easing.exit) : castWebType(theme.motion.easing.emphasized))
        })
      }
    };
  },
  // eslint-disable-next-line react-hooks/exhaustive-deps
  [enterDirection, exitDirection, isEnterDirectionHorizontal, isExitDirectionHorizontal, theme.name]);
  return /*#__PURE__*/jsx(BaseMotionEntryExit, {
    motionVariants: moveVariants,
    children: children,
    type: type,
    isVisible: isVisible,
    motionTriggers: motionTriggers,
    shouldUnmountWhenHidden: shouldUnmountWhenHidden
  });
};

export { Slide };
