// @razorpay/blade/build/lib/web/production/components/InfoGroup/infoGroupTokens.js
import '../../tokens/global/index.js';
import { size } from '../../tokens/global/size.js';

// Gap between leading/trailing elements and content
var elementGap = {
  xsmall: 'spacing.1',
  small: 'spacing.2',
  medium: 'spacing.2',
  large: 'spacing.3'
};

// Size mapping for InfoItemKey
var titleTextSize = {
  xsmall: 'xsmall',
  small: 'small',
  medium: 'medium',
  large: 'large'
};

// Text sizes for help text (has different mapping)
var helpTextSize = {
  xsmall: 'xsmall',
  small: 'xsmall',
  medium: 'xsmall',
  large: 'small'
};

// Icon sizes for different InfoGroup sizes
var iconSize = {
  xsmall: 'xsmall',
  small: 'small',
  medium: 'medium',
  large: 'large'
};
var itemTitleHeight = {
  xsmall: size['14'],
  small: size['18'],
  medium: size['20'],
  large: size['24']
};
var avatarAdjustmentPaddingY = {
  xsmall: 'spacing.1',
  small: 'spacing.2',
  medium: 'spacing.2',
  large: 'spacing.3'
};

export { avatarAdjustmentPaddingY, elementGap, helpTextSize, iconSize, itemTitleHeight, titleTextSize };
