// @razorpay/blade/build/lib/web/development/components/ProgressBar/CircularProgressLabel.js
import 'react';
import '../Box/BaseBox/index.js';
import '../Typography/index.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { Text } from '../Typography/Text/Text.js';

var CircularProgressLabel = function CircularProgressLabel(_ref) {
  var _label$trim;
  var progressPercent = _ref.progressPercent,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'small' : _ref$size,
    label = _ref.label,
    _ref$showPercentage = _ref.showPercentage,
    showPercentage = _ref$showPercentage === void 0 ? true : _ref$showPercentage;
  var hasLabel = label && ((_label$trim = label.trim()) === null || _label$trim === void 0 ? void 0 : _label$trim.length) > 0;
  return /*#__PURE__*/jsxs(BaseBox, {
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
    children: [hasLabel && /*#__PURE__*/jsx(Text, {
      marginTop: "spacing.3",
      variant: "body",
      weight: "regular",
      size: "small",
      children: label
    }), showPercentage && size === 'small' && /*#__PURE__*/jsxs(Text, {
      marginTop: hasLabel ? 'spacing.0' : 'spacing.3',
      variant: "body",
      weight: "semibold",
      size: "small",
      children: [progressPercent, "%"]
    })]
  });
};

export { CircularProgressLabel };
