// @razorpay/blade/build/lib/web/development/components/Tooltip/getTooltipContentWrapperStyles.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import '../../utils/index.js';
import { makeBorderSize } from '../../utils/makeBorderSize/makeBorderSize.js';
import { isReactNative } from '../../utils/platform/isReactNative.js';
import { castWebType } from '../../utils/platform/castUtils.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var getTooltipContentWrapperStyles = function getTooltipContentWrapperStyles(_ref) {
  var theme = _ref.theme,
    styles = _ref.styles;
  return _objectSpread({
    backgroundColor: theme.colors.popup.background.intense,
    borderWidth: makeBorderSize(theme.border.width.thin),
    borderRadius: makeBorderSize(theme.border.radius.medium),
    borderColor: theme.colors.popup.border.intense,
    borderStyle: 'solid',
    boxShadow: isReactNative() ? undefined : castWebType(theme.elevation.lowRaised)
  }, styles);
};

export { getTooltipContentWrapperStyles };
