// @razorpay/blade/build/lib/web/development/components/Tooltip/TooltipInteractiveWrapper.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import React__default from 'react';
import { useTooltipContext } from './TooltipContext.js';
import '../Box/BaseBox/index.js';
import '../../utils/metaAttribute/index.js';
import { jsx } from 'react/jsx-runtime';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var TooltipInteractiveWrapper = /*#__PURE__*/React__default.forwardRef(function (props, ref) {
  useTooltipContext();
  return /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread({
    ref: ref,
    tabIndex: -1,
    display: "inline-block"
  }, metaAttribute({
    testID: 'tooltip-interactive-wrapper',
    name: MetaConstants.TooltipInteractiveWrapper
  })), props));
});

export { TooltipInteractiveWrapper };
