// @razorpay/blade/build/lib/web/production/components/BladeProvider/useBladeProvider.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _typeof from '@babel/runtime/helpers/typeof';
import '../../utils/index.js';
import { colorSchemeNamesInput } from '../../tokens/theme/theme.js';
import '../../utils/toTitleCase/index.js';
import '../../utils/logger/index.js';
import { throwBladeError } from '../../utils/logger/logger.js';
import { useColorScheme } from '../../utils/useColorScheme/useColorScheme.js';
import { useBreakpoint } from '../../utils/useBreakpoint/useBreakpoint.js';
import { toTitleCase } from '../../utils/toTitleCase/toTitleCase.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
/**
 * Reusable hook to be used in BladeProvider.native & BladeProvider.web file
 *
 * This hook processes incoming themeTokens & initialColorScheme
 * And validates & returns the theme values
 */
var useBladeProvider = function useBladeProvider(_ref) {
  var themeTokens = _ref.themeTokens,
    initialColorScheme = _ref.initialColorScheme;
  if (false) {
    if (!themeTokens) {
      throwBladeError({
        message: "Expected valid themeTokens of type ThemeTokens to be passed but found ".concat(_typeof(themeTokens)),
        moduleName: 'BladeProvider'
      });
    }
    if (initialColorScheme && !colorSchemeNamesInput.includes(initialColorScheme)) {
      throwBladeError({
        message: "Expected color scheme to be one of [".concat(colorSchemeNamesInput.toString(), "] but received ").concat(initialColorScheme),
        moduleName: 'BladeProvider'
      });
    }
  }
  var _useColorScheme = useColorScheme(initialColorScheme),
    colorScheme = _useColorScheme.colorScheme,
    setColorScheme = _useColorScheme.setColorScheme;
  var _useBreakpoint = useBreakpoint({
      breakpoints: themeTokens.breakpoints
    }),
    matchedDeviceType = _useBreakpoint.matchedDeviceType;
  var onColorMode = "on".concat(toTitleCase(colorScheme));
  var onDeviceType = "on".concat(toTitleCase(matchedDeviceType));
  var theme = _objectSpread(_objectSpread({}, themeTokens), {}, {
    colors: themeTokens.colors[onColorMode],
    elevation: themeTokens.elevation[onColorMode],
    typography: themeTokens.typography[onDeviceType]
  });
  var themeContextValue = {
    theme: theme,
    colorScheme: colorScheme,
    setColorScheme: setColorScheme,
    platform: onDeviceType
  };
  return {
    themeContextValue: themeContextValue,
    theme: theme
  };
};

export { useBladeProvider };
