// @razorpay/blade/build/lib/web/production/utils/makeAccessible/accessibilityMapWeb.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var accessibilityValue = {
  valueMax: 'aria-valuemax',
  valueMin: 'aria-valuemin',
  valueNow: 'aria-valuenow',
  valueText: 'aria-valuetext'
};
var accessibilityState = {
  selected: 'aria-selected',
  disabled: 'aria-disabled',
  expanded: 'aria-expanded',
  busy: 'aria-busy',
  checked: 'aria-checked'
};

// TODO:
// accessibilityViewIsModal
var accessibilityMap = _objectSpread(_objectSpread(_objectSpread({}, accessibilityState), accessibilityValue), {}, {
  activeDescendant: 'aria-activedescendant',
  atomic: 'aria-atomic',
  autoComplete: 'aria-autocomplete',
  colCount: 'aria-colcount',
  colIndex: 'aria-colindex',
  colSpan: 'aria-colspan',
  controls: 'aria-controls',
  describedBy: 'aria-describedby',
  details: 'aria-details',
  errorMessage: 'aria-errormessage',
  flowTo: 'aria-flowto',
  hasPopup: 'aria-haspopup',
  hidden: 'aria-hidden',
  invalid: 'aria-invalid',
  keyShortcuts: 'aria-keyshortcuts',
  label: 'aria-label',
  labelledBy: 'aria-labelledby',
  liveRegion: 'aria-live',
  modal: 'aria-modal',
  multiline: 'aria-multiline',
  multiSelectable: 'aria-multiselectable',
  orientation: 'aria-orientation',
  owns: 'aria-owns',
  placeholder: 'aria-placeholder',
  posInSet: 'aria-posinset',
  pressed: 'aria-pressed',
  readOnly: 'aria-readonly',
  required: 'aria-required',
  role: 'role',
  roleDescription: 'aria-roledescription',
  rowCount: 'aria-rowcount',
  rowIndex: 'aria-rowindex',
  rowSpan: 'aria-rowspan',
  setSize: 'aria-setsize',
  sort: 'aria-sort',
  current: 'aria-current',
  dropEffect: 'aria-dropeffect',
  grabbed: 'aria-grabbed',
  level: 'aria-level',
  relevant: 'aria-relevant'
});

export { accessibilityMap, accessibilityState, accessibilityValue };
