// @razorpay/blade/build/lib/web/development/components/Avatar/StyledAvatar.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import styled from 'styled-components';
import { avatarSizeTokens, avatarBorderRadiusTokens } from './avatarTokens.js';
import '../Box/BaseBox/index.js';
import '../../utils/index.js';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { makeSize } from '../../utils/makeSize/makeSize.js';
import { makeBorderSize } from '../../utils/makeBorderSize/makeBorderSize.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StyledAvatar = /*#__PURE__*/styled(BaseBox).withConfig({
  displayName: "StyledAvatar",
  componentId: "sc-127ewkg-0"
})(function (_ref) {
  var theme = _ref.theme,
    variant = _ref.variant,
    size = _ref.size,
    isInteractive = _ref.isInteractive;
  return _objectSpread({
    display: 'flex',
    width: makeSize(avatarSizeTokens[size]),
    height: makeSize(avatarSizeTokens[size]),
    borderRadius: makeBorderSize(theme.border.radius[avatarBorderRadiusTokens[variant]]),
    outline: "".concat(makeBorderSize(theme.border.width.thinner), " solid ").concat(theme.colors.surface.border.gray.subtle)
  }, isInteractive ? {
    '&:hover': {
      outline: "".concat(makeBorderSize(theme.border.width.thick), " solid ").concat(theme.colors.surface.border.gray.muted),
      borderColor: theme.colors.surface.border.gray.muted,
      backgroundColor: theme.colors.surface.background.gray.moderate
    }
  } : {});
});

export { StyledAvatar };
