// @razorpay/blade/build/lib/web/development/components/Tabs/TabIndicator.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import React__default from 'react';
import styled from 'styled-components';
import { useTabsContext } from './TabsContext.js';
import '../BladeProvider/index.js';
import '../../utils/index.js';
import { useIsomorphicLayoutEffect } from '../../utils/useIsomorphicLayoutEffect.js';
import '../Box/BaseBox/index.js';
import '../../utils/metaAttribute/index.js';
import { jsx } from 'react/jsx-runtime';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import useTheme from '../BladeProvider/useTheme.js';
import { castWebType } from '../../utils/platform/castUtils.js';
import { makeMotionTime } from '../../utils/makeMotionTime/makeMotionTime.web.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StyledTabIndicator = /*#__PURE__*/styled(BaseBox).withConfig({
  displayName: "TabIndicatorweb__StyledTabIndicator",
  componentId: "sc-1kc65o8-0"
})(function (_ref) {
  var theme = _ref.theme;
  return {
    cursor: 'pointer',
    '&:hover': {
      backgroundColor: theme.colors.interactive.background.primary.fadedHighlighted
    }
  };
});
var TabIndicator = function TabIndicator(_ref2) {
  var tabListContainerRef = _ref2.tabListContainerRef;
  var _useTheme = useTheme(),
    theme = _useTheme.theme;
  var _useTabsContext = useTabsContext(),
    selectedValue = _useTabsContext.selectedValue,
    baseId = _useTabsContext.baseId,
    variant = _useTabsContext.variant;
  var _React$useState = React__default.useState(false),
    _React$useState2 = _slicedToArray(_React$useState, 2),
    hasMeasured = _React$useState2[0],
    setHasMeasured = _React$useState2[1];
  var _React$useState3 = React__default.useState({
      width: 0,
      height: 0,
      x: 0,
      y: 0
    }),
    _React$useState4 = _slicedToArray(_React$useState3, 2),
    activeElementDimensions = _React$useState4[0],
    setActiveElementDimensions = _React$useState4[1];

  // Update the dimensions of the active element
  var updateDimensions = React__default.useCallback(function () {
    var tabItemId = "".concat(baseId, "-").concat(selectedValue, "-tabitem");
    var activeTabItem = document.getElementById(tabItemId);
    if (!activeTabItem) return;
    setActiveElementDimensions({
      width: activeTabItem.offsetWidth,
      height: activeTabItem.offsetHeight,
      x: activeTabItem.offsetLeft,
      y: variant === 'filled' ?
      // on filled variant the indicator is positioned on top of the tab item
      // so no need to add offsetHeight
      activeTabItem.offsetTop : activeTabItem.offsetTop + activeTabItem.offsetHeight - 1.5
    });
  }, [baseId, selectedValue, tabListContainerRef, variant]);

  // Update the dimensions when the selected value changes
  useIsomorphicLayoutEffect(function () {
    if (!selectedValue) return;
    updateDimensions();
    var id = requestAnimationFrame(function () {
      setHasMeasured(true);
    });
    return function () {
      if (id) {
        cancelAnimationFrame(id);
      }
    };
  }, [baseId, selectedValue]);

  // Update the dimensions when the window resizes or when the font loads
  React__default.useEffect(function () {
    if (!tabListContainerRef.current) return;

    // check for FontFace API support
    // FontFaceAPI is widely supported but better to be safe than sorry
    if ('fonts' in document) {
      try {
        // wait for fonts to be loaded and then recalculate the dimensions
        void document.fonts.ready.then(function () {
          updateDimensions();
        });
      } catch (err) {
        /* empty */
      }
    }
    window.addEventListener('resize', updateDimensions);
    return function () {
      if (!tabListContainerRef.current) return;
      window.removeEventListener('resize', updateDimensions);
    };
  }, [tabListContainerRef, updateDimensions]);
  var transitionProps = {
    transitionProperty: 'transform, width, background-color',
    transitionDuration: hasMeasured ? castWebType(makeMotionTime(theme.motion.duration.gentle)) : '0ms',
    transitionTimingFunction: castWebType(theme.motion.easing.standard)
  };
  if (variant === 'filled') {
    return /*#__PURE__*/jsx(StyledTabIndicator, _objectSpread({
      pointerEvents: "none",
      position: "absolute",
      left: "0px",
      top: "0px",
      borderRadius: "small",
      backgroundColor: "interactive.background.primary.faded",
      style: _objectSpread(_objectSpread({}, transitionProps), {}, {
        width: "".concat(activeElementDimensions.width, "px"),
        height: "".concat(activeElementDimensions.height, "px"),
        transform: "translate(".concat(activeElementDimensions.x, "px, ").concat(activeElementDimensions.y, "px)")
      })
    }, metaAttribute({
      name: MetaConstants.TabIndicator
    })));
  }
  return /*#__PURE__*/jsx(StyledTabIndicator, _objectSpread({
    pointerEvents: "none",
    position: "absolute",
    left: "0%",
    top: "-0.5px",
    height: "2px",
    backgroundColor: "interactive.background.primary.default",
    style: _objectSpread(_objectSpread({}, transitionProps), {}, {
      width: "".concat(activeElementDimensions.width, "px"),
      transform: "translate(".concat(activeElementDimensions.x, "px, ").concat(activeElementDimensions.y, "px)")
    })
  }, metaAttribute({
    name: MetaConstants.TabIndicator
  })));
};

export { TabIndicator };
