// @razorpay/blade/build/lib/web/development/components/Tabs/TabPanel.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import 'react';
import { useTabsContext } from './TabsContext.js';
import '../../utils/makeAccessible/index.js';
import '../Box/BaseBox/index.js';
import '../../utils/metaAttribute/index.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import { jsx, Fragment } from 'react/jsx-runtime';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { makeAccessible } from '../../utils/makeAccessible/makeAccessible.web.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';

var _excluded = ["children", "value"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var TabPanel = function TabPanel(_ref) {
  var children = _ref.children,
    value = _ref.value,
    rest = _objectWithoutProperties(_ref, _excluded);
  var _useTabsContext = useTabsContext(),
    selectedValue = _useTabsContext.selectedValue,
    baseId = _useTabsContext.baseId,
    isLazy = _useTabsContext.isLazy;
  var isSelected = selectedValue === value;
  var panelId = "".concat(baseId, "-").concat(value, "-tabpanel");
  var tabItemId = "".concat(baseId, "-").concat(value, "-tabitem");
  if (isLazy) {
    return /*#__PURE__*/jsx(Fragment, {
      children: isSelected ? /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread(_objectSpread(_objectSpread({
        id: panelId,
        tabIndex: 0
      }, metaAttribute({
        name: MetaConstants.TabPanel
      })), makeAccessible({
        role: 'tabpanel',
        labelledBy: tabItemId
      })), makeAnalyticsAttribute(rest)), {}, {
        children: children
      })) : null
    });
  }
  return /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread(_objectSpread(_objectSpread({
    id: panelId,
    tabIndex: 0,
    display: isSelected ? 'block' : 'none'
  }, metaAttribute({
    name: MetaConstants.TabPanel
  })), makeAccessible({
    role: 'tabpanel',
    labelledBy: tabItemId,
    hidden: !isSelected
  })), makeAnalyticsAttribute(rest)), {}, {
    children: children
  }));
};

export { TabPanel };
