// @razorpay/blade/build/lib/web/development/components/Tabs/Tabs.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default from 'react';
import { TabsContext } from './TabsContext.js';
import { useControllableState } from '../../utils/useControllable.js';
import { useId } from '../../utils/useId.js';
import '../Box/BaseBox/index.js';
import '../../utils/metaAttribute/index.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import { jsx } from 'react/jsx-runtime';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';

var _excluded = ["children", "defaultValue", "value", "onChange", "orientation", "size", "variant", "isFullWidthTabItem", "isLazy"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _Tabs = function _Tabs(_ref, ref) {
  var children = _ref.children,
    defaultValue = _ref.defaultValue,
    value = _ref.value,
    _onChange = _ref.onChange,
    _ref$orientation = _ref.orientation,
    orientation = _ref$orientation === void 0 ? 'horizontal' : _ref$orientation,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'medium' : _ref$size,
    _ref$variant = _ref.variant,
    variant = _ref$variant === void 0 ? 'bordered' : _ref$variant,
    _ref$isFullWidthTabIt = _ref.isFullWidthTabItem,
    isFullWidthTabItem = _ref$isFullWidthTabIt === void 0 ? false : _ref$isFullWidthTabIt,
    _ref$isLazy = _ref.isLazy,
    isLazy = _ref$isLazy === void 0 ? false : _ref$isLazy,
    rest = _objectWithoutProperties(_ref, _excluded);
  var baseId = useId('tabs');
  var _useControllableState = useControllableState({
      defaultValue: defaultValue,
      value: value,
      onChange: function onChange(value) {
        _onChange === null || _onChange === void 0 || _onChange(value);
      }
    }),
    _useControllableState2 = _slicedToArray(_useControllableState, 2),
    selectedValue = _useControllableState2[0],
    setSelectedValue = _useControllableState2[1];
  var isVertical = orientation === 'vertical';
  var contextValue = React__default.useMemo(function () {
    return {
      baseId: baseId,
      selectedValue: selectedValue,
      isVertical: isVertical,
      size: size,
      variant: variant,
      isFullWidthTabItem: isFullWidthTabItem,
      setSelectedValue: setSelectedValue,
      isLazy: isLazy
    };
  }, [isFullWidthTabItem, baseId, isVertical, selectedValue, setSelectedValue, size, variant, isLazy]);
  return /*#__PURE__*/jsx(TabsContext.Provider, {
    value: contextValue,
    children: /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread(_objectSpread({
      ref: ref,
      display: "flex",
      flexDirection: isVertical ? 'row' : 'column'
    }, metaAttribute({
      name: MetaConstants.Tabs
    })), makeAnalyticsAttribute(rest)), {}, {
      children: children
    }))
  });
};

/**
 * ### Tabs
 *
 * Check out the [Tab Stories & Examples](https://blade.razorpay.com/?path=/docs/components-tabs--default)
 *
 * ----
 * ### Basic Usage
 *
 * ```jsx
 * import { Tabs, TabList, TabItem, TabPanel } from '@razorpay/blade/components';
 *
 * <Tabs variant="bordered" orientation="horizontal">
 *   <TabList>
 *     <TabItem value="subscriptions">Subscription</TabItem>
 *     <TabItem value="plans">Plans</TabItem>
 *     <TabItem value="settings">Settings</TabItem>
 *   </TabList>
 *
 *   <TabPanel value="subscriptions">
 *     <Text>Subscriptions Panel</Text>
 *   </TabPanel>
 *   <TabPanel value="plans">
 *     <Text>Plans Panel</Text>
 *   </TabPanel>
 *   <TabPanel value="settings">
 *     <Text>Settings Panel</Text>
 *   </TabPanel>
 * </Tabs>
 * ```
 */
var Tabs = /*#__PURE__*/React__default.forwardRef(_Tabs);

export { Tabs };
