// @razorpay/blade/build/lib/web/production/components/Form/formTokens.js
import '../../tokens/global/index.js';
import { size } from '../../tokens/global/size.js';

var labelTextSize = {
  top: {
    xsmall: 'small',
    small: 'small',
    medium: 'small',
    large: 'medium'
  },
  left: {
    xsmall: 'small',
    small: 'small',
    medium: 'medium',
    large: 'large'
  }
};
var labelOptionalIndicatorTextSize = {
  xsmall: 'small',
  small: 'small',
  medium: 'small',
  large: 'medium'
};
var hintTextSize = {
  xsmall: 'small',
  small: 'small',
  medium: 'small',
  large: 'medium'
};
var hintIconSize = {
  xsmall: 'small',
  small: 'small',
  medium: 'small',
  large: 'medium'
};
var hintMarginTop = {
  xsmall: 'spacing.2',
  small: 'spacing.2',
  medium: 'spacing.2',
  large: 'spacing.3'
};
var labelMarginBottom = {
  xsmall: 'spacing.2',
  small: 'spacing.2',
  medium: 'spacing.2',
  large: 'spacing.3'
};
var labelWidth = {
  xsmall: size[120],
  small: size[120],
  medium: size[120],
  large: size[176]
};
var labelLeftMarginRight = {
  xsmall: 'spacing.3',
  small: 'spacing.3',
  medium: 'spacing.4',
  large: 'spacing.5'
};

export { hintIconSize, hintMarginTop, hintTextSize, labelLeftMarginRight, labelMarginBottom, labelOptionalIndicatorTextSize, labelTextSize, labelWidth };
