// @razorpay/blade/build/lib/web/development/components/Link/Link/Link.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default from 'react';
import '../BaseLink/index.js';
import '../../Box/styledProps/index.js';
import '../../../utils/assignWithoutSideEffects/index.js';
import '../../../utils/makeAnalyticsAttribute/index.js';
import { jsx } from 'react/jsx-runtime';
import BaseLink from '../BaseLink/BaseLink.js';
import { getStyledProps } from '../../Box/styledProps/getStyledProps.js';
import { makeAnalyticsAttribute } from '../../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { assignWithoutSideEffects } from '../../../utils/assignWithoutSideEffects/assignWithoutSideEffects.js';

var _excluded = ["children", "icon", "iconPosition", "isDisabled", "onClick", "variant", "color", "href", "target", "rel", "accessibilityLabel", "size", "testID", "hitSlop", "htmlTitle", "onBlur", "onFocus", "onMouseLeave", "onMouseMove", "onPointerDown", "onPointerEnter", "onTouchStart", "onTouchEnd", "onMouseDown", "onMouseUp"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _Link = function _Link(_ref, ref) {
  var children = _ref.children,
    icon = _ref.icon,
    _ref$iconPosition = _ref.iconPosition,
    iconPosition = _ref$iconPosition === void 0 ? 'left' : _ref$iconPosition,
    _ref$isDisabled = _ref.isDisabled,
    isDisabled = _ref$isDisabled === void 0 ? false : _ref$isDisabled,
    onClick = _ref.onClick,
    _ref$variant = _ref.variant,
    variant = _ref$variant === void 0 ? 'anchor' : _ref$variant,
    _ref$color = _ref.color,
    color = _ref$color === void 0 ? 'primary' : _ref$color,
    href = _ref.href,
    target = _ref.target,
    rel = _ref.rel,
    accessibilityLabel = _ref.accessibilityLabel,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'medium' : _ref$size,
    testID = _ref.testID,
    hitSlop = _ref.hitSlop,
    htmlTitle = _ref.htmlTitle,
    onBlur = _ref.onBlur,
    onFocus = _ref.onFocus,
    onMouseLeave = _ref.onMouseLeave,
    onMouseMove = _ref.onMouseMove,
    onPointerDown = _ref.onPointerDown,
    onPointerEnter = _ref.onPointerEnter,
    onTouchStart = _ref.onTouchStart,
    onTouchEnd = _ref.onTouchEnd,
    onMouseDown = _ref.onMouseDown,
    onMouseUp = _ref.onMouseUp,
    rest = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/jsx(BaseLink, _objectSpread(_objectSpread(_objectSpread(_objectSpread({}, icon ? {
    icon: icon,
    children: children
  } : {
    children: children
  }), variant === 'anchor' ? {
    variant: variant,
    href: href,
    target: target,
    rel: rel
  } : {
    variant: variant,
    isDisabled: isDisabled
  }), {}, {
    ref: ref,
    iconPosition: iconPosition,
    onClick: onClick,
    accessibilityProps: {
      label: accessibilityLabel,
      describedBy: rest['aria-describedby']
    },
    size: size,
    color: color,
    testID: testID,
    hitSlop: hitSlop,
    htmlTitle: htmlTitle,
    onBlur: onBlur,
    onFocus: onFocus,
    onMouseLeave: onMouseLeave,
    onMouseMove: onMouseMove,
    onPointerDown: onPointerDown,
    onPointerEnter: onPointerEnter,
    onTouchStart: onTouchStart,
    onTouchEnd: onTouchEnd,
    onMouseDown: onMouseDown,
    onMouseUp: onMouseUp
  }, getStyledProps(rest)), makeAnalyticsAttribute(rest)));
};
var Link = /*#__PURE__*/assignWithoutSideEffects(/*#__PURE__*/React__default.forwardRef(_Link), {
  displayName: 'Link',
  componentId: 'Link'
});

export { Link as default };
