// @razorpay/blade/build/lib/web/production/components/List/listTokens.js
import '../../utils/makeSize/index.js';
import '../../tokens/global/index.js';
import { makeSize } from '../../utils/makeSize/makeSize.js';
import { size } from '../../tokens/global/size.js';

var listItemMarginLeft = {
  unordered: {
    1: 'spacing.2',
    2: 'spacing.5',
    3: 'spacing.5'
  },
  unorderedWithIcon: {
    1: 'spacing.0',
    2: 'spacing.5',
    3: 'spacing.5'
  },
  ordered: {
    1: 'spacing.0',
    2: 'spacing.6',
    3: 'spacing.6'
  },
  'ordered-filled': {
    1: 'spacing.0',
    2: 'spacing.6',
    3: 'spacing.6'
  }
};
var listItemBulletMarginRight = {
  unordered: 'spacing.3',
  ordered: 'spacing.2',
  'ordered-filled': 'spacing.3'
};
var listItemBulletMarginTop = {
  // We need hard-coded non-tokenized spacing for bullet alignment in List
  // These size mappings can be found in here:
  // figma.com/file/jubmQL9Z8V7881ayUD95ps/Blade---Payment-Light?node-id=23239-466987&t=easqqrPtzgMwCEtn-0
  unordered: {
    onDesktop: {
      small: makeSize(6),
      medium: makeSize(7),
      large: makeSize(10)
    },
    onMobile: {
      small: makeSize(6),
      medium: makeSize(7),
      large: makeSize(10)
    }
  },
  unorderedWithIcon: {
    onDesktop: {
      small: 'spacing.2',
      medium: 'spacing.1',
      large: makeSize(5)
    },
    onMobile: {
      small: 'spacing.2',
      medium: 'spacing.1',
      large: makeSize(5)
    }
  },
  ordered: {
    onDesktop: {
      small: makeSize(0),
      medium: makeSize(0),
      large: makeSize(0)
    },
    onMobile: {
      small: makeSize(0),
      medium: makeSize(0),
      large: makeSize(0)
    }
  },
  'ordered-filled': {
    // TODO: fix these
    onDesktop: {
      small: makeSize(0),
      medium: makeSize(0),
      large: makeSize(0)
    },
    onMobile: {
      small: makeSize(0),
      medium: makeSize(0),
      large: makeSize(0)
    }
  }
};
var listItemUnorderedBulletSize = {
  onDesktop: {
    small: makeSize(size[6]),
    medium: makeSize(size[6]),
    large: makeSize(size[6])
  },
  onMobile: {
    small: makeSize(size[5]),
    medium: makeSize(size[6]),
    large: makeSize(size[6])
  }
};
var listItemMarginBottom = 'spacing.3';
var listItemOrderedBulletBoxSize = {
  // We need hard-coded non-tokenized box sizes for bullet alignment in List
  ordered: {
    onDesktop: {
      small: makeSize(size[16]),
      medium: makeSize(size[20]),
      large: makeSize(size[24])
    },
    onMobile: {
      small: makeSize(size[16]),
      medium: makeSize(size[24]),
      large: makeSize(size[24])
    }
  },
  'ordered-filled': {
    onDesktop: {
      small: makeSize(size[18]),
      medium: makeSize(size[20]),
      large: makeSize(size[24])
    },
    onMobile: {
      small: makeSize(size[16]),
      medium: makeSize(size[20]),
      large: makeSize(size[24])
    }
  }
};
var listItemCodeSize = {
  small: 'small',
  medium: 'small',
  large: 'medium'
};

export { listItemBulletMarginRight, listItemBulletMarginTop, listItemCodeSize, listItemMarginBottom, listItemMarginLeft, listItemOrderedBulletBoxSize, listItemUnorderedBulletSize };
