// @razorpay/blade/build/lib/web/production/components/CounterInput/token.js
import '../../tokens/global/index.js';
import { size } from '../../tokens/global/size.js';
import { spacing } from '../../tokens/global/spacing.js';

var COUNTER_INPUT_TOKEN = {
  width: {
    xsmall: size[78],
    medium: size[94],
    large: size[122]
  },
  height: {
    xsmall: size[30],
    medium: size[38],
    large: size[50]
  },
  decrementIconMargin: [spacing[2], spacing[0], spacing[2], spacing[2]],
  incrementIconMargin: [spacing[2], spacing[2], spacing[2], spacing[0]],
  iconPadding: {
    xsmall: spacing[2],
    medium: spacing[2],
    large: spacing[3]
  },
  emphasis: {
    subtle: {
      color: 'surface.text.gray.subtle',
      disabledColor: 'surface.text.gray.disabled',
      iconColor: 'interactive.icon.gray.subtle',
      disabledIconColor: 'interactive.icon.gray.disabled',
      borderColor: 'interactive.border.gray.default',
      disabledBorderColor: 'interactive.border.gray.default',
      backgroundColor: 'surface.background.gray.intense',
      loadingOrDisabledBgColor: 'surface.background.gray.subtle',
      progressBarColor: 'neutral'
    },
    intense: {
      color: 'interactive.text.primary.subtle',
      disabledColor: 'interactive.text.primary.disabled',
      iconColor: 'interactive.icon.primary.subtle',
      disabledIconColor: 'interactive.icon.primary.disabled',
      borderColor: 'interactive.border.primary.highlighted',
      disabledBorderColor: 'interactive.border.primary.disabled',
      backgroundColor: 'surface.background.gray.intense',
      loadingOrDisabledBgColor: 'surface.background.gray.subtle',
      progressBarColor: undefined
    }
  }
};

export { COUNTER_INPUT_TOKEN };
