// @razorpay/blade/build/lib/web/production/components/Spinner/Spinner/Spinner.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default from 'react';
import '../BaseSpinner/index.js';
import { jsx } from 'react/jsx-runtime';
import { BaseSpinner } from '../BaseSpinner/BaseSpinner.js';

var _excluded = ["label", "labelPosition", "accessibilityLabel", "color", "size", "testID"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _Spinner = function _Spinner(_ref, ref) {
  var _ref2;
  var label = _ref.label,
    labelPosition = _ref.labelPosition,
    accessibilityLabel = _ref.accessibilityLabel,
    _ref$color = _ref.color,
    color = _ref$color === void 0 ? 'neutral' : _ref$color,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'medium' : _ref$size,
    testID = _ref.testID,
    styledProps = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/jsx(BaseSpinner, _objectSpread({
    ref: ref,
    label: label,
    labelPosition: labelPosition,
    accessibilityLabel: (_ref2 = accessibilityLabel !== null && accessibilityLabel !== void 0 ? accessibilityLabel : label) !== null && _ref2 !== void 0 ? _ref2 : 'Loading',
    color: color,
    size: size,
    testID: testID
  }, styledProps));
};
var Spinner = /*#__PURE__*/React__default.forwardRef(_Spinner);

export { Spinner };
