// @razorpay/blade/build/lib/web/production/components/Modal/ModalFooter.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import React__default from 'react';
import { componentIds } from './constants.js';
import { BaseFooter } from '../BaseHeaderFooter/BaseFooter.js';
import '../Box/index.js';
import '../../utils/metaAttribute/index.js';
import '../../utils/assignWithoutSideEffects/index.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import { jsx } from 'react/jsx-runtime';
import { Box } from '../Box/Box.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { assignWithoutSideEffects } from '../../utils/assignWithoutSideEffects/assignWithoutSideEffects.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _ModalFooter = function _ModalFooter(props) {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  var containerRef = React__default.useRef(null);
  return /*#__PURE__*/jsx(Box, {
    ref: containerRef,
    children: /*#__PURE__*/jsx(BaseFooter, _objectSpread({
      metaComponentName: MetaConstants.ModalFooter,
      children: props.children,
      showDivider: true
    }, makeAnalyticsAttribute(props)))
  });
};
var ModalFooter = /*#__PURE__*/assignWithoutSideEffects(_ModalFooter, {
  componentId: componentIds.ModalFooter
});

export { ModalFooter };
