// @razorpay/blade/build/lib/web/production/components/Modal/ModalBackdrop.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import 'react';
import styled from 'styled-components';
import { FloatingOverlay } from '@floating-ui/react';
import { useModalContext } from './ModalContext.js';
import '../../utils/index.js';
import '../../utils/metaAttribute/index.js';
import { jsx } from 'react/jsx-runtime';
import { makeMotionTime } from '../../utils/makeMotionTime/makeMotionTime.web.js';
import { castWebType } from '../../utils/platform/castUtils.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StyledModalBackdrop = /*#__PURE__*/styled(FloatingOverlay).withConfig({
  displayName: "ModalBackdrop__StyledModalBackdrop",
  componentId: "sc-69uedf-0"
})(function (_ref) {
  var theme = _ref.theme;
  var _useModalContext = useModalContext(),
    isVisible = _useModalContext.isVisible;
  return {
    transitionDuration: "".concat(makeMotionTime(theme.motion.duration.moderate)),
    transitionTimingFunction: isVisible ? castWebType(theme.motion.easing.entrance) : castWebType(theme.motion.easing.exit),
    transitionProperty: 'opacity',
    opacity: isVisible ? 1 : 0,
    backgroundColor: theme.colors.overlay.background.subtle
  };
});
var ModalBackdrop = function ModalBackdrop() {
  var _useModalContext2 = useModalContext(),
    close = _useModalContext2.close,
    isDismissible = _useModalContext2.isDismissible;
  return /*#__PURE__*/jsx(StyledModalBackdrop, _objectSpread(_objectSpread({}, metaAttribute({
    name: MetaConstants.ModalBackdrop,
    testID: MetaConstants.ModalBackdrop
  })), {}, {
    onClick: function onClick() {
      if (isDismissible) {
        close === null || close === void 0 || close();
      }
    },
    lockScroll: true
  }));
};

export { ModalBackdrop };
