// @razorpay/blade/build/lib/web/production/components/Modal/Modal.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default, { useEffect } from 'react';
import styled, { css } from 'styled-components';
import { useFloating, FloatingPortal, FloatingFocusManager } from '@floating-ui/react';
import usePresence from 'use-presence';
export { ModalHeader } from './ModalHeader.web.js';
export { ModalFooter } from './ModalFooter.web.js';
export { ModalBody } from './ModalBody.web.js';
import { ModalContext } from './ModalContext.js';
import { ModalBackdrop } from './ModalBackdrop.js';
import { modalMargin, modalMaxWidth, modalMinWidth, modalMaxHeight, modalResponsiveScreenGap, modalBorderRadius } from './modalTokens.js';
import '../../utils/index.js';
import '../Box/BaseBox/index.js';
import '../BladeProvider/index.js';
import '../Box/index.js';
import '../../utils/metaAttribute/index.js';
import '../../utils/makeAccessible/index.js';
import '../../utils/logger/index.js';
import { componentZIndices } from '../../utils/componentZIndices.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { castWebType } from '../../utils/platform/castUtils.js';
import { makeMotionTime } from '../../utils/makeMotionTime/makeMotionTime.web.js';
import { makeSize } from '../../utils/makeSize/makeSize.js';
import useTheme from '../BladeProvider/useTheme.js';
import { logger } from '../../utils/logger/logger.js';
import { Box } from '../Box/Box.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { makeAccessible } from '../../utils/makeAccessible/makeAccessible.web.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';

var _excluded = ["isOpen", "children", "onDismiss", "isDismissible", "initialFocusRef", "size", "accessibilityLabel", "zIndex"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var ModalContent = /*#__PURE__*/styled(BaseBox).withConfig({
  displayName: "Modalweb__ModalContent",
  componentId: "sc-107gfd9-0"
})(function (_ref) {
  var isVisible = _ref.isVisible,
    theme = _ref.theme,
    size = _ref.size;
  var scale = isVisible ? 1 : 0.9;
  var transform = size !== 'full' ? "translate(-50%, -50%) scale(".concat(scale, ")") : "";
  return css(["box-shadow:", ";opacity:", ";position:fixed;transform:", ";transition-property:opacity,transform;transition-duration:", ";transition-timing-function:", ";", ""], theme.elevation.highRaised, isVisible ? 1 : 0, transform, castWebType(makeMotionTime(theme.motion.duration.moderate)), isVisible ? castWebType(theme.motion.easing.entrance) : castWebType(theme.motion.easing.exit), size === 'full' && css(["top:", ";left:", ";right:", ";bottom:", ";"], makeSize(modalMargin[size]), makeSize(modalMargin[size]), makeSize(modalMargin[size]), makeSize(modalMargin[size])));
});
var Modal = function Modal(_ref2) {
  var _ref2$isOpen = _ref2.isOpen,
    isOpen = _ref2$isOpen === void 0 ? false : _ref2$isOpen,
    children = _ref2.children,
    onDismiss = _ref2.onDismiss,
    _ref2$isDismissible = _ref2.isDismissible,
    isDismissible = _ref2$isDismissible === void 0 ? true : _ref2$isDismissible,
    initialFocusRef = _ref2.initialFocusRef,
    _ref2$size = _ref2.size,
    size = _ref2$size === void 0 ? 'small' : _ref2$size,
    accessibilityLabel = _ref2.accessibilityLabel,
    _ref2$zIndex = _ref2.zIndex,
    zIndex = _ref2$zIndex === void 0 ? componentZIndices.modal : _ref2$zIndex,
    rest = _objectWithoutProperties(_ref2, _excluded);
  var _useTheme = useTheme(),
    theme = _useTheme.theme,
    platform = _useTheme.platform;
  var _usePresence = usePresence(isOpen, {
      transitionDuration: theme.motion.duration.moderate,
      initialEnter: true
    }),
    isMounted = _usePresence.isMounted,
    isVisible = _usePresence.isVisible;

  // Warn consumer if modal is opened on mobile
  useEffect(function () {
    if (false) {
      if (platform === 'onMobile') {
        logger({
          message: 'Modal is not supported on mobile devices. Please use BottomSheet instead.',
          moduleName: 'Modal',
          type: 'warn'
        });
      }
    }
  }, [platform]);

  // required by floating ui to handle focus
  var _useFloating = useFloating({
      open: isMounted
    }),
    refs = _useFloating.refs,
    context = _useFloating.context;
  var defaultInitialFocusRef = React__default.useRef(null);
  var modalContext = React__default.useMemo(function () {
    return {
      isOpen: isOpen,
      close: onDismiss,
      isDismissible: isDismissible,
      defaultInitialFocusRef: defaultInitialFocusRef,
      isVisible: isVisible
    };
  }, [isOpen, onDismiss, isDismissible, defaultInitialFocusRef, isVisible]);
  var handleKeyDown = function handleKeyDown(event) {
    // close modal on escape key press only if dismissible
    if (isDismissible && ((event === null || event === void 0 ? void 0 : event.key) === 'Escape' || (event === null || event === void 0 ? void 0 : event.code) === 'Escape')) {
      onDismiss === null || onDismiss === void 0 || onDismiss();
    }
  };
  return /*#__PURE__*/jsx(FloatingPortal, {
    children: /*#__PURE__*/jsx(ModalContext.Provider, {
      value: modalContext,
      children: isMounted ? /*#__PURE__*/jsx(FloatingFocusManager, {
        returnFocus: true,
        initialFocus: initialFocusRef !== null && initialFocusRef !== void 0 ? initialFocusRef : defaultInitialFocusRef,
        context: context,
        modal: true,
        children: /*#__PURE__*/jsxs(Box, {
          zIndex: zIndex,
          position: "fixed",
          testID: "modal-wrapper",
          children: [/*#__PURE__*/jsx(ModalBackdrop, {}), /*#__PURE__*/jsx(ModalContent, _objectSpread(_objectSpread(_objectSpread(_objectSpread({}, metaAttribute({
            name: MetaConstants.Modal
          })), makeAccessible({
            role: 'dialog',
            modal: true,
            label: accessibilityLabel
          })), makeAnalyticsAttribute(rest)), {}, {
            maxWidth: size === 'full' ? '100%' : makeSize(modalMaxWidth[size]),
            minWidth: makeSize(modalMinWidth),
            maxHeight: modalMaxHeight[size],
            width: size === 'full' ? "calc(100vw - ".concat(makeSize(modalMargin[size] * 2), ")") : "calc(100vw - ".concat(makeSize(modalResponsiveScreenGap), ")"),
            borderRadius: modalBorderRadius,
            backgroundColor: "popup.background.subtle",
            display: "flex",
            flexDirection: "column",
            top: "50%",
            left: "50%",
            onKeyDown: handleKeyDown,
            isVisible: isVisible,
            size: size,
            ref: refs.setFloating,
            overflow: "hidden",
            children: children
          }))]
        })
      }) : null
    })
  });
};

export { Modal };
