// @razorpay/blade/build/lib/web/production/components/FilterChip/BaseFilterChip.web.js
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import React__default from 'react';
import styled from 'styled-components';
import { FILTER_CHIP_HEIGHT } from './tokens.js';
import '../Box/index.js';
import '../Box/BaseBox/index.js';
import '../Counter/index.js';
import '../Divider/index.js';
import '../Icons/index.js';
import '../Typography/index.js';
import '../../utils/index.js';
import '../../utils/getFocusRingStyles/index.js';
import '../Box/styledProps/index.js';
import '../../utils/assignWithoutSideEffects/index.js';
import '../../utils/metaAttribute/index.js';
import '../../utils/makeAccessible/index.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { getFocusRingStyles } from '../../utils/getFocusRingStyles/getFocusRingStyles.web.js';
import { makeSpace } from '../../utils/makeSpace/makeSpace.js';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { makeBorderSize } from '../../utils/makeBorderSize/makeBorderSize.js';
import { Box } from '../Box/Box.js';
import { Counter } from '../Counter/Counter.js';
import { Text } from '../Typography/Text/Text.js';
import { makeAccessible } from '../../utils/makeAccessible/makeAccessible.web.js';
import { getStyledProps } from '../Box/styledProps/getStyledProps.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import ChevronDownIcon from '../Icons/ChevronDownIcon/ChevronDownIcon.js';
import { Divider } from '../Divider/Divider.js';
import CloseIcon from '../Icons/CloseIcon/CloseIcon.js';
import { assignWithoutSideEffects } from '../../utils/assignWithoutSideEffects/assignWithoutSideEffects.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';

var _excluded = ["value", "onClearButtonClick", "label", "isDisabled", "selectionType", "onClick", "onKeyDown", "accessibilityProps", "id"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var getInteractiveFilterItemStyles = function getInteractiveFilterItemStyles(_ref) {
  var theme = _ref.theme;
  return {
    display: 'flex',
    height: '100%',
    alignItems: 'center',
    border: 'none',
    cursor: 'pointer',
    color: 'currentcolor',
    '&:not([disabled]):hover': {
      backgroundColor: theme.colors.interactive.background.gray.faded
    },
    '&[disabled]': {
      cursor: 'not-allowed'
    },
    '&:focus-visible': _objectSpread(_objectSpread({}, getFocusRingStyles({
      theme: theme
    })), {}, {
      outlineOffset: makeSpace(theme.spacing[1])
    })
  };
};
var StyledFilterChip = /*#__PURE__*/styled(BaseBox).withConfig({
  displayName: "BaseFilterChipweb__StyledFilterChip",
  componentId: "sc-7fl0tf-0"
})(function (_ref2) {
  var theme = _ref2.theme,
    $isDisabled = _ref2.$isDisabled;
  return {
    borderWidth: makeBorderSize(theme.border.width.thinner),
    borderColor: theme.colors.interactive.border.gray[$isDisabled ? 'disabled' : 'highlighted'],
    height: FILTER_CHIP_HEIGHT,
    borderRadius: theme.border.radius.max,
    display: 'flex',
    borderStyle: 'solid',
    backgroundColor: theme.colors.surface.background.gray.intense,
    color: theme.colors.interactive.text.gray[$isDisabled ? 'disabled' : 'normal'],
    width: 'fit-content'
  };
});
var StyledFilterTrigger = /*#__PURE__*/styled.button.withConfig({
  displayName: "BaseFilterChipweb__StyledFilterTrigger",
  componentId: "sc-7fl0tf-1"
})(function (_ref3) {
  var theme = _ref3.theme,
    $isSelected = _ref3.$isSelected;
  var spacing = theme.spacing;
  return _objectSpread({
    backgroundColor: theme.colors.transparent,
    borderRadius: $isSelected ? theme.border.radius.none : theme.border.radius.max,
    borderTopLeftRadius: theme.border.radius.max,
    borderBottomLeftRadius: theme.border.radius.max,
    paddingLeft: makeSpace(spacing[4]),
    paddingRight: $isSelected ? makeSpace(spacing[2]) : makeSpace(spacing[3]),
    gap: makeSpace(spacing[2])
  }, getInteractiveFilterItemStyles({
    theme: theme
  }));
});
var StyledFilterCloseButton = /*#__PURE__*/styled.button.withConfig({
  displayName: "BaseFilterChipweb__StyledFilterCloseButton",
  componentId: "sc-7fl0tf-2"
})(function (_ref4) {
  var theme = _ref4.theme;
  return _objectSpread({
    backgroundColor: theme.colors.transparent,
    borderTopRightRadius: theme.border.radius.max,
    borderBottomRightRadius: theme.border.radius.max,
    paddingLeft: makeSpace(theme.spacing[2]),
    paddingRight: makeSpace(theme.spacing[3]),
    justifyContent: 'center'
  }, getInteractiveFilterItemStyles({
    theme: theme
  }));
});
var renderValue = function renderValue(selectionType, value, isDisabled) {
  if (selectionType === 'multiple' && Array.isArray(value)) {
    return /*#__PURE__*/jsx(Box, {
      display: "flex",
      alignItems: "center",
      children: /*#__PURE__*/jsx(Counter, {
        value: value.length,
        color: isDisabled ? 'neutral' : 'primary',
        size: "small"
      })
    });
  }
  return /*#__PURE__*/jsx(Text, {
    as: "span",
    size: "xsmall",
    weight: "semibold",
    color: isDisabled ? 'interactive.text.gray.disabled' : 'interactive.text.primary.normal',
    children: value
  });
};
var _BaseFilterChip = function _BaseFilterChip(_ref5, ref) {
  var _accessibilityProps$r;
  var value = _ref5.value,
    onClearButtonClick = _ref5.onClearButtonClick,
    label = _ref5.label,
    isDisabled = _ref5.isDisabled,
    _ref5$selectionType = _ref5.selectionType,
    selectionType = _ref5$selectionType === void 0 ? 'single' : _ref5$selectionType,
    _onClick = _ref5.onClick,
    _onKeyDown = _ref5.onKeyDown,
    accessibilityProps = _ref5.accessibilityProps,
    id = _ref5.id,
    rest = _objectWithoutProperties(_ref5, _excluded);
  var isSelected = selectionType === 'multiple' ? Array.isArray(value) && value.length > 0 : !!value;
  return /*#__PURE__*/jsxs(StyledFilterChip, {
    $isDisabled: isDisabled,
    $isSelected: isSelected,
    ref: ref,
    children: [/*#__PURE__*/jsxs(StyledFilterTrigger, _objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread({
      $isSelected: isSelected,
      disabled: isDisabled,
      id: id,
      onClick: function onClick(e) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        _onClick === null || _onClick === void 0 || _onClick(e);
      },
      onKeyDown: function onKeyDown(e) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        _onKeyDown === null || _onKeyDown === void 0 || _onKeyDown(e);
      }
    }, makeAccessible(_objectSpread(_objectSpread({}, accessibilityProps), {}, {
      role: (_accessibilityProps$r = accessibilityProps === null || accessibilityProps === void 0 ? void 0 : accessibilityProps.role) !== null && _accessibilityProps$r !== void 0 ? _accessibilityProps$r : 'button'
    }))), getStyledProps(rest)), makeAnalyticsAttribute(rest)), metaAttribute({
      name: 'filter-chip-trigger',
      testID: rest.testID
    })), {}, {
      children: [/*#__PURE__*/jsxs(Box, {
        display: "flex",
        gap: "spacing.2",
        whiteSpace: "nowrap",
        children: [/*#__PURE__*/jsxs(Text, {
          size: "xsmall",
          weight: "medium",
          color: "interactive.text.gray.subtle",
          truncateAfterLines: 1,
          children: [label, isSelected ? ':' : null]
        }), isSelected ? renderValue(selectionType, value, isDisabled) : null]
      }), /*#__PURE__*/jsx(Box, {
        display: "flex",
        alignItems: "center",
        paddingRight: "spacing.1",
        children: /*#__PURE__*/jsx(ChevronDownIcon, {
          size: "small",
          color: "interactive.icon.gray.muted"
        })
      })]
    })), isSelected ? /*#__PURE__*/jsxs(Fragment, {
      children: [/*#__PURE__*/jsx(Divider, {
        orientation: "vertical",
        variant: isDisabled ? 'muted' : 'subtle'
      }), /*#__PURE__*/jsx(StyledFilterCloseButton, _objectSpread(_objectSpread({
        "aria-label": "Clear ".concat(label, " value")
        // value can never be undefined because when it's undefined the button itself doesn't render/
        ,
        onClick: function onClick() {
          return onClearButtonClick === null || onClearButtonClick === void 0 ? void 0 : onClearButtonClick({
            value: value !== null && value !== void 0 ? value : ''
          });
        },
        disabled: isDisabled
      }, metaAttribute({
        name: 'filter-chip-close-button'
      })), {}, {
        children: /*#__PURE__*/jsx(CloseIcon, {
          size: "small",
          color: "interactive.icon.gray.muted"
        })
      }))]
    }) : null]
  });
};
var BaseFilterChip = /*#__PURE__*/assignWithoutSideEffects(/*#__PURE__*/React__default.forwardRef(_BaseFilterChip), {
  componentId: MetaConstants.BaseFilterChip
});

export { BaseFilterChip };
