// @razorpay/blade/build/lib/web/production/components/Alert/styles.js
import '../../utils/makeSpace/index.js';
import '../../utils/makeSize/index.js';
import '../../utils/makeBorderSize/index.js';
import '../../tokens/global/index.js';
import { size } from '../../tokens/global/size.js';
import { makeSpace } from '../../utils/makeSpace/makeSpace.js';
import { makeBorderSize } from '../../utils/makeBorderSize/makeBorderSize.js';
import { makeSize } from '../../utils/makeSize/makeSize.js';

var MAX_WIDTH = size[584];
var getCommonStyles = function getCommonStyles(props) {
  var theme = props.theme,
    emphasis = props.emphasis,
    color = props.color,
    isFullWidth = props.isFullWidth,
    isDesktop = props.isDesktop;
  return {
    background: theme.colors.feedback.background[color][emphasis],
    padding: makeSpace(theme.spacing[4]),
    borderRadius: makeBorderSize(isFullWidth ? theme.border.radius.none : theme.border.radius.medium),
    display: 'flex',
    flexDirection: 'row',
    maxWidth: isFullWidth ? 'auto' : makeSize(MAX_WIDTH),
    width: isFullWidth ? '100%' : undefined,
    alignItems: isFullWidth && isDesktop ? 'center' : 'flex-start',
    boxSizing: 'border-box'
  };
};

export { getCommonStyles };
