// @razorpay/blade/build/lib/web/development/components/TopNav/TopNav.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default from 'react';
import '../Box/BaseBox/index.js';
import { SIDE_NAV_EXPANDED_L1_WIDTH_BASE, SIDE_NAV_EXPANDED_L1_WIDTH_XL } from '../SideNav/tokens.js';
import '../../tokens/global/index.js';
import '../../utils/index.js';
import '../../utils/metaAttribute/index.js';
import { componentZIndices } from '../../utils/componentZIndices.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import { jsx } from 'react/jsx-runtime';
import { size } from '../../tokens/global/size.js';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { makeSize } from '../../utils/makeSize/makeSize.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';

var _excluded = ["children"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var TOP_NAV_HEIGHT = size[56];
var CONTENT_RIGHT_GAP = size[80];
var _TopNav = function _TopNav(_ref, ref) {
  var children = _ref.children,
    rest = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread(_objectSpread(_objectSpread({
    ref: ref,
    display: "grid",
    gridTemplateColumns: "auto minmax(0, 1fr) auto",
    alignItems: "center",
    position: "sticky",
    top: "0px",
    width: "100%",
    paddingY: {
      base: 'spacing.3',
      m: 'spacing.0'
    },
    paddingX: {
      base: 'spacing.4',
      m: 'spacing.3'
    },
    height: makeSize(TOP_NAV_HEIGHT),
    zIndex: componentZIndices.topnav
  }, rest), metaAttribute({
    name: MetaConstants.TopNav,
    testID: rest.testID
  })), makeAnalyticsAttribute(rest)), {}, {
    children: children
  }));
};
var TopNav = /*#__PURE__*/React__default.forwardRef(_TopNav);
var TopNavBrand = function TopNavBrand(_ref2) {
  var children = _ref2.children;
  return /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread({
    flexDirection: "row",
    marginTop: "spacing.4",
    width: {
      base: makeSize(SIDE_NAV_EXPANDED_L1_WIDTH_BASE),
      xl: makeSize(SIDE_NAV_EXPANDED_L1_WIDTH_XL)
    }
  }, metaAttribute({
    name: MetaConstants.TopNavBrand
  })), {}, {
    children: /*#__PURE__*/jsx(BaseBox, {
      width: "100%",
      textAlign: "center",
      children: children
    })
  }));
};
var TopNavContent = function TopNavContent(_ref3) {
  var children = _ref3.children;
  return /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread({
    display: "flex",
    alignItems: "center",
    alignSelf: "end",
    paddingRight: {
      base: 'spacing.0',
      m: makeSize(CONTENT_RIGHT_GAP)
    }
  }, metaAttribute({
    name: MetaConstants.TopNavContent
  })), {}, {
    children: children
  }));
};
var TopNavActions = function TopNavActions(_ref4) {
  var children = _ref4.children;
  return /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread({
    alignSelf: "end",
    display: "flex",
    gap: "spacing.3",
    alignItems: "center",
    marginTop: "spacing.1",
    padding: "spacing.3",
    borderTopLeftRadius: "medium",
    borderTopRightRadius: "medium"
  }, metaAttribute({
    name: MetaConstants.TopNavActions
  })), {}, {
    children: children
  }));
};

export { TopNav, TopNavActions, TopNavBrand, TopNavContent };
