// @razorpay/blade/build/lib/web/production/components/BottomNav/BottomNav.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default from 'react';
import styled from 'styled-components';
import '../Box/BaseBox/index.js';
import '../Box/styledProps/index.js';
import '../Typography/index.js';
import '../../utils/index.js';
import { componentZIndices } from '../../utils/componentZIndices.js';
import '../../utils/getFocusRingStyles/index.js';
import '../../utils/logger/index.js';
import '../../utils/makeAccessible/index.js';
import '../../utils/metaAttribute/index.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import { throwBladeError } from '../../utils/logger/logger.js';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { getStyledProps } from '../Box/styledProps/getStyledProps.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { makeSpace } from '../../utils/makeSpace/makeSpace.js';
import { makeMotionTime } from '../../utils/makeMotionTime/makeMotionTime.web.js';
import { getFocusRingStyles } from '../../utils/getFocusRingStyles/getFocusRingStyles.web.js';
import { makeAccessible } from '../../utils/makeAccessible/makeAccessible.web.js';
import { Text } from '../Typography/Text/Text.js';

var _excluded = ["children", "zIndex", "testID"],
  _excluded2 = ["title", "href", "rel", "target", "as", "isActive", "onClick", "icon", "testID"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _BottomNav = function _BottomNav(_ref, ref) {
  var children = _ref.children,
    _ref$zIndex = _ref.zIndex,
    zIndex = _ref$zIndex === void 0 ? componentZIndices.bottomNav : _ref$zIndex,
    testID = _ref.testID,
    rest = _objectWithoutProperties(_ref, _excluded);
  if (false) {
    var childrenCount = React__default.Children.count(children);
    if (childrenCount > 5 && childrenCount < 2) {
      throwBladeError({
        moduleName: 'BottomNav',
        message: 'children cannot be less than 2 and more than 5'
      });
    }
  }
  return /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread(_objectSpread(_objectSpread({
    ref: ref,
    role: "navigation",
    position: "fixed",
    bottom: "spacing.0",
    left: "spacing.0",
    elevation: "midRaised",
    width: "100%",
    backgroundColor: "surface.background.gray.intense",
    borderTopWidth: "thin",
    borderTopColor: "surface.border.gray.muted",
    paddingX: "spacing.2",
    display: "flex",
    flexDirection: "row"
  }, getStyledProps(rest)), {}, {
    zIndex: zIndex
  }, metaAttribute({
    testID: testID,
    name: MetaConstants.BottomNav
  })), makeAnalyticsAttribute(rest)), {}, {
    children: children
  }));
};
var StyledBottomNavItem = /*#__PURE__*/styled(BaseBox).withConfig({
  displayName: "BottomNavweb__StyledBottomNavItem",
  componentId: "sc-1ve2e40-0"
})(function (props) {
  return {
    textDecoration: 'none',
    color: props.theme.colors.interactive.text.gray.subtle,
    backgroundColor: 'transparent',
    border: 'none',
    paddingLeft: makeSpace(props.theme.spacing[0]),
    paddingRight: makeSpace(props.theme.spacing[0]),
    transition: "color ".concat(makeMotionTime(props.theme.motion.duration['2xquick']), " ").concat(props.theme.motion.easing.standard),
    '&[aria-current="page"]': {
      color: props.theme.colors.interactive.text.primary.subtle
    },
    '&:focus-visible': _objectSpread({}, getFocusRingStyles({
      theme: props.theme
    }))
  };
});
var BottomNavItem = function BottomNavItem(_ref2) {
  var title = _ref2.title,
    href = _ref2.href,
    rel = _ref2.rel,
    target = _ref2.target,
    as = _ref2.as,
    isActive = _ref2.isActive,
    onClick = _ref2.onClick,
    Icon = _ref2.icon,
    testID = _ref2.testID,
    rest = _objectWithoutProperties(_ref2, _excluded2);
  var isRouterLink = as && href;
  var defaultRenderElement = href ? 'a' : 'button';
  var defaultRel = target === '_blank' ? 'noreferrer noopener' : undefined;
  return /*#__PURE__*/jsxs(StyledBottomNavItem, _objectSpread(_objectSpread(_objectSpread(_objectSpread({
    as: isRouterLink ? as : defaultRenderElement,
    href: as ? undefined : href,
    to: href // for react router
    ,
    rel: rel !== null && rel !== void 0 ? rel : defaultRel,
    target: target,
    paddingTop: "spacing.5",
    paddingBottom: "spacing.4",
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    flexDirection: "column",
    flex: "1",
    cursor: "pointer",
    onClick: onClick,
    gap: "spacing.1"
  }, makeAccessible({
    current: isActive ? 'page' : undefined
  })), metaAttribute({
    name: MetaConstants.BottomNavItem,
    testID: testID
  })), makeAnalyticsAttribute(rest)), {}, {
    children: [/*#__PURE__*/jsx(Icon, {
      color: "currentColor",
      size: "large"
    }), /*#__PURE__*/jsx(Text, {
      truncateAfterLines: 1,
      color: "currentColor",
      size: "xsmall",
      weight: "semibold",
      children: title
    })]
  }));
};
var BottomNav = /*#__PURE__*/React__default.forwardRef(_BottomNav);

export { BottomNav, BottomNavItem };
