// @razorpay/blade/build/lib/web/production/components/SideNav/SideNavItems/SideNavItem.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import 'react';
import styled from 'styled-components';
import { getNavItemTransition, NAV_ITEM_HEIGHT, classes } from '../tokens.js';
import '../../Box/index.js';
import '../../Box/BaseBox/index.js';
import '../../Typography/index.js';
import '../../../utils/index.js';
import '../../../utils/makeAnalyticsAttribute/index.js';
import { TooltipifyComponent } from '../../../utils/TooltipifyComponent.js';
import '../../../utils/isIconComponent/index.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import { BaseBox } from '../../Box/BaseBox/BaseBox.web.js';
import { isIconComponent } from '../../../utils/isIconComponent/isIconComponent.js';
import { makeSize } from '../../../utils/makeSize/makeSize.js';
import { makeAnalyticsAttribute } from '../../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { Box } from '../../Box/Box.js';
import { Text } from '../../Typography/Text/Text.js';

var _excluded = ["leading", "trailing", "title", "backgroundColor", "tooltip", "as"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var SideNavItemContainer = /*#__PURE__*/styled(BaseBox).withConfig({
  displayName: "SideNavItemweb__SideNavItemContainer",
  componentId: "nb93ri-0"
})(function (props) {
  return {
    transition: getNavItemTransition(props.theme)
  };
});
var SideNavItem = function SideNavItem(_ref) {
  var leading = _ref.leading,
    trailing = _ref.trailing,
    title = _ref.title,
    backgroundColor = _ref.backgroundColor,
    tooltip = _ref.tooltip,
    _ref$as = _ref.as,
    as = _ref$as === void 0 ? 'div' : _ref$as,
    rest = _objectWithoutProperties(_ref, _excluded);
  var isIcon = typeof leading === 'function' && isIconComponent(leading);
  var Icon = isIcon ? leading : undefined;
  return /*#__PURE__*/jsx(TooltipifyComponent, {
    tooltip: tooltip,
    children: /*#__PURE__*/jsxs(SideNavItemContainer, _objectSpread(_objectSpread({
      display: "flex",
      flexDirection: "row",
      justifyContent: "space-between",
      alignItems: "center",
      paddingX: "spacing.3",
      height: makeSize(NAV_ITEM_HEIGHT),
      backgroundColor: backgroundColor,
      borderRadius: "medium",
      as: as,
      cursor: as === 'label' ? 'pointer' : undefined
    }, makeAnalyticsAttribute(rest)), {}, {
      children: [/*#__PURE__*/jsxs(Box, {
        display: "inline-flex",
        alignItems: "center",
        gap: "spacing.3",
        children: [Icon ? /*#__PURE__*/jsx(BaseBox, {
          display: "flex",
          alignItems: "center",
          paddingX: "spacing.2",
          children: /*#__PURE__*/jsx(Icon, {
            size: "medium",
            color: "interactive.icon.gray.subtle"
          })
        }) : leading, /*#__PURE__*/jsx(BaseBox, {
          className: classes.HIDE_WHEN_COLLAPSED,
          children: /*#__PURE__*/jsx(Text, {
            truncateAfterLines: 1,
            weight: "medium",
            size: "medium",
            color: "surface.text.gray.subtle",
            children: title
          })
        })]
      }), /*#__PURE__*/jsx(BaseBox, {
        className: classes.HIDE_WHEN_COLLAPSED,
        children: trailing
      })]
    }))
  });
};

export { SideNavItem };
