// @razorpay/blade/build/lib/web/production/components/FileUpload/fileUploadTokens.js
import '../../tokens/global/index.js';
import { size } from '../../tokens/global/size.js';

var getFileUploadInputHoverTokens = function getFileUploadInputHoverTokens() {
  return {
    "default": {
      background: {
        checked: 'colors.transparent',
        unchecked: 'colors.transparent'
      },
      border: {
        checked: 'colors.interactive.border.gray.default',
        unchecked: 'colors.interactive.border.gray.default'
      }
    }
  };
};
var fileUploadMotionTokens = {
  duration: 'duration.2xquick',
  easing: 'easing.standard'
};
var fileUploadHeightTokens = {
  medium: size['56'],
  large: size['64']
};
var fileUploadColorTokens = {
  text: {
    "default": 'surface.text.gray.subtle',
    disabled: 'surface.text.gray.disabled'
  },
  border: {
    "default": 'interactive.border.gray.default',
    disabled: 'interactive.border.gray.disabled'
  },
  background: {
    hover: 'interactive.background.gray.highlighted',
    active: 'interactive.background.primary.faded'
  },
  icon: {
    "default": 'interactive.icon.primary.subtle',
    disabled: 'interactive.icon.primary.disabled'
  },
  link: {
    "default": 'interactive.text.primary.subtle',
    disabled: 'interactive.text.primary.disabled'
  }
};
var fileUploadItemBackgroundColors = {
  success: {
    "default": 'interactive.background.gray.faded',
    hover: 'interactive.background.gray.fadedHighlighted'
  },
  error: {
    "default": 'feedback.background.negative.subtle',
    hover: 'interactive.background.negative.fadedHighlighted'
  },
  uploading: {
    "default": 'surface.background.gray.intense',
    hover: 'surface.background.gray.intense'
  }
};
var fileUploadLinkBorderTokens = {
  color: {
    "default": 'surface.border.primary.normal',
    disabled: 'surface.border.primary.muted'
  },
  width: {
    "default": 'thin'
  }
};

export { fileUploadColorTokens, fileUploadHeightTokens, fileUploadItemBackgroundColors, fileUploadLinkBorderTokens, fileUploadMotionTokens, getFileUploadInputHoverTokens };
