// @razorpay/blade/build/lib/web/production/components/FileUpload/StyledFileUploadWrapper.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import styled from 'styled-components';
import { fileUploadMotionTokens, fileUploadColorTokens } from './fileUploadTokens.js';
import getIn from '../../utils/lodashButBetter/get.js';
import '../Box/BaseBox/index.js';
import '../../utils/makeMotionTime/index.web.js';
import '../../utils/index.js';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { castWebType } from '../../utils/platform/castUtils.js';
import { makeMotionTime } from '../../utils/makeMotionTime/makeMotionTime.web.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StyledFileUploadWrapper = /*#__PURE__*/styled(BaseBox).withConfig({
  displayName: "StyledFileUploadWrapper",
  componentId: "sc-1wln1sx-0"
})(function (_ref) {
  var theme = _ref.theme,
    isDisabled = _ref.isDisabled,
    isActive = _ref.isActive;
  var easing = getIn(theme.motion, fileUploadMotionTokens.easing);
  var duration = castWebType(makeMotionTime(getIn(theme.motion, fileUploadMotionTokens.duration)));
  return _objectSpread(_objectSpread({
    display: 'flex',
    borderStyle: 'dashed',
    transitionProperty: 'background-color',
    transitionTimingFunction: easing,
    transitionDuration: duration,
    borderColor: getIn(theme.colors, fileUploadColorTokens.border[isDisabled ? 'disabled' : 'default'])
  }, isActive && {
    backgroundColor: getIn(theme.colors, fileUploadColorTokens.background.active)
  }), !isDisabled && !isActive && {
    '&:hover': {
      backgroundColor: getIn(theme.colors, fileUploadColorTokens.background.hover)
    }
  });
});

export { StyledFileUploadWrapper };
