// @razorpay/blade/build/lib/web/production/components/FileUpload/StyledFileUploadItemWrapper.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import styled from 'styled-components';
import { fileUploadMotionTokens, fileUploadHeightTokens, fileUploadItemBackgroundColors } from './fileUploadTokens.js';
import getIn from '../../utils/lodashButBetter/get.js';
import '../Box/BaseBox/index.js';
import '../../utils/makeMotionTime/index.web.js';
import '../../utils/index.js';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { castWebType } from '../../utils/platform/castUtils.js';
import { makeMotionTime } from '../../utils/makeMotionTime/makeMotionTime.web.js';
import { makeSize } from '../../utils/makeSize/makeSize.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StyledFileUploadItemWrapper = /*#__PURE__*/styled(BaseBox).withConfig({
  displayName: "StyledFileUploadItemWrapper",
  componentId: "sc-1rxjypg-0"
})(function (_ref) {
  var theme = _ref.theme,
    status = _ref.status,
    size = _ref.size;
  var easing = getIn(theme.motion, fileUploadMotionTokens.easing);
  var duration = castWebType(makeMotionTime(getIn(theme.motion, fileUploadMotionTokens.duration)));
  return _objectSpread({
    position: 'relative',
    overflow: 'hidden',
    display: 'flex',
    justifyContent: 'space-between',
    borderStyle: 'solid',
    minHeight: makeSize(fileUploadHeightTokens[size === 'variable' ? 'large' : size]),
    width: '100%',
    backgroundColor: getIn(theme.colors, fileUploadItemBackgroundColors[status]["default"]),
    transitionProperty: 'background-color',
    transitionTimingFunction: easing,
    transitionDuration: duration,
    borderColor: theme.colors.interactive.border.neutral.faded,
    wordBreak: 'break-all'
  }, status !== 'uploading' && {
    '&:hover': {
      backgroundColor: getIn(theme.colors, fileUploadItemBackgroundColors[status].hover)
    }
  });
});

export { StyledFileUploadItemWrapper };
