// @razorpay/blade/build/lib/web/production/utils/useColorScheme/useColorScheme.js
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import { useState, useCallback } from 'react';
import '../getColorScheme/index.js';
import { colorSchemeNamesInput } from '../../tokens/theme/theme.js';
import '../logger/index.js';
import { getColorScheme } from '../getColorScheme/getColorScheme.web.js';
import { throwBladeError } from '../logger/logger.js';

var useColorScheme = function useColorScheme() {
  var initialColorScheme = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'light';
  // if colorScheme defined use that else fallback to 'light'
  var _useState = useState(function () {
      return getColorScheme(initialColorScheme);
    }),
    _useState2 = _slicedToArray(_useState, 2),
    colorSchemeState = _useState2[0],
    setColorSchemeState = _useState2[1];
  var setColorScheme = useCallback(function setThemeMode(colorScheme) {
    if (false) {
      if (!colorSchemeNamesInput.includes(colorScheme)) {
        throwBladeError({
          message: "Expected color scheme to be one of [".concat(colorSchemeNamesInput.toString(), "] but received ").concat(colorScheme),
          moduleName: 'useColorScheme'
        });
      }
    }
    setColorSchemeState(getColorScheme(colorScheme));
  }, []);
  return {
    colorScheme: colorSchemeState,
    setColorScheme: setColorScheme
  };
};

export { useColorScheme };
