// @razorpay/blade/build/lib/web/production/components/ButtonGroup/StyledButtonGroup.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import styled from 'styled-components';
import '../Box/BaseBox/index.js';
import '../../utils/index.js';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { makeBorderSize } from '../../utils/makeBorderSize/makeBorderSize.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StyledButtonGroup = /*#__PURE__*/styled(BaseBox).withConfig({
  displayName: "StyledButtonGroup",
  componentId: "gr4xbj-0"
})(function (_ref) {
  var theme = _ref.theme,
    _ref$variant = _ref.variant,
    variant = _ref$variant === void 0 ? 'primary' : _ref$variant,
    isFullWidth = _ref.isFullWidth;
  return _objectSpread({
    display: 'flex',
    width: isFullWidth ? '100%' : 'fit-content',
    borderWidth: makeBorderSize(theme.border.width.thin),
    borderRadius: makeBorderSize(theme.border.radius.medium),
    borderStyle: 'solid',
    borderColor: 'transparent',
    overflow: 'hidden',
    'button[role="button"]': {
      flex: isFullWidth ? 1 : 'auto',
      borderRadius: 0
    }
  }, variant === 'secondary' && {
    // Handles both direct buttons and buttons wrapped in components like Dropdown
    '> button[role="button"]:first-child, > *:first-child button[role="button"]': {
      borderRight: 'none',
      borderTopLeftRadius: makeBorderSize(theme.border.radius.medium),
      borderBottomLeftRadius: makeBorderSize(theme.border.radius.medium)
    },
    // Targets middle buttons, whether direct children or wrapped in container elements
    '> button[role="button"]:not(:first-child):not(:last-child), > *:not(:first-child):not(:last-child) button[role="button"]': {
      borderLeft: 'none',
      borderRight: 'none'
    },
    // Handles last button in group, accounting for wrapper components like Dropdown
    '> button[role="button"]:last-child, > *:last-child button[role="button"]': {
      borderLeft: 'none',
      borderTopRightRadius: makeBorderSize(theme.border.radius.medium),
      borderBottomRightRadius: makeBorderSize(theme.border.radius.medium)
    }
  });
});

export { StyledButtonGroup };
