// @razorpay/blade/build/lib/web/development/components/Radio/RadioGroup/RadioGroup.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default from 'react';
import { radioSizes } from '../radioTokens.js';
import { RadioGroupProvider } from './RadioContext.js';
import { useRadioGroup } from './useRadioGroup.js';
import '../../Box/BaseBox/index.js';
import '../../Form/index.js';
import { SelectorGroupField } from '../../Form/Selector/SelectorGroupField.js';
import '../../Box/styledProps/index.js';
import '../../../utils/index.js';
import '../../BladeProvider/index.js';
import '../../../utils/makeAnalyticsAttribute/index.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import useTheme from '../../BladeProvider/useTheme.js';
import { useBreakpoint } from '../../../utils/useBreakpoint/useBreakpoint.js';
import { BaseBox } from '../../Box/BaseBox/BaseBox.web.js';
import { getStyledProps } from '../../Box/styledProps/getStyledProps.js';
import { makeAnalyticsAttribute } from '../../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { FormLabel } from '../../Form/FormLabel.js';
import { FormHint } from '../../Form/FormHint.js';

var _excluded = ["children", "label", "helpText", "isDisabled", "isRequired", "necessityIndicator", "labelPosition", "validationState", "errorText", "name", "defaultValue", "onChange", "value", "size", "orientation", "testID", "flexWrap"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var RadioGroup = function RadioGroup(_ref) {
  var children = _ref.children,
    label = _ref.label,
    helpText = _ref.helpText,
    _ref$isDisabled = _ref.isDisabled,
    isDisabled = _ref$isDisabled === void 0 ? false : _ref$isDisabled,
    _ref$isRequired = _ref.isRequired,
    isRequired = _ref$isRequired === void 0 ? false : _ref$isRequired,
    _ref$necessityIndicat = _ref.necessityIndicator,
    necessityIndicator = _ref$necessityIndicat === void 0 ? 'none' : _ref$necessityIndicat,
    _ref$labelPosition = _ref.labelPosition,
    labelPosition = _ref$labelPosition === void 0 ? 'top' : _ref$labelPosition,
    _ref$validationState = _ref.validationState,
    validationState = _ref$validationState === void 0 ? 'none' : _ref$validationState,
    errorText = _ref.errorText,
    name = _ref.name,
    defaultValue = _ref.defaultValue,
    onChange = _ref.onChange,
    value = _ref.value,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'medium' : _ref$size,
    _ref$orientation = _ref.orientation,
    orientation = _ref$orientation === void 0 ? 'vertical' : _ref$orientation,
    testID = _ref.testID,
    _ref$flexWrap = _ref.flexWrap,
    flexWrap = _ref$flexWrap === void 0 ? 'nowrap' : _ref$flexWrap,
    rest = _objectWithoutProperties(_ref, _excluded);
  var _useRadioGroup = useRadioGroup({
      defaultValue: defaultValue,
      isDisabled: isDisabled,
      isRequired: isRequired,
      labelPosition: labelPosition,
      name: name,
      necessityIndicator: necessityIndicator,
      onChange: onChange,
      validationState: validationState,
      value: value,
      size: size
    }),
    contextValue = _useRadioGroup.contextValue,
    ids = _useRadioGroup.ids;
  var _useTheme = useTheme(),
    theme = _useTheme.theme;
  var _useBreakpoint = useBreakpoint({
      breakpoints: theme.breakpoints
    }),
    matchedDeviceType = _useBreakpoint.matchedDeviceType;
  var showError = validationState === 'error' && errorText;
  var showHelpText = !showError && helpText;
  var accessibilityText = "".concat(showError ? errorText : '', " ").concat(showHelpText ? helpText : '').trim();
  var gap = radioSizes.group.gap[size][matchedDeviceType];
  return /*#__PURE__*/jsx(RadioGroupProvider, {
    value: contextValue,
    children: /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread({}, getStyledProps(rest)), {}, {
      children: /*#__PURE__*/jsxs(SelectorGroupField, _objectSpread(_objectSpread({
        position: labelPosition,
        labelledBy: ids.labelId,
        accessibilityRole: "radiogroup",
        componentName: "radio-group",
        testID: testID
      }, makeAnalyticsAttribute(rest)), {}, {
        children: [label ? /*#__PURE__*/jsx(FormLabel, {
          as: "span",
          necessityIndicator: necessityIndicator,
          position: labelPosition,
          id: ids.labelId,
          accessibilityText: accessibilityText && ",".concat(accessibilityText),
          size: size,
          children: label
        }) : null, /*#__PURE__*/jsxs(BaseBox, {
          children: [/*#__PURE__*/jsx(BaseBox, {
            display: "flex",
            flexDirection: orientation === 'vertical' ? 'column' : 'row',
            gap: gap,
            flexWrap: flexWrap,
            children: React__default.Children.map(children, function (child, index) {
              return /*#__PURE__*/jsx(BaseBox, {
                children: child
              }, index);
            })
          }), /*#__PURE__*/jsx(FormHint, {
            size: size,
            type: validationState === 'error' ? 'error' : 'help',
            errorText: errorText,
            helpText: helpText
          })]
        })]
      }))
    }))
  });
};

export { RadioGroup };
