// @razorpay/blade/build/lib/web/development/components/List/ListItemText.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import { useListContext } from './ListContext.js';
import '../Typography/index.js';
import '../../utils/assignWithoutSideEffects/index.js';
import '../../utils/metaAttribute/index.js';
import { jsx } from 'react/jsx-runtime';
import { Text } from '../Typography/Text/Text.js';
import { assignWithoutSideEffects } from '../../utils/assignWithoutSideEffects/assignWithoutSideEffects.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';

var _excluded = ["children", "testID"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _ListItemText = function _ListItemText(_ref) {
  var children = _ref.children,
    testID = _ref.testID,
    props = _objectWithoutProperties(_ref, _excluded);
  var _useListContext = useListContext(),
    size = _useListContext.size;
  return /*#__PURE__*/jsx(Text, _objectSpread(_objectSpread({
    as: "span"
  }, props), {}, {
    size: size,
    children: children,
    testID: testID
  }));
};
var ListItemText = /*#__PURE__*/assignWithoutSideEffects(_ListItemText, {
  componentId: MetaConstants.ListItemText
});

export { ListItemText };
