// @razorpay/blade/build/lib/web/development/components/List/ListItemLink.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { useListContext } from './ListContext.js';
import '../Link/index.js';
import '../../utils/metaAttribute/index.js';
import '../../utils/assignWithoutSideEffects/index.js';
import { jsx } from 'react/jsx-runtime';
import Link from '../Link/Link/Link.js';
import { assignWithoutSideEffects } from '../../utils/assignWithoutSideEffects/assignWithoutSideEffects.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _ListItemLink = function _ListItemLink(_ref) {
  var accessibilityLabel = _ref.accessibilityLabel,
    children = _ref.children,
    href = _ref.href,
    icon = _ref.icon,
    iconPosition = _ref.iconPosition,
    onClick = _ref.onClick,
    rel = _ref.rel,
    target = _ref.target,
    testID = _ref.testID;
  var _useListContext = useListContext(),
    size = _useListContext.size;
  return /*#__PURE__*/jsx(Link, _objectSpread(_objectSpread({
    size: size,
    variant: "anchor"
  }, icon ? {
    icon: icon,
    children: children
  } : {
    children: children
  }), {}, {
    href: href,
    target: target,
    rel: rel,
    iconPosition: iconPosition,
    onClick: onClick,
    accessibilityLabel: accessibilityLabel,
    testID: testID
  }));
};
var ListItemLink = /*#__PURE__*/assignWithoutSideEffects(_ListItemLink, {
  componentId: MetaConstants.ListItemLink
});

export { ListItemLink };
