// @razorpay/blade/build/lib/web/production/components/Popover/PopoverInteractiveWrapper.web.js
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import styled from 'styled-components';
import React__default from 'react';
import { usePopoverContext } from './PopoverContext.js';
import '../../utils/metaAttribute/index.js';
import '../../utils/makeAccessible/index.js';
import { useMemoizedStyles } from '../Box/BaseBox/useMemoizedStyles.web.js';
import '../../utils/getFocusRingStyles/index.js';
import { jsx } from 'react/jsx-runtime';
import { getFocusRingStyles } from '../../utils/getFocusRingStyles/getFocusRingStyles.web.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { makeAccessible } from '../../utils/makeAccessible/makeAccessible.web.js';

var _excluded = ["accessibilityLabel"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StyledPopoverInteractiveWrapper = /*#__PURE__*/styled.button.withConfig({
  displayName: "PopoverInteractiveWrapperweb__StyledPopoverInteractiveWrapper",
  componentId: "sc-1lanizc-0"
})(function (props) {
  var cssObject = useMemoizedStyles(props);
  return _objectSpread(_objectSpread({
    all: 'unset'
  }, cssObject), {}, {
    appearance: 'none',
    cursor: 'pointer',
    lineHeight: 0,
    '&:focus-visible': _objectSpread({}, getFocusRingStyles({
      theme: props.theme
    }))
  });
});
var PopoverInteractiveWrapper = /*#__PURE__*/React__default.forwardRef(function (_ref, ref) {
  var accessibilityLabel = _ref.accessibilityLabel,
    props = _objectWithoutProperties(_ref, _excluded);
  usePopoverContext();
  return /*#__PURE__*/jsx(StyledPopoverInteractiveWrapper, _objectSpread(_objectSpread(_objectSpread({
    ref: ref,
    display: props.display || 'inline-block'
  }, props), metaAttribute({
    testID: 'popover-interactive-wrapper',
    name: MetaConstants.PopoverInteractiveWrapper
  })), makeAccessible({
    label: accessibilityLabel
  })));
});

export { PopoverInteractiveWrapper };
