// @razorpay/blade/build/lib/web/production/components/Indicator/Indicator.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default from 'react';
import { indicatorDotSizes, textSizeMapping } from './indicatorTokens.js';
import '../BladeProvider/index.js';
import '../Box/BaseBox/index.js';
import '../Icons/_Svg/index.js';
import '../Icons/_Svg/Circle/index.js';
import '../Typography/index.js';
import '../../utils/getStringChildren/index.js';
import '../../utils/index.js';
import '../../utils/metaAttribute/index.js';
import '../Box/styledProps/index.js';
import '../../utils/makeAccessible/index.js';
import '../../utils/assignWithoutSideEffects/index.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import useTheme from '../BladeProvider/useTheme.js';
import { getStringFromReactText } from '../../utils/getStringChildren/getStringChildren.js';
import { isReactNative } from '../../utils/platform/isReactNative.js';
import { makeAccessible } from '../../utils/makeAccessible/makeAccessible.web.js';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { getStyledProps } from '../Box/styledProps/getStyledProps.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import Svg from '../Icons/_Svg/Svg/Svg.web.js';
import Circle from '../Icons/_Svg/Circle/Circle.web.js';
import { Text } from '../Typography/Text/Text.js';
import { assignWithoutSideEffects } from '../../utils/assignWithoutSideEffects/assignWithoutSideEffects.js';

var _excluded = ["accessibilityLabel", "children", "size", "color", "emphasis", "testID"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _Indicator = function _Indicator(_ref, ref) {
  var accessibilityLabel = _ref.accessibilityLabel,
    children = _ref.children,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'medium' : _ref$size,
    _ref$color = _ref.color,
    color = _ref$color === void 0 ? 'neutral' : _ref$color,
    _ref$emphasis = _ref.emphasis,
    emphasis = _ref$emphasis === void 0 ? 'subtle' : _ref$emphasis,
    testID = _ref.testID,
    rest = _objectWithoutProperties(_ref, _excluded);
  var _useTheme = useTheme(),
    theme = _useTheme.theme;
  var childrenString = getStringFromReactText(children);
  var isIntense = emphasis === 'intense';
  var isPrimary = color === 'primary';
  var fillColorOuter = isPrimary ? theme.colors.surface.background.primary.subtle : theme.colors.feedback.background[color].subtle;
  var fillColorInner = isPrimary ? theme.colors.surface.icon.primary.normal : theme.colors.feedback.icon[color].intense;
  var isWeb = !isReactNative();
  var a11yProps = makeAccessible(_objectSpread({
    label: accessibilityLabel !== null && accessibilityLabel !== void 0 ? accessibilityLabel : childrenString
  }, isWeb && {
    role: 'status'
  }));
  var svgSize = isIntense ? indicatorDotSizes[emphasis][size].outer : indicatorDotSizes[emphasis][size].inner;
  return /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread({
    ref: ref,
    display: isWeb ? 'inline-flex' : 'flex'
  }, a11yProps), metaAttribute({
    name: MetaConstants.Indicator,
    testID: testID
  })), getStyledProps(rest)), makeAnalyticsAttribute(rest)), {}, {
    children: /*#__PURE__*/jsxs(BaseBox, {
      display: "flex",
      flexDirection: "row",
      alignItems: "center",
      children: [/*#__PURE__*/jsx(Svg, {
        width: String(svgSize),
        height: String(svgSize),
        viewBox: "0 0 10 10",
        fill: "none",
        children: isIntense ? /*#__PURE__*/jsxs(Fragment, {
          children: [/*#__PURE__*/jsx(Circle, {
            cx: "5",
            cy: "5",
            r: "5",
            fill: fillColorOuter
          }), /*#__PURE__*/jsx(Circle, {
            cx: "5",
            cy: "5",
            r: "2.5",
            fill: fillColorInner
          })]
        }) : /*#__PURE__*/jsx(Circle, {
          cx: "5",
          cy: "5",
          r: "5",
          fill: fillColorInner
        })
      }), /*#__PURE__*/jsx(BaseBox, {
        marginLeft: childrenString ? 'spacing.2' : 'spacing.0',
        children: /*#__PURE__*/jsx(Text, {
          weight: "medium",
          color: "surface.text.gray.subtle",
          textAlign: "left",
          size: textSizeMapping[size],
          children: children
        })
      })]
    })
  }));
};
var Indicator = /*#__PURE__*/assignWithoutSideEffects(/*#__PURE__*/React__default.forwardRef(_Indicator), {
  componentId: 'Indicator'
});

export { Indicator };
