// @razorpay/blade/build/lib/web/development/components/Menu/VisualSubComponents/MenuHeaderFooter.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import 'react';
import { overlayPaddingX, headerMarginBottom, footerPaddingTop } from '../tokens.js';
import { MenuDivider } from './MenuDivider.web.js';
import { BaseFooter } from '../../BaseHeaderFooter/BaseFooter.js';
import { BaseHeader } from '../../BaseHeaderFooter/BaseHeader.js';
import '../../../utils/assignWithoutSideEffects/index.js';
import { MetaConstants } from '../../../utils/metaAttribute/metaConstants.js';
import '../../Box/index.js';
import '../../../utils/makeAnalyticsAttribute/index.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import { Box } from '../../Box/Box.js';
import { makeAnalyticsAttribute } from '../../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { assignWithoutSideEffects } from '../../../utils/assignWithoutSideEffects/assignWithoutSideEffects.js';

var _excluded = ["title", "subtitle", "leading", "titleSuffix", "trailing", "testID"],
  _excluded2 = ["children", "testID"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _MenuHeader = function _MenuHeader(_ref) {
  var title = _ref.title,
    subtitle = _ref.subtitle,
    leading = _ref.leading,
    titleSuffix = _ref.titleSuffix,
    trailing = _ref.trailing,
    testID = _ref.testID,
    rest = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/jsxs(Box, {
    marginBottom: "spacing.3",
    children: [/*#__PURE__*/jsx(BaseHeader, _objectSpread({
      title: title,
      subtitle: subtitle,
      leading: leading,
      trailing: trailing,
      titleSuffix: titleSuffix,
      metaComponentName: MetaConstants.MenuHeader,
      testID: testID
      // back button
      ,
      showBackButton: false
      // close button
      ,
      showCloseButton: false,
      paddingX: overlayPaddingX,
      marginTop: "spacing.0",
      marginBottom: headerMarginBottom,
      showDivider: false
    }, makeAnalyticsAttribute(rest))), /*#__PURE__*/jsx(MenuDivider, {})]
  });
};
var MenuHeader = /*#__PURE__*/assignWithoutSideEffects(_MenuHeader, {
  componentId: 'MenuHeader'
});
var _MenuFooter = function _MenuFooter(_ref2) {
  var children = _ref2.children,
    testID = _ref2.testID,
    rest = _objectWithoutProperties(_ref2, _excluded2);
  return /*#__PURE__*/jsxs(Box, {
    marginTop: "spacing.3",
    children: [/*#__PURE__*/jsx(MenuDivider, {}), /*#__PURE__*/jsx(BaseFooter, _objectSpread(_objectSpread({
      metaComponentName: MetaConstants.MenuFooter,
      showDivider: false,
      padding: [footerPaddingTop, overlayPaddingX, 'spacing.0', overlayPaddingX],
      testID: testID
    }, makeAnalyticsAttribute(rest)), {}, {
      children: children
    }))]
  });
};
var MenuFooter = /*#__PURE__*/assignWithoutSideEffects(_MenuFooter, {
  componentId: 'MenuFooter'
});

export { MenuFooter, MenuHeader };
