// @razorpay/blade/build/lib/web/development/components/Stagger/Stagger.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _typeof from '@babel/runtime/helpers/typeof';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default from 'react';
import { AnimatePresence } from 'framer-motion';
import { StaggerContext } from './StaggerProvider.js';
import '../BaseMotion/index.js';
import { msToSeconds } from '../../utils/msToSeconds.js';
import '../../utils/index.js';
import { jsx } from 'react/jsx-runtime';
import useTheme from '../BladeProvider/useTheme.js';
import { BaseMotionBox } from '../BaseMotion/BaseMotion.js';

var _excluded = ["children", "isVisible", "type", "shouldUnmountWhenHidden", "delay", "motionTriggers"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Stagger = function Stagger(_ref) {
  var children = _ref.children,
    _ref$isVisible = _ref.isVisible,
    isVisible = _ref$isVisible === void 0 ? true : _ref$isVisible,
    _ref$type = _ref.type,
    type = _ref$type === void 0 ? 'inout' : _ref$type,
    _ref$shouldUnmountWhe = _ref.shouldUnmountWhenHidden,
    shouldUnmountWhenHidden = _ref$shouldUnmountWhe === void 0 ? false : _ref$shouldUnmountWhe,
    delay = _ref.delay,
    motionTriggers = _ref.motionTriggers,
    boxProps = _objectWithoutProperties(_ref, _excluded);
  var _useTheme = useTheme(),
    theme = _useTheme.theme;
  // Only need AnimatePresence when we have to unmount the component
  var AnimateWrapper = shouldUnmountWhenHidden ? AnimatePresence : React__default.Fragment;
  // keep it always mounted when shouldUnmountWhenHidden is false
  var isMounted = shouldUnmountWhenHidden ? isVisible : true;
  var enterDelay = _typeof(delay) === 'object' ? delay.enter : delay;
  var exitDelay = _typeof(delay) === 'object' ? delay.exit : delay;
  var staggerVariants = {
    initial: {},
    animate: {
      transition: _objectSpread(_objectSpread({}, enterDelay ? {
        delayChildren: msToSeconds(theme.motion.delay[enterDelay])
      } : {}), {}, {
        staggerChildren: msToSeconds(theme.motion.duration['2xquick'])
      })
    },
    exit: {
      transition: _objectSpread(_objectSpread({}, exitDelay ? {
        delayChildren: msToSeconds(theme.motion.delay[exitDelay])
      } : {}), {}, {
        staggerChildren: msToSeconds(theme.motion.duration['2xquick'])
      })
    }
  };
  return /*#__PURE__*/jsx(AnimateWrapper, {
    children: isMounted ? /*#__PURE__*/jsx(BaseMotionBox, _objectSpread(_objectSpread(_objectSpread({
      type: type,
      motionVariants: staggerVariants,
      motionTriggers: motionTriggers
    }, shouldUnmountWhenHidden ? {} : {
      animateVisibility: isVisible ? 'animate' : 'exit'
    }), boxProps), {}, {
      children: /*#__PURE__*/jsx(StaggerContext.Provider, {
        value: {
          isInsideStaggerContainer: true,
          staggerType: type
        },
        children: children
      })
    })) : null
  });
};

export { Stagger };
