// @razorpay/blade/build/lib/web/development/components/DatePicker/FilterChipDatePicker/DatePickerFilterChip.web.js
import _objectDestructuringEmpty from '@babel/runtime/helpers/objectDestructuringEmpty';
import _extends from '@babel/runtime/helpers/extends';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import { forwardRef } from 'react';
import { useDatesContext } from '@mantine/dates';
import { getFormattedDate } from '../utils.js';
import { BaseFilterChip } from '../../FilterChip/BaseFilterChip.web.js';
import { jsx } from 'react/jsx-runtime';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var formatDateRange = function formatDateRange(date, format, locale, selectionType) {
  var formatOptions = {
    date: date,
    format: format,
    labelSeparator: '-',
    locale: locale,
    type: 'default'
  };
  if (selectionType === 'single' && date instanceof Date) {
    return getFormattedDate(formatOptions);
  }
  if (Array.isArray(date)) {
    var _date = _slicedToArray(date, 2),
      startDate = _date[0],
      endDate = _date[1];
    if (startDate) {
      return "".concat(getFormattedDate(_objectSpread(_objectSpread({}, formatOptions), {}, {
        date: startDate
      })), "  ").concat(endDate ? ' - ' : '', " ").concat(getFormattedDate(_objectSpread(_objectSpread({}, formatOptions), {}, {
        date: endDate
      })));
    }
  }
  return '';
};
var _DatePickerFilterChip = function _DatePickerFilterChip(_ref, ref) {
  var props = _extends({}, (_objectDestructuringEmpty(_ref), _ref));
  var referenceProps = props.referenceProps,
    label = props.label,
    selectionType = props.selectionType,
    onClearButtonChange = props.onClearButtonChange,
    accessibilityLabel = props.accessibilityLabel,
    date = props.date,
    format = props.format,
    isDisabled = props.isDisabled;
  var _useDatesContext = useDatesContext(),
    locale = _useDatesContext.locale;
  var dateValue = formatDateRange(date, format, locale, selectionType);
  return /*#__PURE__*/jsx(BaseFilterChip, _objectSpread({
    ref: ref,
    label: label,
    value: dateValue,
    onClearButtonClick: onClearButtonChange,
    accessibilityProps: {
      label: accessibilityLabel !== null && accessibilityLabel !== void 0 ? accessibilityLabel : label,
      hasPopup: referenceProps['aria-haspopup'],
      expanded: referenceProps['aria-expanded'],
      controls: referenceProps['aria-controls'],
      role: 'combobox'
    },
    isDisabled: isDisabled
  }, referenceProps));
};
var DatePickerFilterChip = /*#__PURE__*/forwardRef(_DatePickerFilterChip);

export { DatePickerFilterChip };
