// @razorpay/blade/build/lib/web/development/components/Dropdown/InputDropdownButton.web.js
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import 'react';
import styled from 'styled-components';
import { useDropdown } from './useDropdown.js';
import { dropdownComponentIds } from './dropdownComponentIds.js';
import '../../utils/assignWithoutSideEffects/index.js';
import '../../utils/index.js';
import '../Icons/index.js';
import '../Box/index.js';
import '../../utils/getFocusRingStyles/index.js';
import '../Typography/index.js';
import '../../utils/makeAccessible/index.js';
import { getActionListContainerRole } from '../ActionList/getA11yRoles.js';
import { useId } from '../../utils/useId.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import '../../utils/metaAttribute/index.js';
import '../../utils/useControlledDropdownInput/index.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import { makeSpace } from '../../utils/makeSpace/makeSpace.js';
import { getFocusRingStyles } from '../../utils/getFocusRingStyles/getFocusRingStyles.web.js';
import { useControlledDropdownInput } from '../../utils/useControlledDropdownInput/useControlledDropdownInput.js';
import { makeAccessible } from '../../utils/makeAccessible/makeAccessible.web.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { Box } from '../Box/Box.js';
import { Text } from '../Typography/Text/Text.js';
import ChevronUpDownIcon from '../Icons/ChevronUpDownIcon/ChevronUpDownIcon.js';
import { assignWithoutSideEffects } from '../../utils/assignWithoutSideEffects/assignWithoutSideEffects.js';

var _excluded = ["onClick", "onBlur", "onKeyDown", "accessibilityLabel", "_isInsideSearchInput", "isDisabled", "onChange", "name", "testID", "value", "defaultValue", "icon"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StyledSearchTrailingDropdown = /*#__PURE__*/styled.button.withConfig({
  displayName: "InputDropdownButtonweb__StyledSearchTrailingDropdown",
  componentId: "gq7bp3-0"
})(function (_ref) {
  var theme = _ref.theme,
    isDisabled = _ref.isDisabled;
  var spacing = theme.spacing;
  return {
    backgroundColor: theme.colors.transparent,
    gap: makeSpace(spacing[2]),
    display: 'flex',
    height: '100%',
    alignItems: 'center',
    border: 'none',
    cursor: isDisabled ? 'not-allowed' : 'pointer',
    '&[disabled]': {
      cursor: 'not-allowed',
      pointerEvents: 'none'
    },
    '&:focus': _objectSpread(_objectSpread({}, getFocusRingStyles({
      theme: theme
    })), {}, {
      backgroundColor: theme.colors.interactive.background.gray.faded
    }),
    '&:hover': {
      backgroundColor: theme.colors.interactive.background.gray.faded
    },
    '&:focus-visible': {
      outlineOffset: makeSpace(theme.spacing[0])
    },
    borderRadius: theme.border.radius.small
  };
});
var _InputDropdownButton = function _InputDropdownButton(_ref2) {
  var _onClick = _ref2.onClick,
    _onBlur = _ref2.onBlur,
    _onKeyDown = _ref2.onKeyDown,
    accessibilityLabel = _ref2.accessibilityLabel,
    _ref2$_isInsideSearch = _ref2._isInsideSearchInput,
    _isInsideSearchInput = _ref2$_isInsideSearch === void 0 ? false : _ref2$_isInsideSearch,
    isDisabled = _ref2.isDisabled,
    _onChange = _ref2.onChange,
    name = _ref2.name,
    testID = _ref2.testID,
    value = _ref2.value,
    defaultValue = _ref2.defaultValue,
    Icon = _ref2.icon,
    rest = _objectWithoutProperties(_ref2, _excluded);
  var idBase = useId('input-drop-down-button');
  var _useDropdown = useDropdown(),
    onTriggerClick = _useDropdown.onTriggerClick,
    onTriggerKeydown = _useDropdown.onTriggerKeydown,
    dropdownBaseId = _useDropdown.dropdownBaseId,
    isOpen = _useDropdown.isOpen,
    activeIndex = _useDropdown.activeIndex,
    hasFooterAction = _useDropdown.hasFooterAction,
    triggererRef = _useDropdown.triggererRef,
    displayValue = _useDropdown.displayValue;
  useControlledDropdownInput({
    onChange: function onChange(_ref3) {
      var name = _ref3.name,
        values = _ref3.values;
      _onChange === null || _onChange === void 0 || _onChange({
        name: name || idBase,
        value: values[0]
      });
    },
    name: name,
    value: value,
    defaultValue: defaultValue,
    triggererRef: triggererRef,
    isSelectInput: false
  });
  if (!displayValue) {
    return null;
  }
  return /*#__PURE__*/jsx(StyledSearchTrailingDropdown, _objectSpread(_objectSpread(_objectSpread(_objectSpread({
    type: "button",
    onClick: function onClick(e) {
      if (isDisabled) return;
      onTriggerClick();
      // Setting it for web fails it on native typecheck and vice versa
      _onClick === null || _onClick === void 0 || _onClick(e);
      // Since this dropdown is inside another dropdown we should stop event stopPropagation.
      e === null || e === void 0 || e.stopPropagation();
    },
    onBlur: function onBlur(e) {
      if (isDisabled) return;
      // With button trigger, there is no "value" as such. It's just clickable items
      // Setting it for web fails it on native typecheck and vice versa
      _onBlur === null || _onBlur === void 0 || _onBlur(e);
      e === null || e === void 0 || e.stopPropagation();
    },
    onKeyDown: function onKeyDown(e) {
      if (isDisabled) return;
      onTriggerKeydown === null || onTriggerKeydown === void 0 || onTriggerKeydown({
        event: e
      });
      // Setting it for web fails it on native typecheck and vice versa
      _onKeyDown === null || _onKeyDown === void 0 || _onKeyDown(e);
      e === null || e === void 0 || e.stopPropagation();
    }
  }, makeAccessible({
    label: accessibilityLabel !== null && accessibilityLabel !== void 0 ? accessibilityLabel : "change ".concat(displayValue, " filter"),
    hasPopup: getActionListContainerRole(hasFooterAction, 'InputDropdownButton'),
    expanded: isOpen,
    controls: "".concat(dropdownBaseId, "-actionlist"),
    activeDescendant: activeIndex >= 0 ? "".concat(dropdownBaseId, "-").concat(activeIndex) : undefined
  })), {}, {
    ref: triggererRef,
    isDisabled: isDisabled
  }, makeAnalyticsAttribute(rest)), metaAttribute({
    name: 'InputDropdownButton',
    testID: testID
  })), {}, {
    children: /*#__PURE__*/jsxs(Box, {
      padding: "spacing.2",
      display: "flex",
      gap: "spacing.2",
      alignItems: "center",
      children: [_isInsideSearchInput && /*#__PURE__*/jsxs(Text, {
        variant: "body",
        size: "medium",
        weight: "regular",
        color: isDisabled ? 'surface.text.gray.disabled' : 'surface.text.gray.muted',
        children: [' ', "in"]
      }), Icon && /*#__PURE__*/jsx(Icon, {
        size: "medium",
        color: isDisabled ? 'surface.icon.gray.disabled' : 'surface.icon.gray.muted'
      }), /*#__PURE__*/jsx(Text, {
        variant: "body",
        size: "medium",
        weight: "regular",
        color: isDisabled ? 'surface.text.gray.disabled' : 'surface.text.gray.subtle',
        children: displayValue
      }), /*#__PURE__*/jsx(ChevronUpDownIcon, {
        color: isDisabled ? 'surface.icon.gray.disabled' : 'surface.icon.gray.muted'
      })]
    })
  }));
};
var InputDropdownButton = /*#__PURE__*/assignWithoutSideEffects(_InputDropdownButton, {
  componentId: dropdownComponentIds.triggers.InputDropdownButton,
  displayName: 'InputDropDown'
});

export { InputDropdownButton };
