// @razorpay/blade/build/lib/web/development/components/Input/BaseInput/BaseInputWrapper.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default from 'react';
import { AnimatedBaseInputWrapper } from './AnimatedBaseInputWrapper.web.js';
import { jsx } from 'react/jsx-runtime';

var _excluded = ["children", "validationState", "currentInteraction", "isLabelLeftPositioned", "isTextArea", "showAllTags", "setShowAllTagsWithAnimation", "maxTagRows", "numberOfLines", "isTableInputCell"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _BaseInputWrapper = function _BaseInputWrapper(_ref, ref) {
  var children = _ref.children,
    validationState = _ref.validationState,
    currentInteraction = _ref.currentInteraction,
    isLabelLeftPositioned = _ref.isLabelLeftPositioned,
    isTextArea = _ref.isTextArea,
    showAllTags = _ref.showAllTags,
    setShowAllTagsWithAnimation = _ref.setShowAllTagsWithAnimation,
    maxTagRows = _ref.maxTagRows,
    numberOfLines = _ref.numberOfLines,
    isTableInputCell = _ref.isTableInputCell,
    props = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/jsx(AnimatedBaseInputWrapper
  // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unnecessary-type-assertion
  , _objectSpread(_objectSpread({
    ref: ref,
    isTextArea: isTextArea,
    validationState: validationState,
    currentInteraction: currentInteraction,
    showAllTags: showAllTags,
    maxTagRows: maxTagRows,
    setShowAllTagsWithAnimation: setShowAllTagsWithAnimation,
    numberOfLines: numberOfLines,
    isTableInputCell: isTableInputCell
  }, props), {}, {
    children: children
  }));
};
var BaseInputWrapper = /*#__PURE__*/React__default.forwardRef(_BaseInputWrapper);

export { BaseInputWrapper };
