// @razorpay/blade/build/lib/web/production/tokens/global/colors.js
import { opacity } from './opacity.js';

var colors = {
  chromatic: {
    azure: {
      50: "hsla(222, 100%, 98%, ".concat(opacity[1200], ")"),
      100: "hsla(221, 90%, 92%, ".concat(opacity[1200], ")"),
      200: "hsla(219, 95%, 85%, ".concat(opacity[1200], ")"),
      300: "hsla(220, 100%, 73%, ".concat(opacity[1200], ")"),
      400: "hsla(223, 100%, 65%, ".concat(opacity[1200], ")"),
      500: "hsla(227, 100%, 59%, ".concat(opacity[1200], ")"),
      600: "hsla(227, 71%, 51%, ".concat(opacity[1200], ")"),
      700: "hsla(227, 69%, 42%, ".concat(opacity[1200], ")"),
      800: "hsla(227, 69%, 34%, ".concat(opacity[1200], ")"),
      900: "hsla(227, 69%, 25%, ".concat(opacity[1200], ")"),
      1000: "hsla(227, 69%, 17%, ".concat(opacity[1200], ")"),
      a50: "hsla(227, 100%, 59%, ".concat(opacity[100], ")"),
      a100: "hsla(227, 100%, 59%, ".concat(opacity[300], ")"),
      a150: "hsla(227, 100%, 59%, ".concat(opacity[400], ")"),
      a200: "hsla(227, 100%, 59%, ".concat(opacity[500], ")"),
      a400: "hsla(227, 100%, 59%, ".concat(opacity[800], ")")
    },
    emerald: {
      50: "hsla(152, 60%, 95%, ".concat(opacity[1200], ")"),
      100: "hsla(151, 57%, 91%, ".concat(opacity[1200], ")"),
      200: "hsla(150, 59%, 82%, ".concat(opacity[1200], ")"),
      300: "hsla(150, 59%, 73%, ".concat(opacity[1200], ")"),
      400: "hsla(150, 59%, 55%, ".concat(opacity[1200], ")"),
      500: "hsla(150, 100%, 37%, ".concat(opacity[1200], ")"),
      600: "hsla(150, 100%, 32%, ".concat(opacity[1200], ")"),
      700: "hsla(150, 100%, 27%, ".concat(opacity[1200], ")"),
      800: "hsla(150, 100%, 21%, ".concat(opacity[1200], ")"),
      900: "hsla(150, 100%, 16%, ".concat(opacity[1200], ")"),
      1000: "hsla(150, 100%, 11%, ".concat(opacity[1200], ")"),
      a50: "hsla(150, 100%, 32%, ".concat(opacity[100], ")"),
      a100: "hsla(150, 100%, 32%, ".concat(opacity[300], ")"),
      a150: "hsla(150, 100%, 32%, ".concat(opacity[400], ")"),
      a200: "hsla(150, 100%, 32%, ".concat(opacity[500], ")"),
      a400: "hsla(150, 100%, 32%, ".concat(opacity[800], ")")
    },
    crimson: {
      50: "hsla(0, 100%, 98%, ".concat(opacity[1200], ")"),
      100: "hsla(4, 93%, 94%, ".concat(opacity[1200], ")"),
      200: "hsla(3, 97%, 88%, ".concat(opacity[1200], ")"),
      300: "hsla(4, 96%, 79%, ".concat(opacity[1200], ")"),
      400: "hsla(4, 93%, 68%, ".concat(opacity[1200], ")"),
      500: "hsla(4, 86%, 58%, ".concat(opacity[1200], ")"),
      600: "hsla(4, 74%, 49%, ".concat(opacity[1200], ")"),
      700: "hsla(4, 77%, 40%, ".concat(opacity[1200], ")"),
      800: "hsla(0, 83%, 33%, ".concat(opacity[1200], ")"),
      900: "hsla(0, 84%, 29%, ".concat(opacity[1200], ")"),
      1000: "hsla(0, 84%, 25%, ".concat(opacity[1200], ")"),
      a50: "hsla(4, 74%, 49%, ".concat(opacity[100], ")"),
      a100: "hsla(4, 74%, 49%, ".concat(opacity[300], ")"),
      a150: "hsla(4, 74%, 49%, ".concat(opacity[400], ")"),
      a200: "hsla(4, 74%, 49%, ".concat(opacity[500], ")"),
      a400: "hsla(4, 74%, 49%, ".concat(opacity[800], ")")
    },
    cider: {
      50: "hsla(24, 100%, 96%, ".concat(opacity[1200], ")"),
      100: "hsla(25, 100%, 90%, ".concat(opacity[1200], ")"),
      200: "hsla(25, 100%, 80%, ".concat(opacity[1200], ")"),
      300: "hsla(25, 100%, 72%, ".concat(opacity[1200], ")"),
      400: "hsla(25, 100%, 63%, ".concat(opacity[1200], ")"),
      500: "hsla(25, 100%, 55%, ".concat(opacity[1200], ")"),
      600: "hsla(25, 90%, 48%, ".concat(opacity[1200], ")"),
      700: "hsla(25, 85%, 42%, ".concat(opacity[1200], ")"),
      800: "hsla(25, 82%, 35%, ".concat(opacity[1200], ")"),
      900: "hsla(25, 80%, 28%, ".concat(opacity[1200], ")"),
      1000: "hsla(25, 82%, 20%, ".concat(opacity[1200], ")"),
      a50: "hsla(25, 90%, 48%, ".concat(opacity[100], ")"),
      a100: "hsla(25, 90%, 48%, ".concat(opacity[300], ")"),
      a150: "hsla(25, 90%, 48%, ".concat(opacity[400], ")"),
      a200: "hsla(25, 90%, 48%, ".concat(opacity[500], ")"),
      a400: "hsla(25, 90%, 48%, ".concat(opacity[800], ")")
    },
    sapphire: {
      50: "hsla(201, 92%, 95%, ".concat(opacity[1200], ")"),
      100: "hsla(200, 88%, 90%, ".concat(opacity[1200], ")"),
      200: "hsla(200, 89%, 82%, ".concat(opacity[1200], ")"),
      300: "hsla(200, 90%, 72%, ".concat(opacity[1200], ")"),
      400: "hsla(200, 90%, 65%, ".concat(opacity[1200], ")"),
      500: "hsla(198, 90%, 52%, ".concat(opacity[1200], ")"),
      600: "hsla(200, 84%, 44%, ".concat(opacity[1200], ")"),
      700: "hsla(200, 84%, 37%, ".concat(opacity[1200], ")"),
      800: "hsla(200, 84%, 29%, ".concat(opacity[1200], ")"),
      900: "hsla(200, 84%, 22%, ".concat(opacity[1200], ")"),
      1000: "hsla(199, 84%, 15%, ".concat(opacity[1200], ")"),
      a50: "hsla(200, 84%, 44%, ".concat(opacity[100], ")"),
      a100: "hsla(200, 84%, 44%, ".concat(opacity[300], ")"),
      a150: "hsla(200, 84%, 44%, ".concat(opacity[400], ")"),
      a200: "hsla(200, 84%, 44%, ".concat(opacity[500], ")"),
      a400: "hsla(200, 84%, 44%, ".concat(opacity[800], ")")
    },
    sea: {
      50: "hsla(180, 39%, 95%, ".concat(opacity[1200], ")"),
      100: "hsla(180, 42%, 92%, ".concat(opacity[1200], ")"),
      200: "hsla(180, 33%, 82%, ".concat(opacity[1200], ")"),
      300: "hsla(180, 35%, 70%, ".concat(opacity[1200], ")"),
      400: "hsla(180, 30%, 52%, ".concat(opacity[1200], ")"),
      500: "hsla(180, 45%, 40%, ".concat(opacity[1200], ")"),
      600: "hsla(180, 60%, 30%, ".concat(opacity[1200], ")"),
      700: "hsla(180, 55%, 25%, ".concat(opacity[1200], ")"),
      800: "hsla(180, 61%, 20%, ".concat(opacity[1200], ")"),
      900: "hsla(180, 91%, 13%, ".concat(opacity[1200], ")"),
      1000: "hsla(180, 91%, 8%, ".concat(opacity[1200], ")"),
      a50: "hsla(180, 60%, 30%, ".concat(opacity[100], ")"),
      a100: "hsla(180, 60%, 30%, ".concat(opacity[300], ")"),
      a150: "hsla(180, 60%, 30%, ".concat(opacity[400], ")"),
      a200: "hsla(180, 60%, 30%, ".concat(opacity[500], ")"),
      a400: "hsla(180, 60%, 30%, ".concat(opacity[800], ")")
    },
    cloud: {
      50: "hsla(198, 39%, 95%, ".concat(opacity[1200], ")"),
      100: "hsla(201, 39%, 93%, ".concat(opacity[1200], ")"),
      200: "hsla(200, 35%, 85%, ".concat(opacity[1200], ")"),
      300: "hsla(200, 35%, 70%, ".concat(opacity[1200], ")"),
      400: "hsla(200, 30%, 52%, ".concat(opacity[1200], ")"),
      500: "hsla(200, 45%, 40%, ".concat(opacity[1200], ")"),
      600: "hsla(200, 60%, 30%, ".concat(opacity[1200], ")"),
      700: "hsla(201, 55%, 25%, ".concat(opacity[1200], ")"),
      800: "hsla(200, 61%, 20%, ".concat(opacity[1200], ")"),
      900: "hsla(200, 91%, 13%, ".concat(opacity[1200], ")"),
      1000: "hsla(200, 91%, 8%, ".concat(opacity[1200], ")"),
      a50: "hsla(200, 60%, 30%, ".concat(opacity[100], ")"),
      a100: "hsla(200, 60%, 30%, ".concat(opacity[300], ")"),
      a150: "hsla(200, 60%, 30%, ".concat(opacity[400], ")"),
      a200: "hsla(200, 60%, 30%, ".concat(opacity[500], ")"),
      a400: "hsla(200, 60%, 30%, ".concat(opacity[800], ")")
    },
    forest: {
      50: "hsla(152, 60%, 95%, ".concat(opacity[1200], ")"),
      100: "hsla(151, 57%, 91%, ".concat(opacity[1200], ")"),
      200: "hsla(150, 59%, 82%, ".concat(opacity[1200], ")"),
      300: "hsla(150, 59%, 73%, ".concat(opacity[1200], ")"),
      400: "hsla(150, 59%, 55%, ".concat(opacity[1200], ")"),
      500: "hsla(155, 100%, 37%, ".concat(opacity[1200], ")"),
      600: "hsla(155, 100%, 31%, ".concat(opacity[1200], ")"),
      700: "hsla(155, 100%, 27%, ".concat(opacity[1200], ")"),
      800: "hsla(155, 100%, 21%, ".concat(opacity[1200], ")"),
      900: "hsla(155, 100%, 16%, ".concat(opacity[1200], ")"),
      1000: "hsla(155, 100%, 11%, ".concat(opacity[1200], ")"),
      a50: "hsla(155, 100%, 31%, ".concat(opacity[100], ")"),
      a100: "hsla(155, 100%, 31%, ".concat(opacity[300], ")"),
      a150: "hsla(155, 100%, 31%, ".concat(opacity[400], ")"),
      a200: "hsla(155, 100%, 31%, ".concat(opacity[500], ")"),
      a400: "hsla(155, 100%, 31%, ".concat(opacity[800], ")")
    },
    orchid: {
      50: "hsla(267, 100%, 95%, ".concat(opacity[1200], ")"),
      100: "hsla(264, 100%, 89%, ".concat(opacity[1200], ")"),
      200: "hsla(263, 100%, 84%, ".concat(opacity[1200], ")"),
      300: "hsla(261, 100%, 79%, ".concat(opacity[1200], ")"),
      400: "hsla(259, 100%, 74%, ".concat(opacity[1200], ")"),
      500: "hsla(258, 93%, 68%, ".concat(opacity[1200], ")"),
      600: "hsla(257, 69%, 58%, ".concat(opacity[1200], ")"),
      700: "hsla(258, 54%, 48%, ".concat(opacity[1200], ")"),
      800: "hsla(257, 58%, 37%, ".concat(opacity[1200], ")"),
      900: "hsla(257, 62%, 26%, ".concat(opacity[1200], ")"),
      1000: "hsla(257, 65%, 14%, ".concat(opacity[1200], ")"),
      a50: "hsla(258, 93%, 68%, ".concat(opacity[100], ")"),
      a100: "hsla(258, 93%, 68%, ".concat(opacity[300], ")"),
      a150: "hsla(258, 93%, 68%, ".concat(opacity[400], ")"),
      a200: "hsla(258, 93%, 68%, ".concat(opacity[500], ")"),
      a400: "hsla(258, 93%, 68%, ".concat(opacity[800], ")")
    },
    magenta: {
      50: "hsla(309, 100%, 94%, ".concat(opacity[1200], ")"),
      100: "hsla(312, 100%, 87%, ".concat(opacity[1200], ")"),
      200: "hsla(316, 100%, 81%, ".concat(opacity[1200], ")"),
      300: "hsla(316, 90%, 73%, ".concat(opacity[1200], ")"),
      400: "hsla(316, 74%, 64%, ".concat(opacity[1200], ")"),
      500: "hsla(317, 60%, 55%, ".concat(opacity[1200], ")"),
      600: "hsla(317, 56%, 46%, ".concat(opacity[1200], ")"),
      700: "hsla(316, 63%, 36%, ".concat(opacity[1200], ")"),
      800: "hsla(316, 70%, 27%, ".concat(opacity[1200], ")"),
      900: "hsla(316, 76%, 18%, ".concat(opacity[1200], ")"),
      1000: "hsla(318, 80%, 10%, ".concat(opacity[1200], ")"),
      a50: "hsla(317, 60%, 55%, ".concat(opacity[100], ")"),
      a100: "hsla(317, 60%, 55%, ".concat(opacity[300], ")"),
      a150: "hsla(317, 60%, 55%, ".concat(opacity[400], ")"),
      a200: "hsla(317, 60%, 55%, ".concat(opacity[500], ")"),
      a400: "hsla(317, 60%, 55%, ".concat(opacity[800], ")")
    },
    topaz: {
      50: "hsla(46, 86%, 83%, ".concat(opacity[1200], ")"),
      100: "hsla(46, 80%, 68%, ".concat(opacity[1200], ")"),
      200: "hsla(46, 75%, 55%, ".concat(opacity[1200], ")"),
      300: "hsla(46, 93%, 43%, ".concat(opacity[1200], ")"),
      400: "hsla(44, 100%, 37%, ".concat(opacity[1200], ")"),
      500: "hsla(41, 100%, 33%, ".concat(opacity[1200], ")"),
      600: "hsla(39, 100%, 28%, ".concat(opacity[1200], ")"),
      700: "hsla(38, 100%, 24%, ".concat(opacity[1200], ")"),
      800: "hsla(37, 100%, 18%, ".concat(opacity[1200], ")"),
      900: "hsla(35, 100%, 13%, ".concat(opacity[1200], ")"),
      1000: "hsla(33, 100%, 7%, ".concat(opacity[1200], ")"),
      a50: "hsla(41, 100%, 33%, ".concat(opacity[100], ")"),
      a100: "hsla(41, 100%, 33%, ".concat(opacity[300], ")"),
      a150: "hsla(41, 100%, 33%, ".concat(opacity[400], ")"),
      a200: "hsla(41, 100%, 33%, ".concat(opacity[500], ")"),
      a400: "hsla(41, 100%, 33%, ".concat(opacity[800], ")")
    }
  },
  neutral: {
    blueGrayLight: {
      0: "hsla(0, 0%, 100%, ".concat(opacity[1200], ")"),
      50: "hsla(210, 40%, 98%, ".concat(opacity[1200], ")"),
      100: "hsla(213, 47%, 96%, ".concat(opacity[1200], ")"),
      200: "hsla(214, 40%, 92%, ".concat(opacity[1200], ")"),
      300: "hsla(214, 28%, 84%, ".concat(opacity[1200], ")"),
      400: "hsla(211, 27%, 76%, ".concat(opacity[1200], ")"),
      500: "hsla(211, 24%, 65%, ".concat(opacity[1200], ")"),
      600: "hsla(211, 22%, 56%, ".concat(opacity[1200], ")"),
      700: "hsla(211, 20%, 52%, ".concat(opacity[1200], ")"),
      800: "hsla(211, 23%, 45%, ".concat(opacity[1200], ")"),
      900: "hsla(211, 26%, 34%, ".concat(opacity[1200], ")"),
      1000: "hsla(211, 29%, 26%, ".concat(opacity[1200], ")"),
      1100: "hsla(211, 33%, 21%, ".concat(opacity[1200], ")"),
      1200: "hsla(212, 39%, 16%, ".concat(opacity[1200], ")"),
      1300: "hsla(211, 53%, 10%, ".concat(opacity[1200], ")"),
      a25: "hsla(211, 20%, 52%, ".concat(opacity[50], ")"),
      a50: "hsla(211, 20%, 52%, ".concat(opacity[100], ")"),
      a75: "hsla(211, 20%, 52%, ".concat(opacity[200], ")"),
      a100: "hsla(211, 20%, 52%, ".concat(opacity[300], ")"),
      a200: "hsla(211, 20%, 52%, ".concat(opacity[500], ")"),
      a400: "hsla(211, 20%, 52%, ".concat(opacity[800], ")")
    },
    blueGrayDark: {
      0: "hsla(240, 20%, 99%, ".concat(opacity[1200], ")"),
      50: "hsla(220, 22%, 92%, ".concat(opacity[1200], ")"),
      100: "hsla(214, 20%, 84%, ".concat(opacity[1200], ")"),
      200: "hsla(214, 17%, 76%, ".concat(opacity[1200], ")"),
      300: "hsla(216, 18%, 62%, ".concat(opacity[1200], ")"),
      400: "hsla(215, 20%, 43%, ".concat(opacity[1200], ")"),
      500: "hsla(216, 22%, 36%, ".concat(opacity[1200], ")"),
      600: "hsla(216, 24%, 31%, ".concat(opacity[1200], ")"),
      700: "hsla(216, 27%, 24%, ".concat(opacity[1200], ")"),
      800: "hsla(217, 27%, 21%, ".concat(opacity[1200], ")"),
      900: "hsla(218, 26%, 18%, ".concat(opacity[1200], ")"),
      1000: "hsla(217, 27%, 15%, ".concat(opacity[1200], ")"),
      1100: "hsla(216, 30%, 13%, ".concat(opacity[1200], ")"),
      1200: "hsla(215, 37%, 10%, ".concat(opacity[1200], ")"),
      1300: "hsla(214, 44%, 7%, ".concat(opacity[1200], ")"),
      a25: "hsla(214, 20%, 84%, ".concat(opacity[50], ")"),
      a50: "hsla(214, 20%, 84%, ".concat(opacity[100], ")"),
      a75: "hsla(214, 20%, 84%, ".concat(opacity[200], ")"),
      a100: "hsla(214, 20%, 84%, ".concat(opacity[300], ")"),
      a200: "hsla(214, 20%, 84%, ".concat(opacity[500], ")"),
      a400: "hsla(214, 20%, 84%, ".concat(opacity[800], ")")
    },
    ashGrayLight: {
      0: "hsla(0, 0%, 100%, ".concat(opacity[1200], ")"),
      50: "hsla(240, 9%, 98%, ".concat(opacity[1200], ")"),
      100: "hsla(210, 9%, 96%, ".concat(opacity[1200], ")"),
      200: "hsla(210, 6%, 94%, ".concat(opacity[1200], ")"),
      300: "hsla(210, 4%, 89%, ".concat(opacity[1200], ")"),
      400: "hsla(214, 7%, 81%, ".concat(opacity[1200], ")"),
      500: "hsla(216, 6%, 69%, ".concat(opacity[1200], ")"),
      600: "hsla(215, 6%, 60%, ".concat(opacity[1200], ")"),
      700: "hsla(214, 6%, 55%, ".concat(opacity[1200], ")"),
      800: "hsla(216, 7%, 45%, ".concat(opacity[1200], ")"),
      900: "hsla(218, 9%, 36%, ".concat(opacity[1200], ")"),
      1000: "hsla(219, 12%, 28%, ".concat(opacity[1200], ")"),
      1100: "hsla(214, 15%, 18%, ".concat(opacity[1200], ")"),
      1200: "hsla(216, 15%, 13%, ".concat(opacity[1200], ")"),
      1300: "hsla(214, 24%, 6%, ".concat(opacity[1200], ")"),
      a25: "hsla(214, 6%, 55%, ".concat(opacity[50], ")"),
      a50: "hsla(214, 6%, 55%, ".concat(opacity[100], ")"),
      a75: "hsla(214, 6%, 55%, ".concat(opacity[200], ")"),
      a100: "hsla(214, 6%, 55%, ".concat(opacity[300], ")"),
      a200: "hsla(214, 6%, 55%, ".concat(opacity[500], ")"),
      a400: "hsla(214, 6%, 55%, ".concat(opacity[500], ")")
    },
    ashGrayDark: {
      0: "hsla(0, 0%, 99%, ".concat(opacity[1200], ")"),
      50: "hsla(240, 2%, 92%, ".concat(opacity[1200], ")"),
      100: "hsla(240, 1%, 84%, ".concat(opacity[1200], ")"),
      200: "hsla(228, 4%, 76%, ".concat(opacity[1200], ")"),
      300: "hsla(227, 4%, 60%, ".concat(opacity[1200], ")"),
      400: "hsla(229, 4%, 50%, ".concat(opacity[1200], ")"),
      500: "hsla(233, 4%, 40%, ".concat(opacity[1200], ")"),
      600: "hsla(233, 5%, 32%, ".concat(opacity[1200], ")"),
      700: "hsla(230, 6%, 22%, ".concat(opacity[1200], ")"),
      800: "hsla(230, 6%, 19%, ".concat(opacity[1200], ")"),
      900: "hsla(230, 7%, 17%, ".concat(opacity[1200], ")"),
      1000: "hsla(230, 8%, 15%, ".concat(opacity[1200], ")"),
      1100: "hsla(231, 12%, 12%, ".concat(opacity[1200], ")"),
      1200: "hsla(231, 17%, 8%, ".concat(opacity[1200], ")"),
      1300: "hsla(240, 5%, 4%, ".concat(opacity[1200], ")"),
      a25: "hsla(228, 4%, 76%, ".concat(opacity[50], ")"),
      a50: "hsla(228, 4%, 76%, ".concat(opacity[100], ")"),
      a75: "hsla(228, 4%, 76%, ".concat(opacity[200], ")"),
      a100: "hsla(228, 4%, 76%, ".concat(opacity[300], ")"),
      a200: "hsla(228, 4%, 76%, ".concat(opacity[500], ")"),
      a400: "hsla(228, 4%, 76%, ".concat(opacity[500], ")")
    },
    white: {
      10: "hsla(0, 0%, 100%, ".concat(opacity[100], ")"),
      25: "hsla(0, 0%, 100%, ".concat(opacity[200], ")"),
      50: "hsla(0, 0%, 100%, ".concat(opacity[300], ")"),
      100: "hsla(0, 0%, 100%, ".concat(opacity[500], ")"),
      200: "hsla(0, 0%, 100%, ".concat(opacity[600], ")"),
      300: "hsla(0, 0%, 100%, ".concat(opacity[800], ")"),
      400: "hsla(0, 0%, 100%, ".concat(opacity[1000], ")"),
      450: "hsla(0, 0%, 100%, ".concat(opacity[1100], ")"),
      500: "hsla(0, 0%, 100%, ".concat(opacity[1200], ")")
    },
    black: {
      10: "hsla(0, 0%, 0%, ".concat(opacity[100], ")"),
      25: "hsla(0, 0%, 0%, ".concat(opacity[200], ")"),
      50: "hsla(0, 0%, 0%, ".concat(opacity[300], ")"),
      100: "hsla(0, 0%, 0%, ".concat(opacity[500], ")"),
      200: "hsla(0, 0%, 0%, ".concat(opacity[700], ")"),
      300: "hsla(0, 0%, 0%, ".concat(opacity[900], ")"),
      400: "hsla(0, 0%, 0%, ".concat(opacity[1000], ")"),
      450: "hsla(0, 0%, 0%, ".concat(opacity[1100], ")"),
      500: "hsla(0, 0%, 0%, ".concat(opacity[1200], ")")
    }
  }
};

export { colors };
