// @razorpay/blade/build/lib/web/production/components/Dropdown/DropdownHeaderFooter.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default from 'react';
import { useDropdown } from './useDropdown.js';
import { BaseFooter } from '../BaseHeaderFooter/BaseFooter.js';
import { BaseHeader } from '../BaseHeaderFooter/BaseHeader.js';
import '../Box/BaseBox/index.js';
import '../../utils/index.js';
import '../../utils/assignWithoutSideEffects/index.js';
import '../../utils/makeAccessible/index.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import { jsx } from 'react/jsx-runtime';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { isReactNative } from '../../utils/platform/isReactNative.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { assignWithoutSideEffects } from '../../utils/assignWithoutSideEffects/assignWithoutSideEffects.js';
import { makeAccessible } from '../../utils/makeAccessible/makeAccessible.web.js';

var _excluded = ["title", "subtitle", "leading", "titleSuffix", "trailing", "testID", "children"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _DropdownHeader = function _DropdownHeader(_ref) {
  var title = _ref.title,
    subtitle = _ref.subtitle,
    leading = _ref.leading,
    titleSuffix = _ref.titleSuffix,
    trailing = _ref.trailing,
    testID = _ref.testID,
    children = _ref.children,
    rest = _objectWithoutProperties(_ref, _excluded);
  var _useDropdown = useDropdown(),
    hasAutoCompleteInHeader = _useDropdown.hasAutoCompleteInHeader,
    setShouldIgnoreBlurAnimation = _useDropdown.setShouldIgnoreBlurAnimation;
  return /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread({
    flexShrink: 0
  }, isReactNative() ? {} : {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    onMouseDown: function onMouseDown(e) {
      // we don't want focus to ever move on header because its static element except when autocomplete is present
      if (!hasAutoCompleteInHeader) {
        e.preventDefault();
      } else {
        setShouldIgnoreBlurAnimation(false);
      }
    }
  }), {}, {
    children: /*#__PURE__*/jsx(BaseHeader, _objectSpread(_objectSpread({
      title: title,
      subtitle: subtitle,
      leading: leading,
      trailing: trailing,
      titleSuffix: titleSuffix,
      metaComponentName: MetaConstants.DropdownHeader,
      testID: testID
      // back button
      ,
      showBackButton: false
      // close button
      ,
      showCloseButton: false
    }, makeAnalyticsAttribute(rest)), {}, {
      children: children
    }))
  }));
};
var DropdownHeader = /*#__PURE__*/assignWithoutSideEffects(_DropdownHeader, {
  componentId: 'DropdownHeader'
});
var _DropdownFooter = function _DropdownFooter(_ref2) {
  var children = _ref2.children,
    testID = _ref2.testID;
  var _useDropdown2 = useDropdown(),
    setHasFooterAction = _useDropdown2.setHasFooterAction,
    isOpen = _useDropdown2.isOpen;
  var footerRef = React__default.useRef(null);
  React__default.useEffect(function () {
    setHasFooterAction(true);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  return /*#__PURE__*/jsx(BaseBox
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  , _objectSpread(_objectSpread({
    ref: footerRef
  }, makeAccessible({
    role: isReactNative() ? undefined : 'group'
  })), {}, {
    children: /*#__PURE__*/jsx(BaseFooter, {
      metaComponentName: MetaConstants.DropdownFooter,
      testID: testID,
      children: isOpen ? children : null
    })
  }));
};
var DropdownFooter = /*#__PURE__*/assignWithoutSideEffects(_DropdownFooter, {
  componentId: 'DropdownFooter'
});

export { DropdownFooter, DropdownHeader };
