// @razorpay/blade/build/lib/web/production/components/Dropdown/FilterChipGroup.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import { useState } from 'react';
import { useListViewFilterContext } from '../ListView/ListViewFiltersContext.web.js';
import { FilterChipGroupProvider } from './FilterChipGroupContext.web.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import '../Box/BaseBox/index.js';
import '../../utils/metaAttribute/index.js';
import '../Link/index.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import Link from '../Link/Link/Link.js';

var _excluded = ["testID", "children", "showClearButton", "onClearButtonClick"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var FilterChipGroup = function FilterChipGroup(_ref) {
  var testID = _ref.testID,
    children = _ref.children,
    _ref$showClearButton = _ref.showClearButton,
    showClearButton = _ref$showClearButton === void 0 ? true : _ref$showClearButton,
    onClearButtonClick = _ref.onClearButtonClick,
    rest = _objectWithoutProperties(_ref, _excluded);
  var _useState = useState([]),
    _useState2 = _slicedToArray(_useState, 2),
    filterChipGroupSelectedFilters = _useState2[0],
    setFilterChipGroupSelectedFilters = _useState2[1];
  var _useState3 = useState(0),
    _useState4 = _slicedToArray(_useState3, 2),
    clearFilterCallbackTriggerer = _useState4[0],
    setClearFilterCallbackTriggerer = _useState4[1];
  var _useListViewFilterCon = useListViewFilterContext(),
    selectedFiltersCount = _useListViewFilterCon.selectedFiltersCount,
    setListViewSelectedFilters = _useListViewFilterCon.setListViewSelectedFilters;
  var handleClearButtonClick = function handleClearButtonClick() {
    onClearButtonClick === null || onClearButtonClick === void 0 || onClearButtonClick();
    setListViewSelectedFilters({});
    setFilterChipGroupSelectedFilters([]);
    setClearFilterCallbackTriggerer(function (prev) {
      return prev + 1;
    });
  };
  return /*#__PURE__*/jsx(FilterChipGroupProvider, {
    value: {
      filterChipGroupSelectedFilters: filterChipGroupSelectedFilters,
      setFilterChipGroupSelectedFilters: setFilterChipGroupSelectedFilters,
      clearFilterCallbackTriggerer: clearFilterCallbackTriggerer,
      setClearFilterCallbackTriggerer: setClearFilterCallbackTriggerer
    },
    children: /*#__PURE__*/jsxs(BaseBox, _objectSpread(_objectSpread(_objectSpread({}, metaAttribute({
      name: MetaConstants.FilterChipGroup,
      testID: testID
    })), makeAnalyticsAttribute(rest)), {}, {
      display: "flex",
      padding: ['spacing.4', 'spacing.1'],
      alignItems: "center",
      justifyContent: "flex-start",
      width: "100%",
      gap: "spacing.3",
      flexWrap: "wrap",
      children: [children, showClearButton && (filterChipGroupSelectedFilters.length > 0 || selectedFiltersCount > 0) ? /*#__PURE__*/jsx(Link, {
        size: "xsmall",
        color: "neutral",
        onClick: handleClearButtonClick,
        children: "Clear Filter".concat(filterChipGroupSelectedFilters.length > 1 || selectedFiltersCount > 1 ? 's' : '')
      }) : null]
    }))
  });
};

export { FilterChipGroup };
