// @razorpay/blade/build/lib/web/development/components/BottomSheet/BottomSheetHeader.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default from 'react';
import { ComponentIds } from './componentIds.js';
import { useBottomSheetContext, useBottomSheetAndDropdownGlue } from './BottomSheetContext.js';
import { BottomSheetEmptyHeader } from './BottomSheetCommon.js';
import { BaseHeader } from '../BaseHeaderFooter/BaseHeader.js';
import '../Box/BaseBox/index.js';
import '../../utils/metaAttribute/index.js';
import { useIsomorphicLayoutEffect } from '../../utils/useIsomorphicLayoutEffect.js';
import '../../utils/assignWithoutSideEffects/index.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import { jsx } from 'react/jsx-runtime';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { assignWithoutSideEffects } from '../../utils/assignWithoutSideEffects/assignWithoutSideEffects.js';

var _excluded = ["title", "subtitle", "leading", "trailing", "titleSuffix", "showBackButton", "onBackButtonClick", "children"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _BottomSheetHeader = function _BottomSheetHeader(_ref) {
  var title = _ref.title,
    subtitle = _ref.subtitle,
    leading = _ref.leading,
    trailing = _ref.trailing,
    titleSuffix = _ref.titleSuffix,
    _ref$showBackButton = _ref.showBackButton,
    showBackButton = _ref$showBackButton === void 0 ? false : _ref$showBackButton,
    onBackButtonClick = _ref.onBackButtonClick,
    children = _ref.children,
    dataAnalyticsProps = _objectWithoutProperties(_ref, _excluded);
  var _useBottomSheetContex = useBottomSheetContext(),
    setHeaderHeight = _useBottomSheetContex.setHeaderHeight,
    isOpen = _useBottomSheetContex.isOpen,
    close = _useBottomSheetContex.close,
    isDismissible = _useBottomSheetContex.isDismissible,
    bind = _useBottomSheetContex.bind,
    setIsHeaderEmpty = _useBottomSheetContex.setIsHeaderEmpty,
    defaultInitialFocusRef = _useBottomSheetContex.defaultInitialFocusRef;
  var bottomSheetAndDropdownGlue = useBottomSheetAndDropdownGlue();
  var ref = React__default.useRef(null);
  useIsomorphicLayoutEffect(function () {
    if (!ref.current) return;
    if (!title && !subtitle) return;
    setHeaderHeight(ref.current.getBoundingClientRect().height);
  }, [ref, isOpen]);
  var isHeaderEmpty = !(title || subtitle || leading || trailing || showBackButton || children);
  React__default.useEffect(function () {
    setIsHeaderEmpty(isHeaderEmpty);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isHeaderEmpty]);
  return /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread(_objectSpread({
    ref: ref,
    overflow: isHeaderEmpty ? 'visible' : 'auto',
    flexShrink: 0
  }, metaAttribute({
    name: MetaConstants.BottomSheetHeader
  })), makeAnalyticsAttribute(dataAnalyticsProps)), {}, {
    children: isHeaderEmpty ? /*#__PURE__*/jsx(BottomSheetEmptyHeader, {
      ref: defaultInitialFocusRef
    }) : /*#__PURE__*/jsx(BaseHeader, _objectSpread(_objectSpread({
      title: title,
      subtitle: subtitle,
      leading: leading,
      trailing: trailing,
      titleSuffix: titleSuffix
      // we don't set focus on close button when it has AutoComplete inside.
      // We set focus on AutoComplete instead inside AutoComplete component
      ,
      closeButtonRef: bottomSheetAndDropdownGlue !== null && bottomSheetAndDropdownGlue !== void 0 && bottomSheetAndDropdownGlue.hasAutoCompleteInHeader ? undefined : defaultInitialFocusRef
      // back button
      ,
      showBackButton: showBackButton,
      onBackButtonClick: onBackButtonClick
      // close button
      ,
      showCloseButton: isDismissible,
      onCloseButtonClick: close
    }, bind === null || bind === void 0 ? void 0 : bind()), {}, {
      children: children
    }))
  }));
};
var BottomSheetHeader = /*#__PURE__*/assignWithoutSideEffects(_BottomSheetHeader, {
  componentId: ComponentIds.BottomSheetHeader
});

export { BottomSheetHeader };
