// @razorpay/blade/build/lib/web/development/components/BottomSheet/getBottomSheetGrabHandleStyles.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import '../../utils/index.js';
import '../../utils/makeSize/index.js';
import '../../utils/makeSpace/index.js';
import '../../tokens/global/index.js';
import { makeSize } from '../../utils/makeSize/makeSize.js';
import { size } from '../../tokens/global/size.js';
import { makeSpace } from '../../utils/makeSpace/makeSpace.js';
import { isReactNative } from '../../utils/platform/isReactNative.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var getHandlePartStyles = function getHandlePartStyles(_ref) {
  var theme = _ref.theme;
  return {
    margin: 'auto',
    content: "''",
    width: makeSize(size[56]),
    height: makeSize(size[4]),
    backgroundColor: theme.colors.interactive.background.gray.faded,
    // TODO: we do not have 16px radius token
    borderRadius: makeSpace(theme.spacing[5])
  };
};
var getBottomSheetGrabHandleStyles = function getBottomSheetGrabHandleStyles(_ref2) {
  var theme = _ref2.theme,
    isHeaderFloating = _ref2.isHeaderFloating;
  return _objectSpread({
    position: isHeaderFloating ? 'absolute' : 'relative',
    flexShrink: 0,
    paddingTop: makeSpace(theme.spacing[4]),
    marginBottom: makeSpace(theme.spacing[2]),
    touchAction: 'none',
    width: '100%',
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
    zIndex: 100
  }, isReactNative() ? undefined : {
    ':after': getHandlePartStyles({
      theme: theme
    })
  });
};

export { getBottomSheetGrabHandleStyles, getHandlePartStyles };
