// @razorpay/blade/build/lib/web/development/components/Drawer/DrawerSubcomponents.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default from 'react';
import { drawerComponentIds } from './drawerComponentIds.js';
import { DrawerContext } from './DrawerContext.js';
import { useDrawerStack } from './StackProvider.js';
import { BaseHeader } from '../BaseHeaderFooter/BaseHeader.js';
import { BaseFooter } from '../BaseHeaderFooter/BaseFooter.js';
import '../Box/index.js';
import '../../utils/assignWithoutSideEffects/index.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import '../../utils/index.js';
import { jsx } from 'react/jsx-runtime';
import useTheme from '../BladeProvider/useTheme.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { assignWithoutSideEffects } from '../../utils/assignWithoutSideEffects/assignWithoutSideEffects.js';
import { Box } from '../Box/Box.js';

var _excluded = ["title", "subtitle", "leading", "trailing", "titleSuffix", "children", "color", "showDivider"],
  _excluded2 = ["children", "showDivider"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _DrawerHeader = function _DrawerHeader(_ref) {
  var title = _ref.title,
    subtitle = _ref.subtitle,
    leading = _ref.leading,
    trailing = _ref.trailing,
    titleSuffix = _ref.titleSuffix,
    children = _ref.children,
    _ref$color = _ref.color,
    color = _ref$color === void 0 ? 'information' : _ref$color,
    _ref$showDivider = _ref.showDivider,
    showDivider = _ref$showDivider === void 0 ? true : _ref$showDivider,
    rest = _objectWithoutProperties(_ref, _excluded);
  var _React$useContext = React__default.useContext(DrawerContext),
    close = _React$useContext.close,
    closeButtonRef = _React$useContext.closeButtonRef,
    stackingLevel = _React$useContext.stackingLevel,
    isExiting = _React$useContext.isExiting;
  var _useDrawerStack = useDrawerStack(),
    drawerStack = _useDrawerStack.drawerStack;
  var _useTheme = useTheme(),
    theme = _useTheme.theme;
  var closeAllDrawers = function closeAllDrawers() {
    for (var _i = 0, _Object$values = Object.values(drawerStack); _i < _Object$values.length; _i++) {
      var onDismiss = _Object$values[_i];
      onDismiss();
    }
  };
  var isStackedDrawer = stackingLevel && stackingLevel > 1;
  var isAtleastOneDrawerOpen = Object.keys(drawerStack).length > 0;
  var backgroundGradient = "radial-gradient(150% 100% at 50% 100%, ".concat(theme.colors.transparent, " 0%, ").concat(theme.colors.feedback.background[color].subtle, " 100%)");
  // This condition is to avoid back button disappear while stacked drawer is in the exiting transition
  var isDrawerExiting = isAtleastOneDrawerOpen && isExiting && stackingLevel !== 1;
  return /*#__PURE__*/jsx(BaseHeader
  // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
  , _objectSpread(_objectSpread({
    showBackButton: isStackedDrawer || isDrawerExiting,
    showCloseButton: true,
    closeButtonRef: closeButtonRef,
    onCloseButtonClick: function onCloseButtonClick() {
      return closeAllDrawers();
    },
    onBackButtonClick: function onBackButtonClick() {
      return close();
    },
    title: title,
    size: "xlarge",
    titleSuffix: titleSuffix,
    subtitle: subtitle,
    leading: leading,
    trailing: trailing,
    backgroundImage: backgroundGradient,
    showDivider: showDivider
  }, makeAnalyticsAttribute(rest)), {}, {
    children: children
  }));
};

/**
 * #### Usage
 *
 * ```jsx
 * <DrawerHeader
 *  title="Announcements"
 *  subtitle="Checkout what's new in Razorpay"
 *  titleSuffix={<DrawerHeaderBadge>New</DrawerHeaderBadge>}
 *  leading={<DrawerHeaderIcon icon={AnnouncementIcon} />}
 *  trailing={<Button icon={DownloadIcon} />}
 * />
 * ```
 *
 */
var DrawerHeader = /*#__PURE__*/assignWithoutSideEffects(_DrawerHeader, {
  componentId: drawerComponentIds.DrawerHeader
});
var drawerPadding = 'spacing.6';
var _DrawerBody = function _DrawerBody(_ref2) {
  var children = _ref2.children;
  return /*#__PURE__*/jsx(Box, {
    padding: drawerPadding,
    overflow: "auto",
    flex: "1",
    children: children
  });
};
var DrawerBody = /*#__PURE__*/assignWithoutSideEffects(_DrawerBody, {
  componentId: drawerComponentIds.DrawerBody
});
var _DrawerFooter = function _DrawerFooter(_ref3) {
  var children = _ref3.children,
    _ref3$showDivider = _ref3.showDivider,
    showDivider = _ref3$showDivider === void 0 ? true : _ref3$showDivider,
    rest = _objectWithoutProperties(_ref3, _excluded2);
  return /*#__PURE__*/jsx(Box, _objectSpread(_objectSpread({
    position: "sticky"
  }, makeAnalyticsAttribute(rest)), {}, {
    children: /*#__PURE__*/jsx(BaseFooter, {
      showDivider: showDivider,
      children: children
    })
  }));
};

/**
 * #### Usage
 *
 * ```jsx
 * <DrawerFooter>
 *   <Button variant="primary" isFullWidth>
 *     Continue
 *   </Button>
 * </DrawerFooter>
 * ```
 *
 */
var DrawerFooter = /*#__PURE__*/assignWithoutSideEffects(_DrawerFooter, {
  componentId: drawerComponentIds.DrawerFooter
});

export { DrawerBody, DrawerFooter, DrawerHeader, drawerPadding };
