// @razorpay/blade/build/lib/web/development/components/Link/BaseLink/StyledBaseLink.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import styled from 'styled-components';
import getStyledLinkStyles from './getStyledLinkStyles.js';
import getIn from '../../../utils/lodashButBetter/get.js';
import '../../../utils/index.js';
import '../../Box/styledProps/index.js';
import '../../../utils/makeBorderSize/index.js';
import '../../../utils/makeMotionTime/index.web.js';
import { omitPropsFromHTML } from '../../../utils/omitPropsFromHTML/index.js';
import { useStyledProps } from '../../Box/styledProps/useStyledProps.js';
import { makeBorderSize } from '../../../utils/makeBorderSize/makeBorderSize.js';
import { castWebType } from '../../../utils/platform/castUtils.js';
import { makeMotionTime } from '../../../utils/makeMotionTime/makeMotionTime.web.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StyledLink = /*#__PURE__*/styled.button.withConfig({
  shouldForwardProp: omitPropsFromHTML,
  displayName: 'StyledLink'
}).attrs(function (props) {
  return _objectSpread({}, props.accessibilityProps);
}).withConfig({
  displayName: "StyledBaseLinkweb__StyledLink",
  componentId: "sc-1yj1z8h-0"
})(function (props) {
  var styledPropsCSSObject = useStyledProps(props);
  return _objectSpread(_objectSpread({}, getStyledLinkStyles(props)), {}, {
    display: 'inline-block',
    borderRadius: makeBorderSize(props.theme.border.radius.small),
    transitionProperty: 'box-shadow',
    transitionTimingFunction: getIn(props.theme.motion, props.motionEasing),
    transitionDuration: castWebType(makeMotionTime(getIn(props.theme.motion, props.motionDuration))),
    '.content-container': {
      width: 'max-content',
      borderRadius: makeBorderSize(props.theme.border.radius.small)
    },
    '&:focus-visible': {
      '.content-container': {
        boxShadow: "0px 0px 0px 4px ".concat(props.focusRingColor)
      }
    },
    '*': {
      transitionProperty: 'color, fill',
      transitionTimingFunction: getIn(props.theme.motion, props.motionEasing),
      transitionDuration: castWebType(makeMotionTime(getIn(props.theme.motion, props.motionDuration)))
    }
  }, styledPropsCSSObject);
});

export { StyledLink as default };
