// @razorpay/blade/build/lib/web/production/components/SideNav/tokens.js
import '../../tokens/global/index.js';
import '../../utils/index.js';
import { size } from '../../tokens/global/size.js';
import { makeMotionTime } from '../../utils/makeMotionTime/makeMotionTime.web.js';

var classes = {
  SHOW_WHEN_COLLAPSED: 'show-when-collapsed',
  HIDE_WHEN_COLLAPSED: 'hide-when-collapsed',
  COLLAPSED: 'collapsed',
  TRANSITIONING: 'transitioning',
  L1_ITEM_WRAPPER: 'l1-item-wrapper',
  SHOW_ON_LINK_HOVER: 'show-on-link-hover',
  STYLED_NAV_LINK: 'styled-nav-link'
};
var SKIP_NAV_ID = 'blade-side-nav-skip';
var COLLAPSED_L1_WIDTH = size['56'];
var SIDE_NAV_EXPANDED_L1_WIDTH_XL = size['264'];
var SIDE_NAV_EXPANDED_L1_WIDTH_BASE = size['240'];
var NAV_ITEM_HEIGHT = size['40'];

// This is the delay after which transition cleanup happens for rare cases where transitionEnd is not triggered
var TRANSITION_CLEANUP_DELAY = 300; // A little more than the duration of transition end

var HOVER_AGAIN_DELAY = 500;

// Delay between mouse out from L1 and L1 collapsing again
var L1_EXIT_HOVER_DELAY = 150;
var getNavItemTransition = function getNavItemTransition(_ref) {
  var motion = _ref.motion;
  return "background-color ".concat(makeMotionTime(motion.duration['2xquick']), " ").concat(motion.easing.standard);
};

export { COLLAPSED_L1_WIDTH, HOVER_AGAIN_DELAY, L1_EXIT_HOVER_DELAY, NAV_ITEM_HEIGHT, SIDE_NAV_EXPANDED_L1_WIDTH_BASE, SIDE_NAV_EXPANDED_L1_WIDTH_XL, SKIP_NAV_ID, TRANSITION_CLEANUP_DELAY, classes, getNavItemTransition };
