// @razorpay/blade/build/lib/web/production/components/Carousel/NavigationButton/NavigationButton.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import 'react';
import { StyledNavigationButton } from './StyledNavigationButton.web.js';
import '../../Icons/index.js';
import '../../BladeProvider/index.js';
import '../../../utils/index.js';
import '../../../utils/metaAttribute/index.js';
import '../../../utils/makeAccessible/index.js';
import { jsx } from 'react/jsx-runtime';
import useTheme from '../../BladeProvider/useTheme.js';
import { isReactNative } from '../../../utils/platform/isReactNative.js';
import { metaAttribute } from '../../../utils/metaAttribute/metaAttribute.web.js';
import { makeAccessible } from '../../../utils/makeAccessible/makeAccessible.web.js';
import ChevronRightIcon from '../../Icons/ChevronRightIcon/ChevronRightIcon.js';
import ChevronLeftIcon from '../../Icons/ChevronLeftIcon/ChevronLeftIcon.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var NavigationButton = function NavigationButton(_ref) {
  var type = _ref.type,
    variant = _ref.variant,
    onClick = _ref.onClick;
  var _useTheme = useTheme(),
    platform = _useTheme.platform;
  var isMobile = platform === 'onMobile';
  var iconSize = isMobile ? 'small' : 'medium';

  // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
  var getIconColor = function getIconColor() {
    if (!isReactNative()) {
      return 'currentColor';
    }

    // on react native we cannot rely on currentColor thus need to manually pass iconColor
    var iconColor = {
      filled: 'interactive.icon.staticBlack.muted',
      stroked: 'interactive.icon.gray.normal'
    };
    return iconColor[variant];
  };
  return /*#__PURE__*/jsx(StyledNavigationButton, _objectSpread(_objectSpread(_objectSpread({
    onClick: onClick,
    variant: variant
  }, metaAttribute({
    name: 'NavigationButton'
  })), makeAccessible({
    label: type === 'previous' ? 'Previous Slide' : 'Next Slide'
  })), {}, {
    children: type === 'next' ? /*#__PURE__*/jsx(ChevronRightIcon, {
      size: iconSize,
      color: getIconColor()
    }) : /*#__PURE__*/jsx(ChevronLeftIcon, {
      size: iconSize,
      color: getIconColor()
    })
  }));
};

export { NavigationButton };
