// @razorpay/blade/build/lib/web/development/components/ListView/ListViewFilters.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default, { useState } from 'react';
import styled from 'styled-components';
import { ListViewFiltersProvider } from './ListViewFiltersContext.web.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import '../Box/BaseBox/index.js';
import '../../utils/metaAttribute/index.js';
import '../Box/index.js';
import '../Input/SearchInput/index.js';
import { useId } from '../../utils/useId.js';
import { useIsMobile } from '../../utils/useIsMobile.js';
import '../../utils/isValidAllowedChildren/index.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { getComponentId } from '../../utils/isValidAllowedChildren/isValidAllowedChildren.js';
import { SearchInput } from '../Input/SearchInput/SearchInput.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { Box } from '../Box/Box.js';

var _excluded = ["testID", "children", "quickFilters", "onSearchChange", "searchValue", "searchValuePlaceholder", "searchName", "showQuickFilters", "onShowQuickFiltersChange", "showFilters", "onShowFiltersChange", "onSearchClear", "selectedFiltersCount", "searchTrailing", "actions"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StyledQuickFilterContainer = /*#__PURE__*/styled(BaseBox).withConfig({
  displayName: "ListViewFiltersweb__StyledQuickFilterContainer",
  componentId: "igy2td-0"
})({
  /* For Webkit (Chrome, Safari) */
  '::-webkit-scrollbar': {
    display: 'none'
  },
  /* For Firefox */
  scrollbarWidth: 'none',
  /* For Edge */
  msOverflowStyle: 'none'
});
var ListViewFilters = function ListViewFilters(_ref) {
  var testID = _ref.testID,
    children = _ref.children,
    quickFilters = _ref.quickFilters,
    onSearchChange = _ref.onSearchChange,
    searchValue = _ref.searchValue,
    searchValuePlaceholder = _ref.searchValuePlaceholder,
    searchName = _ref.searchName,
    showQuickFilters = _ref.showQuickFilters,
    onShowQuickFiltersChange = _ref.onShowQuickFiltersChange,
    showFiltersProp = _ref.showFilters,
    onShowFiltersChange = _ref.onShowFiltersChange,
    onSearchClear = _ref.onSearchClear,
    _ref$selectedFiltersC = _ref.selectedFiltersCount,
    selectedFiltersCount = _ref$selectedFiltersC === void 0 ? 0 : _ref$selectedFiltersC,
    searchTrailing = _ref.searchTrailing,
    actions = _ref.actions,
    rest = _objectWithoutProperties(_ref, _excluded);
  var _useState = useState({}),
    _useState2 = _slicedToArray(_useState, 2),
    listViewSelectedFilters = _useState2[0],
    setListViewSelectedFilters = _useState2[1];
  var searchId = useId('search-input');
  var searchNameValue = searchName || searchId;
  var isMobile = useIsMobile();
  var showSearchInput = onSearchChange || onSearchClear || searchValuePlaceholder || searchName;
  var isSearchTrailingDropDown = /*#__PURE__*/React__default.isValidElement(searchTrailing) && getComponentId(searchTrailing) === 'Dropdown';
  return /*#__PURE__*/jsxs(ListViewFiltersProvider, {
    value: {
      listViewSelectedFilters: listViewSelectedFilters,
      setListViewSelectedFilters: setListViewSelectedFilters,
      selectedFiltersCount: selectedFiltersCount
    },
    children: [isMobile && showSearchInput && /*#__PURE__*/jsx(SearchInput, {
      label: "",
      value: searchValue,
      placeholder: searchValuePlaceholder,
      name: searchNameValue || searchId,
      onChange: function onChange(_ref2) {
        var name = _ref2.name,
          value = _ref2.value;
        return onSearchChange === null || onSearchChange === void 0 ? void 0 : onSearchChange({
          name: name,
          value: value
        });
      },
      onClearButtonClick: onSearchClear,
      trailing: searchTrailing
    }), /*#__PURE__*/jsxs(BaseBox, {
      children: [/*#__PURE__*/jsxs(BaseBox, _objectSpread(_objectSpread(_objectSpread({}, metaAttribute({
        name: MetaConstants.ListViewFilter,
        testID: testID
      })), makeAnalyticsAttribute(rest)), {}, {
        display: "flex",
        justifyContent: "space-between",
        children: [/*#__PURE__*/jsx(Box, {
          position: "relative",
          display: "flex",
          flexDirection: "column",
          marginRight: "spacing.3",
          flex: "1",
          minWidth: "0px",
          children: /*#__PURE__*/jsx(StyledQuickFilterContainer, {
            overflow: "scroll",
            width: "100%",
            paddingTop: isMobile ? 'spacing.4' : 'spacing.1',
            paddingBottom: "spacing.1",
            paddingLeft: "spacing.1",
            children: quickFilters
          })
        }), /*#__PURE__*/jsxs(BaseBox, {
          display: "flex",
          alignItems: "center",
          flexShrink: "0",
          children: [!isMobile && showSearchInput && /*#__PURE__*/jsx(Box, {
            width: isSearchTrailingDropDown ? '280px' : '208px',
            children: /*#__PURE__*/jsx(SearchInput, {
              label: "",
              value: searchValue,
              placeholder: searchValuePlaceholder,
              name: searchNameValue || searchId,
              onChange: function onChange(_ref3) {
                var name = _ref3.name,
                  value = _ref3.value;
                return onSearchChange === null || onSearchChange === void 0 ? void 0 : onSearchChange({
                  name: name,
                  value: value
                });
              },
              onClearButtonClick: onSearchClear,
              size: "medium",
              trailing: searchTrailing
            })
          }), actions]
        })]
      })), /*#__PURE__*/jsx(BaseBox, {
        display: "flex",
        children: children
      })]
    })]
  });
};

export { ListViewFilters };
