// @razorpay/blade/build/lib/web/development/components/Carousel/Indicators/getIndicatorButtonStyles.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import '../../../tokens/global/index.js';
import '../../../utils/index.js';
import '../../../utils/getFocusRingStyles/index.js';
import { castWebType } from '../../../utils/platform/castUtils.js';
import { makeMotionTime } from '../../../utils/makeMotionTime/makeMotionTime.web.js';
import { makeSpace } from '../../../utils/makeSpace/makeSpace.js';
import { makeSize } from '../../../utils/makeSize/makeSize.js';
import { size } from '../../../tokens/global/size.js';
import { isReactNative } from '../../../utils/platform/isReactNative.js';
import { getFocusRingStyles } from '../../../utils/getFocusRingStyles/getFocusRingStyles.web.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var getIndicatorButtonStyles = function getIndicatorButtonStyles(_ref) {
  var theme = _ref.theme,
    isActive = _ref.isActive,
    variant = _ref.variant;
  var backgroundColor = {
    gray: theme.colors.interactive.icon.gray.muted,
    white: theme.colors.interactive.icon.staticWhite.normal,
    blue: theme.colors.interactive.icon.primary.subtle
  };
  var easing = castWebType(theme.motion.easing.standard);
  var duration = castWebType(makeMotionTime(theme.motion.duration.gentle));
  return _objectSpread(_objectSpread({
    border: 'none',
    cursor: 'pointer',
    padding: makeSpace(theme.spacing[0]),
    borderRadius: theme.border.radius.max,
    backgroundColor: isActive ? backgroundColor[variant] : theme.colors.overlay.background.moderate,
    width: isActive ? makeSize(size[18]) : makeSize(size[6]),
    height: makeSize(size[6])
  }, isReactNative() && {
    width: undefined
  }), !isReactNative() && {
    transitionProperty: 'width',
    transitionDuration: duration,
    transitionTimingFunction: easing,
    '&:before': {
      content: "\"\"",
      // increasing the hit area, just random numbers
      padding: makeSize(size[10]),
      paddingTop: makeSize(size[20])
    },
    '&:focus-visible': _objectSpread({}, getFocusRingStyles({
      theme: theme
    }))
  });
};

export { getIndicatorButtonStyles };
