// @razorpay/blade/build/lib/web/development/components/Carousel/Indicators/Indicators.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { useCarouselContext } from '../CarouselContext.js';
import { IndicatorButton } from './IndicatorButton.js';
import '../../Box/BaseBox/index.js';
import '../../../utils/makeAccessible/index.js';
import { jsx } from 'react/jsx-runtime';
import { BaseBox } from '../../Box/BaseBox/BaseBox.web.js';
import { makeAccessible } from '../../../utils/makeAccessible/makeAccessible.web.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Indicators = function Indicators(props) {
  var _useCarouselContext = useCarouselContext(),
    carouselId = _useCarouselContext.carouselId,
    isResponsive = _useCarouselContext.isResponsive,
    visibleItems = _useCarouselContext.visibleItems;
  return /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread({
    display: "flex",
    flexDirection: "row",
    gap: "spacing.2"
  }, makeAccessible({
    role: 'tablist',
    label: 'Slides'
  })), {}, {
    children: new Array(props.totalItems).fill(0).map(function (_, idx) {
      var _visibleItems = visibleItems;
      if (isResponsive) {
        _visibleItems = 1;
      }
      return /*#__PURE__*/jsx(IndicatorButton, _objectSpread(_objectSpread({}, makeAccessible({
        role: 'tab',
        label: "Slide ".concat(idx * _visibleItems + 1),
        selected: idx === props.activeIndex,
        controls: "".concat(carouselId, "-carousel-item-").concat(idx * _visibleItems)
      })), {}, {
        slideIndex: idx * _visibleItems,
        isActive: idx === props.activeIndex,
        onClick: function onClick() {
          var _props$onClick;
          return props === null || props === void 0 || (_props$onClick = props.onClick) === null || _props$onClick === void 0 ? void 0 : _props$onClick.call(props, idx);
        },
        variant: props.variant
      }), idx);
    })
  }));
};

export { Indicators };
