// @razorpay/blade/build/lib/web/development/components/ActionList/styles/StyledListBoxWrapper.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import styled from 'styled-components';
import { getBaseListBoxWrapperStyles } from './getBaseListBoxWrapperStyles.js';
import '../../Box/BaseBox/index.js';
import { BaseBox } from '../../Box/BaseBox/BaseBox.web.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StyledListBoxWrapper = /*#__PURE__*/styled(BaseBox).withConfig({
  displayName: "StyledListBoxWrapperweb__StyledListBoxWrapper",
  componentId: "sc-1eo4pr6-0"
})(function (props) {
  return _objectSpread(_defineProperty(_defineProperty({}, "& [role=group]:last-child > [role=separator]:last-child", {
    display: 'none'
  }), "overflowY", 'auto'), getBaseListBoxWrapperStyles({
    theme: props.theme,
    isInBottomSheet: props.isInBottomSheet
  }));
});

export { StyledListBoxWrapper };
