// @razorpay/blade/build/lib/web/development/components/DatePicker/QuickSelection/PresetSideBar.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { QuickSelectionItem } from './QuickSelectionItem.web.js';
import '../../Box/index.js';
import '../../Box/BaseBox/index.js';
import '../../Chip/index.js';
import '../../Divider/index.js';
import '../../../tokens/global/index.js';
import '../../../utils/index.js';
import '../../../utils/makeAccessible/index.js';
import { jsx, Fragment, jsxs } from 'react/jsx-runtime';
import { Box } from '../../Box/Box.js';
import { Divider } from '../../Divider/Divider.js';
import { ChipGroup } from '../../Chip/ChipGroup.js';
import { Chip } from '../../Chip/Chip.js';
import { BaseBox } from '../../Box/BaseBox/BaseBox.web.js';
import { makeSpace } from '../../../utils/makeSpace/makeSpace.js';
import { size } from '../../../tokens/global/size.js';
import { makeAccessible } from '../../../utils/makeAccessible/makeAccessible.web.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var PresetSideBar = function PresetSideBar(_ref) {
  var onSelection = _ref.onSelection,
    isMobile = _ref.isMobile,
    presetStates = _ref.presetStates;
  if (presetStates.length === 0) return /*#__PURE__*/jsx(Fragment, {});
  if (isMobile) {
    return /*#__PURE__*/jsxs(Box, {
      children: [/*#__PURE__*/jsx(Divider, {
        marginTop: "spacing.4"
      }), /*#__PURE__*/jsx(ChipGroup, {
        marginTop: "spacing.7",
        size: "small",
        selectionType: "single",
        accessibilityLabel: "Select Presets",
        onChange: function onChange(_ref2) {
          var values = _ref2.values;
          var selectedPreset = presetStates.find(function (state) {
            return state.preset.label === values[0];
          });
          if (selectedPreset) {
            onSelection(selectedPreset.preset.value);
          }
        },
        children: presetStates.map(function (_ref3, index) {
          var preset = _ref3.preset;
          return /*#__PURE__*/jsx(Chip, {
            value: preset.label,
            "data-analytics-name": preset.label,
            children: preset.label
          }, index);
        })
      })]
    });
  }
  return /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread({
    padding: "spacing.5",
    display: "flex",
    flexDirection: "column",
    gap: "spacing.2",
    backgroundColor: "surface.background.gray.moderate",
    minWidth: makeSpace(size[160]),
    borderRightColor: "surface.border.gray.muted",
    borderRightStyle: "solid",
    borderRightWidth: "thin"
  }, makeAccessible({
    role: 'listbox',
    label: 'Select Presets'
  })), {}, {
    children: presetStates.map(function (_ref4, index) {
      var preset = _ref4.preset,
        isSelected = _ref4.isSelected;
      return /*#__PURE__*/jsx(QuickSelectionItem, {
        isSelected: isSelected,
        onClick: function onClick() {
          return onSelection(preset.value);
        },
        children: preset.label
      }, index);
    })
  }));
};

export { PresetSideBar };
