// @razorpay/blade/build/lib/web/production/components/Button/Button/Button.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default from 'react';
import '../BaseButton/index.js';
import '../../Box/styledProps/index.js';
import '../../../utils/assignWithoutSideEffects/index.js';
import '../../../utils/makeAnalyticsAttribute/index.js';
import { jsx } from 'react/jsx-runtime';
import BaseButton from '../BaseButton/BaseButton.js';
import { getStyledProps } from '../../Box/styledProps/getStyledProps.js';
import { makeAnalyticsAttribute } from '../../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { assignWithoutSideEffects } from '../../../utils/assignWithoutSideEffects/assignWithoutSideEffects.js';

var _excluded = ["children", "icon", "iconPosition", "isDisabled", "isFullWidth", "isLoading", "href", "target", "rel", "onClick", "size", "type", "variant", "color", "accessibilityLabel", "role", "testID", "onBlur", "onFocus", "onMouseLeave", "onMouseMove", "onMouseDown", "onPointerDown", "onPointerEnter", "onTouchStart", "onTouchEnd"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _Button = function _Button(// While adding any prop here, make sure to handle it in DropdownButton as well
_ref, ref) {
  var children = _ref.children,
    icon = _ref.icon,
    _ref$iconPosition = _ref.iconPosition,
    iconPosition = _ref$iconPosition === void 0 ? 'left' : _ref$iconPosition,
    _ref$isDisabled = _ref.isDisabled,
    isDisabled = _ref$isDisabled === void 0 ? false : _ref$isDisabled,
    _ref$isFullWidth = _ref.isFullWidth,
    isFullWidth = _ref$isFullWidth === void 0 ? false : _ref$isFullWidth,
    _ref$isLoading = _ref.isLoading,
    isLoading = _ref$isLoading === void 0 ? false : _ref$isLoading,
    href = _ref.href,
    target = _ref.target,
    rel = _ref.rel,
    onClick = _ref.onClick,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'medium' : _ref$size,
    _ref$type = _ref.type,
    type = _ref$type === void 0 ? 'button' : _ref$type,
    _ref$variant = _ref.variant,
    variant = _ref$variant === void 0 ? 'primary' : _ref$variant,
    _ref$color = _ref.color,
    color = _ref$color === void 0 ? 'primary' : _ref$color,
    accessibilityLabel = _ref.accessibilityLabel,
    role = _ref.role,
    testID = _ref.testID,
    onBlur = _ref.onBlur,
    onFocus = _ref.onFocus,
    onMouseLeave = _ref.onMouseLeave,
    onMouseMove = _ref.onMouseMove,
    onMouseDown = _ref.onMouseDown,
    onPointerDown = _ref.onPointerDown,
    onPointerEnter = _ref.onPointerEnter,
    onTouchStart = _ref.onTouchStart,
    onTouchEnd = _ref.onTouchEnd,
    rest = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/jsx(BaseButton, _objectSpread(_objectSpread(_objectSpread(_objectSpread({}, icon ? {
    icon: icon,
    children: children
  } : {
    children: children
  }), getStyledProps(rest)), makeAnalyticsAttribute(rest)), {}, {
    ref: ref,
    href: href,
    target: target,
    rel: rel,
    accessibilityProps: {
      label: accessibilityLabel,
      describedBy: rest['aria-describedby'],
      expanded: rest['aria-expanded'],
      hasPopup: rest['aria-haspopup'],
      role: role
    },
    iconPosition: iconPosition,
    color: color,
    isDisabled: isDisabled,
    isFullWidth: isFullWidth,
    onClick: onClick,
    size: size,
    type: type,
    variant: variant,
    isLoading: isLoading,
    testID: testID,
    onBlur: onBlur,
    onFocus: onFocus,
    onMouseLeave: onMouseLeave,
    onMouseMove: onMouseMove,
    onMouseDown: onMouseDown,
    onPointerDown: onPointerDown,
    onPointerEnter: onPointerEnter,
    onTouchStart: onTouchStart,
    onTouchEnd: onTouchEnd
  }));
};
var Button = /*#__PURE__*/assignWithoutSideEffects(/*#__PURE__*/React__default.forwardRef(_Button), {
  displayName: 'Button',
  componentId: 'Button'
});

export { Button as default };
