// @razorpay/blade/build/lib/web/production/components/Accordion/AccordionButton.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { StyledAccordionButton } from './StyledAccordionButton.web.js';
import { useAccordion } from './AccordionContext.js';
import { AccordionItemHeader } from './AccordionItemHeader.js';
import '../Box/BaseBox/index.js';
import '../../utils/metaAttribute/index.js';
import '../Typography/index.js';
import { useCollapsible } from '../Collapsible/CollapsibleContext.js';
import '../../utils/makeAccessible/index.js';
import '../../utils/assignWithoutSideEffects/index.js';
import '../../utils/logger/index.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import { Text } from '../Typography/Text/Text.js';
import { throwBladeError } from '../../utils/logger/logger.js';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { makeAccessible } from '../../utils/makeAccessible/makeAccessible.web.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { MAKE_ANALYTICS_CONSTANTS } from '../../utils/makeAnalyticsAttribute/makeAnalyticsConstants.js';
import { assignWithoutSideEffects } from '../../utils/assignWithoutSideEffects/assignWithoutSideEffects.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _AccordionButton = function _AccordionButton(_ref) {
  var index = _ref.index,
    Icon = _ref.icon,
    title = _ref.title,
    isDeprecatedAPI = _ref.isDeprecatedAPI,
    header = _ref.header,
    isDisabled = _ref.isDisabled;
  var _useCollapsible = useCollapsible(),
    onExpandChange = _useCollapsible.onExpandChange,
    isExpanded = _useCollapsible.isExpanded,
    collapsibleBodyId = _useCollapsible.collapsibleBodyId;
  var _useAccordion = useAccordion(),
    showNumberPrefix = _useAccordion.showNumberPrefix,
    expandedIndex = _useAccordion.expandedIndex,
    size = _useAccordion.size;
  var toggleCollapse = function toggleCollapse() {
    return onExpandChange(!isExpanded);
  };
  var onClick = function onClick() {
    return toggleCollapse();
  };
  var _index = typeof index === 'number' && showNumberPrefix ?
  /*#__PURE__*/
  // we have to add -2px margin to align the number with title of BaseHeader
  jsxs(Text, {
    size: size,
    weight: "semibold",
    marginTop: "-2px",
    as: "span",
    children: [index + 1, "."]
  }) : null;
  var _icon = Icon && /*#__PURE__*/jsx(Icon, {
    size: size,
    color: "surface.icon.gray.normal",
    marginY: "spacing.2"
  });
  if (false) {
    if (_index && _icon) {
      throwBladeError({
        message: "showNumberPrefix and icon shouldn't be used together",
        moduleName: 'Accordion'
      });
    }
  }
  var isItemExpanded = expandedIndex === index;
  return /*#__PURE__*/jsx(BaseBox
  // a11y guidelines suggest having an apt heading surround a button but heading level is hardcoded here
  , _objectSpread(_objectSpread({}, makeAccessible({
    role: 'heading',
    level: 3
  })), {}, {
    width: "100%",
    children: /*#__PURE__*/jsx(StyledAccordionButton, _objectSpread(_objectSpread(_objectSpread({
      type: "button",
      isExpanded: isItemExpanded,
      disabled: isDisabled,
      onClick: onClick
    }, makeAccessible({
      expanded: isItemExpanded,
      controls: collapsibleBodyId
    })), metaAttribute({
      name: MetaConstants.AccordionButton
    })), {}, {
      "data-analytics-name": MAKE_ANALYTICS_CONSTANTS.ACCORDION.ACCORDION_ITEM_BUTTON,
      children: isDeprecatedAPI ? /*#__PURE__*/jsx(AccordionItemHeader, {
        title: title,
        leading: _icon !== null && _icon !== void 0 ? _icon : _index
      }) : header
    }))
  }));
};
var AccordionButton = /*#__PURE__*/assignWithoutSideEffects(_AccordionButton, {
  componentId: MetaConstants.AccordionButton
});

export { AccordionButton };
