// @razorpay/blade/build/lib/web/production/components/Accordion/AccordionItemHeader.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default from 'react';
import { useAccordion, useAccordionItemIndex } from './AccordionContext.js';
import { componentIds } from './componentIds.js';
import { BaseHeader } from '../BaseHeaderFooter/BaseHeader.js';
import '../Typography/index.js';
import '../Box/BaseBox/index.js';
import { CollapsibleChevronIcon } from '../Collapsible/CollapsibleChevronIcon.web.js';
import '../../utils/assignWithoutSideEffects/index.js';
import '../../utils/metaAttribute/index.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import '../../utils/isIconComponent/index.js';
import '../../utils/makeSize/index.js';
import '../../tokens/global/index.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import { makeSize } from '../../utils/makeSize/makeSize.js';
import { size } from '../../tokens/global/size.js';
import { isIconComponent } from '../../utils/isIconComponent/isIconComponent.js';
import { Text } from '../Typography/Text/Text.js';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { assignWithoutSideEffects } from '../../utils/assignWithoutSideEffects/assignWithoutSideEffects.js';

var _excluded = ["title", "subtitle", "leading", "children", "trailing", "titleSuffix"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var getLeadingElementMaxHeightAndWidth = function getLeadingElementMaxHeightAndWidth(size$1) {
  if (size$1 === 'large') return makeSize(size['32']);
  return makeSize(size['24']);
};
var _AccordionItemHeader = function _AccordionItemHeader(_ref) {
  var _leading$type$name;
  var title = _ref.title,
    subtitle = _ref.subtitle,
    leading = _ref.leading,
    children = _ref.children,
    trailing = _ref.trailing,
    titleSuffix = _ref.titleSuffix,
    rest = _objectWithoutProperties(_ref, _excluded);
  var _useAccordion = useAccordion(),
    size = _useAccordion.size,
    showNumberPrefix = _useAccordion.showNumberPrefix,
    expandedIndex = _useAccordion.expandedIndex;
  var _useAccordionItemInde = useAccordionItemIndex(),
    index = _useAccordionItemInde.index,
    isDisabled = _useAccordionItemInde.isDisabled;
  var isLeadingIcon = /*#__PURE__*/React__default.isValidElement(leading) && typeof leading.type === 'function' && ((_leading$type$name = leading.type.name) === null || _leading$type$name === void 0 ? void 0 : _leading$type$name.endsWith('Icon'));
  var isLeadingNumberOrIcon = React__default.useMemo(function () {
    // Check if leading is a number
    if (showNumberPrefix && typeof index === 'number') return true;

    // Check if leading is an Icon component (name ends with "Icon")
    if (leading && /*#__PURE__*/React__default.isValidElement(leading) && isIconComponent(leading)) return true;
    if (isLeadingIcon) return true;
    return false;
  }, [showNumberPrefix, index, leading, isLeadingIcon]);
  var shouldAlignHeaderItemsInCenter = Boolean(children || Boolean(leading) && !isLeadingNumberOrIcon);
  var leadingElement = React__default.useMemo(function () {
    if (showNumberPrefix && typeof index === 'number') {
      return /*#__PURE__*/jsxs(Text, {
        size: size,
        weight: "semibold",
        marginTop: "-2px",
        as: "span",
        children: [index + 1, "."]
      });
    }
    if (leading) {
      return /*#__PURE__*/jsx(BaseBox, {
        marginRight: isLeadingIcon ? 'spacing.0' : 'spacing.3',
        marginTop: "spacing.1"
        // So in design, we have set max height and width for the leading element, doing same here
        ,
        maxHeight: getLeadingElementMaxHeightAndWidth(size),
        maxWidth: getLeadingElementMaxHeightAndWidth(size),
        overflow: "hidden",
        children: leading
      });
    }
    return null;
  }, [showNumberPrefix, index, leading, size, isLeadingIcon]);
  return /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread({}, metaAttribute({
    name: MetaConstants.AccordionItemHeader
  })), {}, {
    flex: "1",
    children: /*#__PURE__*/jsx(BaseHeader, _objectSpread(_objectSpread({
      leading: leadingElement,
      title: title,
      subtitle: subtitle,
      trailing: trailing,
      titleSuffix: titleSuffix,
      isDisabled: isDisabled,
      showBackButton: false,
      showCloseButton: false,
      showDivider: expandedIndex === index,
      paddingX: "spacing.5",
      marginY: "spacing.5",
      size: size,
      alignItems: !subtitle && !isLeadingNumberOrIcon ? 'center' : 'flex-start',
      dividerProps: {
        thickness: 'thinner',
        marginX: 'spacing.5'
      },
      trailingInteractionElement: /*#__PURE__*/jsx(CollapsibleChevronIcon, {
        color: isDisabled ? 'interactive.icon.gray.disabled' : 'interactive.icon.gray.muted',
        size: "large"
      }),
      shouldAlignLeadingAndTrailingElementsToCenter: shouldAlignHeaderItemsInCenter
    }, makeAnalyticsAttribute(rest)), {}, {
      children: children
    }))
  }));
};
var AccordionItemHeader = /*#__PURE__*/assignWithoutSideEffects(_AccordionItemHeader, {
  componentId: componentIds.AccordionItemHeader
});

export { AccordionItemHeader };
