// @razorpay/blade/build/lib/web/production/components/Accordion/AccordionItemBody.js
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { useAccordion } from './AccordionContext.js';
import { componentIds } from './componentIds.js';
import '../Box/BaseBox/index.js';
import '../Typography/index.js';
import '../../utils/index.js';
import '../../utils/makeAccessible/index.js';
import '../../utils/assignWithoutSideEffects/index.js';
import '../../utils/metaAttribute/index.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { makeAccessible } from '../../utils/makeAccessible/makeAccessible.web.js';
import { Text } from '../Typography/Text/Text.js';
import { isReactNative } from '../../utils/platform/isReactNative.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { assignWithoutSideEffects } from '../../utils/assignWithoutSideEffects/assignWithoutSideEffects.js';

var _excluded = ["children", "_description"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var BLANK_SPACE = ' ';

/**
 * On React Native if the `AccordionItem` has a lengthy description which renders a `Text` spanning multiple lines,
 * it sometimes messes up the layout calculation (it thinks of multiline as a single line before flowing in the UI).
 * And during the expanding / collapsing animation, text reflows causing words to jump around across lines.
 *
 * Rendering a blank `Text` at the end seems to fix all this 🤯
 */
var reactNativeMultilineTextOverflowFix =
/*#__PURE__*/
// make this hidden from screen readers
jsx(BaseBox, _objectSpread(_objectSpread({}, makeAccessible({
  hidden: true
})), {}, {
  children: /*#__PURE__*/jsx(Text, {
    children: BLANK_SPACE
  })
}));
var descriptionSizeToken = {
  large: 'medium',
  medium: 'small'
};
var _AccordionItemBody = function _AccordionItemBody(_ref) {
  var children = _ref.children,
    _description = _ref._description,
    rest = _objectWithoutProperties(_ref, _excluded);
  var _useAccordion = useAccordion(),
    size = _useAccordion.size;
  var childrenElement = typeof children === 'string' || typeof children === 'number' ? /*#__PURE__*/jsx(Text, {
    size: descriptionSizeToken[size],
    color: "surface.text.gray.subtle",
    children: children
  }) : children;
  var descriptionElement = _description && /*#__PURE__*/jsx(Text, {
    size: descriptionSizeToken[size],
    color: "surface.text.gray.subtle",
    children: _description
  });
  var collapsibleBodyContent = isReactNative() ? /*#__PURE__*/jsxs(BaseBox, {
    marginX: "spacing.5",
    children: [descriptionElement, /*#__PURE__*/jsx(BaseBox, {
      marginTop: _description && children ? 'spacing.5' : 'spacing.0',
      children: childrenElement
    }), reactNativeMultilineTextOverflowFix]
  }) : /*#__PURE__*/jsxs(BaseBox, {
    display: "flex",
    flexDirection: "column",
    gap: "spacing.5",
    marginBottom: "spacing.5",
    marginX: "spacing.5",
    children: [descriptionElement, childrenElement]
  });
  return /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread(_objectSpread({}, metaAttribute({
    name: MetaConstants.AccordionItemBody
  })), makeAnalyticsAttribute(rest)), {}, {
    children: collapsibleBodyContent
  }));
};
var AccordionItemBody = /*#__PURE__*/assignWithoutSideEffects(_AccordionItemBody, {
  componentId: componentIds.AccordionItemBody
});

export { AccordionItemBody };
