// @razorpay/blade/build/lib/web/production/components/Accordion/AccordionContext.js
import { createContext, useContext } from 'react';
import '../../utils/logger/index.js';
import { throwBladeError } from '../../utils/logger/logger.js';

var AccordionContext = /*#__PURE__*/createContext(null);
var AccordionItemContext = /*#__PURE__*/createContext({
  index: undefined
});
var useAccordion = function useAccordion() {
  var accordionContext = useContext(AccordionContext);
  if (false) {
    if (!accordionContext) {
      throwBladeError({
        message: 'useAccordion should be only used within AccordionContext',
        moduleName: 'AccordionContext'
      });
    }
  }
  return accordionContext;
};
var useAccordionItemIndex = function useAccordionItemIndex() {
  var accordionItemContext = useContext(AccordionItemContext);
  if (false) {
    if (!accordionItemContext) {
      throwBladeError({
        message: 'AccordionItem* components should be only used within AccordionItem',
        moduleName: 'AccordionContext'
      });
    }
  }
  return accordionItemContext;
};

export { AccordionContext, AccordionItemContext, useAccordion, useAccordionItemIndex };
