// @razorpay/blade/build/lib/web/production/components/Accordion/StyledAccordionButton.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import styled from 'styled-components';
import { getCommonAccordionButtonStyles, getBackgroundColor, getTransitionDuration, getTransitionEasing } from './commonStyles.js';
import '../../utils/index.js';
import '../../utils/getFocusRingStyles/index.js';
import { castWebType } from '../../utils/platform/castUtils.js';
import { getFocusRingStyles } from '../../utils/getFocusRingStyles/getFocusRingStyles.web.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StyledAccordionButton = /*#__PURE__*/styled.button.withConfig({
  displayName: "StyledAccordionButtonweb__StyledAccordionButton",
  componentId: "sc-1qxmqda-0"
})(function (props) {
  var theme = props.theme,
    isExpanded = props.isExpanded,
    disabled = props.disabled;
  var commonStyles = getCommonAccordionButtonStyles(props);
  return _objectSpread(_objectSpread({}, commonStyles), {}, {
    backgroundColor: getBackgroundColor({
      theme: theme,
      isActive: false
    }),
    transitionProperty: 'background-color, box-shadow, border-radius, color',
    transitionDuration: castWebType(getTransitionDuration(theme)),
    transitionTimingFunction: castWebType(getTransitionEasing(theme)),
    cursor: disabled ? 'not-allowed' : 'pointer',
    color: theme.colors.interactive.icon.gray[isExpanded ? 'subtle' : 'muted'],
    width: '100%',
    border: 'none',
    textAlign: 'left',
    '&:hover, &:focus-visible': {
      backgroundColor: getBackgroundColor({
        theme: theme,
        isActive: true
      }),
      color: theme.colors.interactive.icon.gray.subtle,
      '& [data-blade-component="divider"]': {
        opacity: 0
      }
    },
    '&:focus-visible': _objectSpread(_objectSpread({}, getFocusRingStyles({
      theme: theme
    })), {}, {
      // only need border radius on the focus ring
      borderRadius: theme.border.radius.small
    })
  });
});

export { StyledAccordionButton };
