// @razorpay/blade/build/lib/web/development/components/Button/BaseButton/StyledBaseButton.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import styled from 'styled-components';
import getBaseButtonStyles from './getStyledBaseButtonStyles.js';
import getIn from '../../../utils/lodashButBetter/get.js';
import '../../Box/styledProps/index.js';
import '../../../utils/makeMotionTime/index.web.js';
import { omitPropsFromHTML } from '../../../utils/omitPropsFromHTML/index.js';
import '../../../utils/index.js';
import { useStyledProps } from '../../Box/styledProps/useStyledProps.js';
import { castWebType } from '../../../utils/platform/castUtils.js';
import { makeMotionTime } from '../../../utils/makeMotionTime/makeMotionTime.web.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StyledBaseButton = /*#__PURE__*/styled.button.withConfig({
  shouldForwardProp: omitPropsFromHTML,
  displayName: 'StyledBaseButton'
}).attrs(function (props) {
  return _objectSpread({}, props.accessibilityProps);
}).withConfig({
  displayName: "StyledBaseButtonweb__StyledBaseButton",
  componentId: "sc-26bt38-0"
})(function (props) {
  var styledPropsCSSObject = useStyledProps(props);
  return _objectSpread(_objectSpread({}, getBaseButtonStyles(props)), {}, {
    display: 'inline-flex',
    transitionProperty: 'background-color, border-color, box-shadow',
    transitionTimingFunction: getIn(props.theme.motion, props.motionEasing),
    transitionDuration: castWebType(makeMotionTime(getIn(props.theme.motion, props.motionDuration))),
    position: 'relative',
    '&:hover': _objectSpread(_objectSpread({
      backgroundColor: props.hoverBackgroundColor
    }, props.variant !== 'tertiary' && {
      borderColor: props.hoverBorderColor
    }), props.variant === 'tertiary' && props.color === 'transparent' && {
      '&& [data-blade-component="svg-path"]': {
        fill: props.hoverIconColor
      }
    }),
    '&:active': _objectSpread(_objectSpread({
      backgroundColor: props.focusBackgroundColor
    }, props.variant !== 'tertiary' && {
      borderColor: props.focusBorderColor
    }), props.variant === 'tertiary' && props.color === 'transparent' && {
      '&& [data-blade-component="svg-path"]': {
        fill: props.hoverIconColor
      }
    }),
    '&:focus-visible': _objectSpread(_objectSpread({
      backgroundColor: props.focusBackgroundColor
    }, props.variant !== 'tertiary' && {
      borderColor: props.focusBorderColor
    }), {}, {
      outline: "1px solid ".concat(props.theme.colors.surface.background.primary.subtle),
      boxShadow: "0px 0px 0px 4px ".concat(props.focusRingColor)
    }, props.variant === 'tertiary' && props.color === 'transparent' && {
      '&& [data-blade-component="svg-path"]': {
        fill: props.hoverIconColor
      }
    }),
    '*': {
      transitionProperty: 'color, fill, opacity',
      transitionDuration: castWebType(makeMotionTime(getIn(props.theme.motion, props.motionDuration))),
      transitionTimingFunction: getIn(props.theme.motion, props.motionEasing)
    }
  }, styledPropsCSSObject);
});

export { StyledBaseButton as default };
