// @razorpay/blade/build/lib/web/production/components/Form/Selector/SelectorGroupField.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import 'react';
import '../../../utils/metaAttribute/index.js';
import '../../Box/BaseBox/index.js';
import '../../../utils/makeAccessible/index.js';
import '../../../utils/index.js';
import '../../../utils/makeAnalyticsAttribute/index.js';
import '../../BladeProvider/index.js';
import { jsx } from 'react/jsx-runtime';
import useTheme from '../../BladeProvider/useTheme.js';
import { useBreakpoint } from '../../../utils/useBreakpoint/useBreakpoint.js';
import { getPlatformType } from '../../../utils/getPlatformType/getPlatformType.js';
import { BaseBox } from '../../Box/BaseBox/BaseBox.web.js';
import { makeAccessible } from '../../../utils/makeAccessible/makeAccessible.web.js';
import { metaAttribute } from '../../../utils/metaAttribute/metaAttribute.web.js';
import { makeAnalyticsAttribute } from '../../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';

var _excluded = ["children", "labelledBy", "position", "accessibilityRole", "componentName", "testID"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var SelectorGroupField = function SelectorGroupField(_ref) {
  var children = _ref.children,
    labelledBy = _ref.labelledBy,
    position = _ref.position,
    _ref$accessibilityRol = _ref.accessibilityRole,
    accessibilityRole = _ref$accessibilityRol === void 0 ? 'group' : _ref$accessibilityRol,
    componentName = _ref.componentName,
    testID = _ref.testID,
    props = _objectWithoutProperties(_ref, _excluded);
  var _useTheme = useTheme(),
    theme = _useTheme.theme;
  var _useBreakpoint = useBreakpoint({
      breakpoints: theme.breakpoints
    }),
    matchedDeviceType = _useBreakpoint.matchedDeviceType;
  var isReactNative = getPlatformType() === 'react-native';
  var isLabelLeftPositioned = position === 'left' && matchedDeviceType === 'desktop';
  if (isReactNative) isLabelLeftPositioned = false;
  var role = accessibilityRole === 'group' && isReactNative ? undefined : accessibilityRole;
  return /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread(_objectSpread(_objectSpread({
    display: "flex",
    flexDirection: isLabelLeftPositioned ? 'row' : 'column'
  }, makeAccessible({
    role: role,
    labelledBy: labelledBy
  })), {}, {
    alignItems: componentName === 'chip-group' && isLabelLeftPositioned ? 'baseline' : undefined
  }, metaAttribute({
    name: componentName,
    testID: testID
  })), makeAnalyticsAttribute(props)), {}, {
    children: children
  }));
};

export { SelectorGroupField };
