// @razorpay/blade/build/lib/web/production/components/BottomSheet/BottomSheetBackdrop.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import 'react';
import styled from 'styled-components';
import { useBottomSheetContext } from './BottomSheetContext.js';
import '../Box/BaseBox/index.js';
import '../../utils/index.js';
import '../../utils/metaAttribute/index.js';
import '../../utils/makeMotionTime/index.web.js';
import { jsx } from 'react/jsx-runtime';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { makeMotionTime } from '../../utils/makeMotionTime/makeMotionTime.web.js';
import { castWebType } from '../../utils/platform/castUtils.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StyledBottomSheetBackdrop = /*#__PURE__*/styled(BaseBox).withConfig({
  displayName: "BottomSheetBackdropweb__StyledBottomSheetBackdrop",
  componentId: "sc-198k9f6-0"
})(function (_ref) {
  var theme = _ref.theme,
    isOpen = _ref.isOpen;
  return {
    transitionDuration: "".concat(makeMotionTime(theme.motion.duration.moderate)),
    transitionTimingFunction: isOpen ? castWebType(theme.motion.easing.entrance) : castWebType(theme.motion.easing.exit),
    pointerEvents: isOpen ? 'all' : 'none',
    transitionProperty: 'opacity'
  };
});
var BottomSheetBackdrop = function BottomSheetBackdrop(_ref2) {
  var zIndex = _ref2.zIndex;
  var _useBottomSheetContex = useBottomSheetContext(),
    close = _useBottomSheetContex.close,
    isOpen = _useBottomSheetContex.isOpen,
    isDismissible = _useBottomSheetContex.isDismissible;
  return /*#__PURE__*/jsx(StyledBottomSheetBackdrop, _objectSpread(_objectSpread({}, metaAttribute({
    testID: 'bottomsheet-backdrop'
  })), {}, {
    onClick: function onClick() {
      if (isDismissible) {
        close === null || close === void 0 || close();
      }
    },
    isOpen: isOpen,
    opacity: isOpen ? 1 : 0,
    position: "fixed",
    left: "spacing.0",
    top: "spacing.0",
    bottom: "spacing.0",
    right: "spacing.0",
    zIndex: zIndex,
    backgroundColor: "overlay.background.subtle"
  }));
};

export { BottomSheetBackdrop };
