// @razorpay/blade/build/lib/web/production/components/QuickFilters/QuickFilter.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import { forwardRef } from 'react';
import './QuickFilterGroup/index.js';
import '../Card/index.js';
import '../Box/index.js';
import '../Typography/index.js';
import '../Radio/index.js';
import '../Checkbox/index.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import '../../utils/metaAttribute/index.js';
import '../Box/BaseBox/index.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { Box } from '../Box/Box.js';
import { Radio } from '../Radio/Radio.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { Checkbox } from '../Checkbox/Checkbox.js';
import { Text } from '../Typography/Text/Text.js';
import { useQuickFilterGroupContext } from './QuickFilterGroup/QuickFilterContext.js';
import { Card, CardBody } from '../Card/Card.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';

var _excluded = ["value", "title", "trailing", "selectionType", "isSelected"],
  _excluded2 = ["title", "value", "trailing", "testID"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var QuickFilterContent = function QuickFilterContent(_ref) {
  var value = _ref.value,
    title = _ref.title,
    trailing = _ref.trailing,
    _ref$selectionType = _ref.selectionType,
    selectionType = _ref$selectionType === void 0 ? 'single' : _ref$selectionType,
    _ref$isSelected = _ref.isSelected,
    isSelected = _ref$isSelected === void 0 ? false : _ref$isSelected,
    rest = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/jsxs(BaseBox, {
    display: "flex",
    flexDirection: "row",
    gap: "spacing.3",
    width: "fit-content",
    justifyContent: "center",
    alignItems: "center",
    paddingY: "spacing.3",
    paddingLeft: "spacing.3",
    paddingRight: "spacing.4",
    children: [/*#__PURE__*/jsxs(Box, {
      display: "flex",
      flexDirection: "row",
      gap: "spacing.2",
      justifyContent: "center",
      alignItems: "center",
      children: [selectionType === 'single' ? /*#__PURE__*/jsx(Radio, _objectSpread({
        value: value,
        _hideRadioIcon: true
      }, makeAnalyticsAttribute(rest))) : /*#__PURE__*/jsx(Checkbox, _objectSpread({
        value: value
      }, makeAnalyticsAttribute(rest))), /*#__PURE__*/jsx(Text, {
        variant: "body",
        size: "medium",
        color: isSelected ? 'interactive.text.primary.subtle' : 'interactive.text.gray.subtle',
        weight: "medium",
        children: title
      })]
    }), trailing]
  });
};
var QuickFilter = /*#__PURE__*/forwardRef(function (_ref2, ref) {
  var title = _ref2.title,
    value = _ref2.value,
    trailing = _ref2.trailing,
    testID = _ref2.testID,
    rest = _objectWithoutProperties(_ref2, _excluded2);
  var _useQuickFilterGroupC = useQuickFilterGroupContext(),
    selectedQuickFilters = _useQuickFilterGroupC.selectedQuickFilters,
    selectionType = _useQuickFilterGroupC.selectionType;
  var isQuickFilterSelected = selectedQuickFilters.includes(value);
  return /*#__PURE__*/jsx(Card, _objectSpread(_objectSpread(_objectSpread({
    padding: "spacing.0",
    as: "label",
    accessibilityLabel: title,
    borderRadius: "medium",
    elevation: "none",
    isSelected: isQuickFilterSelected,
    ref: ref
    // Add margin to accommodate the box shadow that appears when selected, preventing visual cutoff
    ,
    marginRight: "spacing.1"
  }, makeAnalyticsAttribute(rest)), metaAttribute({
    testID: testID
  })), {}, {
    children: /*#__PURE__*/jsx(CardBody, {
      children: /*#__PURE__*/jsx(QuickFilterContent, _objectSpread({
        value: value,
        title: title,
        trailing: trailing,
        selectionType: selectionType,
        isSelected: isQuickFilterSelected
      }, rest))
    })
  }));
});

export { QuickFilter };
