// @razorpay/blade/build/lib/web/development/components/Form/Selector/SelectorLabel.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import 'react';
import styled from 'styled-components';
import '../../../utils/metaAttribute/index.js';
import '../../../utils/makeSpace/index.js';
import { jsx } from 'react/jsx-runtime';
import { makeSpace } from '../../../utils/makeSpace/makeSpace.js';
import { metaAttribute } from '../../../utils/metaAttribute/metaAttribute.web.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StyledSelectorLabel = /*#__PURE__*/styled.label.withConfig({
  displayName: "SelectorLabelweb__StyledSelectorLabel",
  componentId: "sc-1yoi09a-0"
})(function (_ref) {
  var theme = _ref.theme;
  return {
    display: 'flex',
    marginTop: makeSpace(theme.spacing[1]),
    marginBottom: makeSpace(theme.spacing[1])
  };
});
var SelectorLabel = function SelectorLabel(_ref2) {
  var children = _ref2.children,
    componentName = _ref2.componentName,
    testID = _ref2.testID,
    onMouseDown = _ref2.onMouseDown,
    onMouseUp = _ref2.onMouseUp,
    onMouseOut = _ref2.onMouseOut,
    onKeyDown = _ref2.onKeyDown,
    onKeyUp = _ref2.onKeyUp,
    onTouchStart = _ref2.onTouchStart,
    onTouchEnd = _ref2.onTouchEnd,
    style = _ref2.style;
  return /*#__PURE__*/jsx(StyledSelectorLabel, _objectSpread(_objectSpread({
    onTouchStart: onTouchStart,
    onTouchEnd: onTouchEnd,
    onMouseDown: onMouseDown,
    onMouseUp: onMouseUp,
    onMouseOut: onMouseOut,
    onKeyDown: onKeyDown,
    onKeyUp: onKeyUp,
    style: style
  }, metaAttribute({
    name: componentName,
    testID: testID
  })), {}, {
    children: children
  }));
};

export { SelectorLabel };
