// @razorpay/blade/build/lib/web/development/components/Form/Selector/SelectorSupportText.js
import '../../Box/index.js';
import '../../Typography/Text/index.js';
import '../../../utils/index.js';
import { jsx } from 'react/jsx-runtime';
import { getPlatformType } from '../../../utils/getPlatformType/getPlatformType.js';
import { Box } from '../../Box/Box.js';
import { castWebType } from '../../../utils/platform/castUtils.js';
import { Text } from '../../Typography/Text/Text.js';

var SelectorSupportText = function SelectorSupportText(_ref) {
  var children = _ref.children,
    id = _ref.id,
    isNegative = _ref.isNegative,
    size = _ref.size;
  var isReactNative = getPlatformType() === 'react-native';
  var sizeMapping = {
    small: 'small',
    medium: 'small',
    large: 'medium'
  };
  return /*#__PURE__*/jsx(Box, {
    id: id,
    display: isReactNative ? undefined : castWebType('contents'),
    children: /*#__PURE__*/jsx(Text, {
      variant: "caption",
      size: sizeMapping[size],
      as: isReactNative ? undefined : 'span',
      color: isNegative ? 'feedback.text.negative.intense' : 'surface.text.gray.muted',
      children: children
    })
  });
};

export { SelectorSupportText };
