// @razorpay/blade/build/lib/web/development/components/BaseMenu/BaseMenuItem/tokens.js
import '../../../tokens/global/index.js';
import { size } from '../../../tokens/global/size.js';

/* eslint-disable @typescript-eslint/restrict-plus-operands */

var itemFirstRowHeight = size[20];
var actionListItemPadding = 8;
var dividerYMargin = 1;
var actionListSectionTitleHeight = size[18] + actionListItemPadding * 2;
var actionListDividerHeight = size[1] + dividerYMargin * 2;
var getItemPadding = function getItemPadding(theme) {
  return {
    itemPaddingMobile: theme.spacing[2],
    itemPaddingDesktop: theme.spacing[3]
  };
};
var getItemMargin = function getItemMargin(theme) {
  return theme.spacing[1];
};
var getActionListItemHeight = function getActionListItemHeight(theme) {
  return {
    itemHeightMobile:
    // eslint-disable-next-line @typescript-eslint/restrict-plus-operands
    itemFirstRowHeight + getItemPadding(theme).itemPaddingMobile * 2 + getItemMargin(theme) * 2,
    itemHeightDesktop:
    // eslint-disable-next-line @typescript-eslint/restrict-plus-operands
    itemFirstRowHeight + getItemPadding(theme).itemPaddingDesktop * 2 + getItemMargin(theme) * 2
  };
};

export { actionListDividerHeight, actionListSectionTitleHeight, getActionListItemHeight, getItemMargin, getItemPadding, itemFirstRowHeight };
