// @razorpay/blade/build/lib/web/development/components/ChatMessage/DefaultMessageBubble.web.js
import 'react';
import Rotate from './Rotate.web.js';
import '../Box/BaseBox/index.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';

var DefaultMessageBubble = function DefaultMessageBubble(_ref) {
  var children = _ref.children,
    leading = _ref.leading,
    isLoading = _ref.isLoading,
    footerActions = _ref.footerActions,
    isChildText = _ref.isChildText;
  return /*#__PURE__*/jsx(BaseBox, {
    children: /*#__PURE__*/jsxs(BaseBox, {
      display: "grid",
      gridTemplateColumns: "auto 1fr",
      gridTemplateRows: "auto auto",
      columnGap: "spacing.4",
      children: [/*#__PURE__*/jsx(BaseBox, {
        padding: "spacing.2",
        children: /*#__PURE__*/jsx(Rotate, {
          animate: isLoading,
          children: leading
        })
      }), /*#__PURE__*/jsx(BaseBox, {
        display: "flex",
        alignItems: "center",
        paddingY: isChildText ? 'spacing.2' : 'spacing.0',
        children: children
      }), /*#__PURE__*/jsx(BaseBox, {
        gridColumn: "2",
        children: footerActions
      })]
    })
  });
};

export { DefaultMessageBubble };
