// @razorpay/blade/build/lib/web/development/components/ChatMessage/Rotate.web.js
import 'react';
import { m } from 'framer-motion';
import '../../utils/index.js';
import '../BladeProvider/index.js';
import { msToSeconds } from '../../utils/msToSeconds.js';
import { cssBezierToArray } from '../../utils/cssBezierToArray.js';
import { jsx } from 'react/jsx-runtime';
import useTheme from '../BladeProvider/useTheme.js';
import { castWebType } from '../../utils/platform/castUtils.js';

var Rotate = function Rotate(_ref) {
  var children = _ref.children,
    animate = _ref.animate;
  var _useTheme = useTheme(),
    theme = _useTheme.theme;
  if (!animate) {
    return children;
  }
  return /*#__PURE__*/jsx(m.div, {
    style: {
      display: 'flex'
    },
    animate: {
      rotate: 90
    },
    transition: {
      duration: msToSeconds(theme.motion.duration.gentle),
      repeat: Infinity,
      ease: cssBezierToArray(castWebType(theme.motion.easing.emphasized)),
      delay: msToSeconds(theme.motion.delay.gentle)
    },
    children: children
  });
};

export { Rotate as default };
