// @razorpay/blade/build/lib/web/production/components/Collapsible/CollapsibleBody.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import { CollapsibleBodyContent } from './CollapsibleBodyContent.web.js';
import { useCollapsible } from './CollapsibleContext.js';
import { componentIds } from './componentIds.js';
import '../Box/BaseBox/index.js';
import '../../utils/metaAttribute/index.js';
import '../../utils/makeAccessible/index.js';
import '../../utils/assignWithoutSideEffects/index.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import { jsx } from 'react/jsx-runtime';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { makeAccessible } from '../../utils/makeAccessible/makeAccessible.web.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { assignWithoutSideEffects } from '../../utils/assignWithoutSideEffects/assignWithoutSideEffects.js';

var _excluded = ["children", "testID", "width", "_hasMargin"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _CollapsibleBody = function _CollapsibleBody(_ref) {
  var children = _ref.children,
    testID = _ref.testID,
    width = _ref.width,
    _ref$_hasMargin = _ref._hasMargin,
    _hasMargin = _ref$_hasMargin === void 0 ? true : _ref$_hasMargin,
    rest = _objectWithoutProperties(_ref, _excluded);
  var _useCollapsible = useCollapsible(),
    collapsibleBodyId = _useCollapsible.collapsibleBodyId,
    isExpanded = _useCollapsible.isExpanded;
  return /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread(_objectSpread(_objectSpread({
    id: collapsibleBodyId,
    width: width
  }, makeAccessible({
    role: 'region',
    hidden: !isExpanded
  })), metaAttribute({
    name: MetaConstants.CollapsibleBody,
    testID: testID
  })), makeAnalyticsAttribute(rest)), {}, {
    children: /*#__PURE__*/jsx(CollapsibleBodyContent, {
      _hasMargin: _hasMargin,
      children: children
    })
  }));
};
var CollapsibleBody = /*#__PURE__*/assignWithoutSideEffects(_CollapsibleBody, {
  componentId: componentIds.CollapsibleBody
});

export { CollapsibleBody };
