// @razorpay/blade/build/lib/web/production/components/Tag/AnimatedTag.web.js
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import React__default from 'react';
import styled, { keyframes, css } from 'styled-components';
import { Tag } from './Tag.js';
import { TAG_OPACITY_START, TAG_MAX_WIDTH_START, TAG_OPACITY_END, TAG_MAX_WIDTH_END } from './tagAnimationConfig.js';
import '../Box/BaseBox/index.js';
import '../BladeProvider/index.js';
import '../../utils/index.js';
import { jsx } from 'react/jsx-runtime';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import useTheme from '../BladeProvider/useTheme.js';
import { makeMotionTime } from '../../utils/makeMotionTime/makeMotionTime.web.js';

var tagDissappearKeyframe = /*#__PURE__*/keyframes(["0%{opacity:", ";max-width:", "}100%{opacity:", ";max-width:", ";}"], TAG_OPACITY_START, TAG_MAX_WIDTH_START, TAG_OPACITY_END, TAG_MAX_WIDTH_END);
var tagShowKeyframe = /*#__PURE__*/keyframes(["0%{opacity:", ";}100%{opacity:", ";}"], TAG_OPACITY_END, TAG_OPACITY_START);
var AnimatedTagContainer = /*#__PURE__*/styled(BaseBox).withConfig({
  displayName: "AnimatedTagweb__AnimatedTagContainer",
  componentId: "sc-1ypkpkt-0"
})(function (props) {
  return css(["", ";display:inline-block;opacity:", ";max-width:", ";flex-shrink:0;"], props.animationProps, props.isVisible ? TAG_OPACITY_START : TAG_OPACITY_END, props.isVisible ? TAG_MAX_WIDTH_START : TAG_MAX_WIDTH_END);
});
var AnimatedTag = function AnimatedTag(_ref) {
  var children = _ref.children,
    currentTagIndex = _ref.currentTagIndex,
    isDisabled = _ref.isDisabled,
    activeTagIndex = _ref.activeTagIndex,
    onDismiss = _ref.onDismiss,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'medium' : _ref$size;
  var _React$useState = React__default.useState(true),
    _React$useState2 = _slicedToArray(_React$useState, 2),
    isTagVisible = _React$useState2[0],
    setIsTagVisible = _React$useState2[1];
  var _useTheme = useTheme(),
    theme = _useTheme.theme;
  var hideTagAnimationProps = /*#__PURE__*/css(["animation:", " ", " ", ";"], tagDissappearKeyframe, makeMotionTime(theme.motion.duration.xquick), String(theme.motion.easing.exit));
  var showTagAnimationProps = /*#__PURE__*/css(["animation:", " ", " ", ";"], tagShowKeyframe, makeMotionTime(theme.motion.duration.xquick), String(theme.motion.easing.entrance));
  return /*#__PURE__*/jsx(AnimatedTagContainer, {
    isVisible: isTagVisible,
    onAnimationEnd: function onAnimationEnd() {
      if (!isTagVisible) {
        onDismiss({
          tagIndex: currentTagIndex,
          tagName: children
        });
      }
    },
    animationProps: isTagVisible ? showTagAnimationProps : hideTagAnimationProps,
    children: /*#__PURE__*/jsx(Tag, {
      _isVirtuallyFocused: currentTagIndex === activeTagIndex,
      _isTagInsideInput: true,
      onDismiss: function onDismiss() {
        setIsTagVisible(false);
      },
      marginRight: "spacing.3",
      marginY: "spacing.2",
      isDisabled: isDisabled,
      size: size,
      children: children
    })
  });
};

export { AnimatedTag };
