// @razorpay/blade/build/lib/web/production/components/Tag/Tag.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default from 'react';
import styled from 'styled-components';
import '../Box/index.js';
import '../Box/styledProps/index.js';
import '../Button/IconButton/index.js';
import '../Icons/index.js';
import '../Typography/index.js';
import '../../utils/metaAttribute/index.js';
import '../../utils/index.js';
import '../../tokens/global/index.js';
import '../Box/BaseBox/index.js';
import { useIsMobile } from '../../utils/useIsMobile.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { isReactNative } from '../../utils/platform/isReactNative.js';
import { makeSize } from '../../utils/makeSize/makeSize.js';
import { size } from '../../tokens/global/size.js';
import { getStyledProps } from '../Box/styledProps/getStyledProps.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { Box } from '../Box/Box.js';
import { Text } from '../Typography/Text/Text.js';
import { IconButton } from '../Button/IconButton/IconButton.js';
import CloseIcon from '../Icons/CloseIcon/CloseIcon.js';

var _excluded = ["size", "icon", "onDismiss", "children", "isDisabled", "testID", "_isVirtuallyFocused", "_isTagInsideInput"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var FocussableTag = /*#__PURE__*/styled(BaseBox).withConfig({
  displayName: "Tag__FocussableTag",
  componentId: "jmwb4r-0"
})(function (props) {
  if (props._isVirtuallyFocused && !isReactNative()) {
    return {
      outline: "".concat(makeSize(size['1']), " solid ").concat(props.theme.colors.surface.background.gray.subtle),
      boxShadow: "0px 0px 0px 4px ".concat(props.theme.colors.interactive.background.primary.faded)
    };
  }
  return {};
});
var _Tag = function _Tag(_ref, ref) {
  var _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'medium' : _ref$size,
    Icon = _ref.icon,
    onDismiss = _ref.onDismiss,
    children = _ref.children,
    isDisabled = _ref.isDisabled,
    testID = _ref.testID,
    _isVirtuallyFocused = _ref._isVirtuallyFocused,
    _isTagInsideInput = _ref._isTagInsideInput,
    rest = _objectWithoutProperties(_ref, _excluded);
  var isMobile = useIsMobile();
  var textColor = isDisabled ? 'interactive.text.gray.disabled' : 'interactive.text.gray.subtle';
  var iconColor = isDisabled ? 'interactive.icon.gray.disabled' : 'interactive.icon.gray.muted';
  var backgroundColor = isDisabled ? 'interactive.background.gray.disabled' : 'interactive.background.gray.default';
  var mediumPadding = {
    base: ['spacing.2', 'spacing.3', 'spacing.2', 'spacing.4'],
    m: ['spacing.1', 'spacing.2', 'spacing.1', 'spacing.3']
  };
  var largePadding = {
    base: ['spacing.2', 'spacing.3', 'spacing.2', 'spacing.4'],
    m: ['spacing.2', 'spacing.3', 'spacing.2', 'spacing.4']
  };
  var assetSize = React__default.useMemo(function () {
    if (isMobile && size === 'large') {
      return 'medium';
    }
    return 'small';
  }, [isMobile, size]);
  return /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread(_objectSpread(_objectSpread({
    ref: ref,
    display: isReactNative() ? 'flex' : 'inline-flex'
  }, getStyledProps(rest)), metaAttribute({
    name: MetaConstants.Tag,
    testID: testID
  })), makeAnalyticsAttribute(rest)), {}, {
    children: /*#__PURE__*/jsxs(FocussableTag, {
      display: isReactNative() ? 'flex' : 'inline-flex',
      alignSelf: isReactNative() ? 'center' : undefined,
      flexDirection: "row",
      flexWrap: "nowrap",
      flexShrink: 0,
      backgroundColor: backgroundColor,
      borderRadius: "max",
      padding: size === 'medium' ? mediumPadding : largePadding,
      _isVirtuallyFocused: _isVirtuallyFocused,
      children: [Icon ? /*#__PURE__*/jsx(Box, {
        display: "flex",
        flexDirection: "row",
        alignItems: "center",
        children: /*#__PURE__*/jsx(Icon, {
          color: iconColor,
          size: assetSize,
          marginRight: "spacing.2"
        })
      }) : null, /*#__PURE__*/jsx(Box, {
        display: "flex",
        flexDirection: "row",
        children: /*#__PURE__*/jsx(Text, {
          textAlign: "center",
          truncateAfterLines: 1,
          marginRight: "spacing.2",
          color: textColor,
          size: "small",
          children: children
        })
      }), /*#__PURE__*/jsx(Box, {
        display: "flex",
        flexDirection: "row",
        alignItems: "center",
        justifyContent: "center",
        children: /*#__PURE__*/jsx(IconButton, {
          size: assetSize,
          icon: CloseIcon,
          accessibilityLabel: "Close ".concat(children, " tag"),
          isDisabled: isDisabled,
          _tabIndex: _isTagInsideInput ? -1 : undefined,
          onClick: function onClick(e) {
            // Inside tag input, we stop propagation to avoid opening dropdown on click of close on tag
            if (_isTagInsideInput) {
              e.stopPropagation();
            }
            onDismiss();
          }
        })
      })]
    })
  }));
};

/**
 * ## Tags
 *
 * Tag component can be used to display selected items on UI.
 *
 * ### Usage
 *
 * ***Note:*** _Make sure to handle state when using Tag_
 *
 * ```jsx
 * const [showTag, setShowTag] = React.useState(true);
 *
 * // ...
 *
 * {showTag && (
 *  <Tag
 *    icon={CheckIcon}
 *    onDismiss={() => setShowTag(false)}
 *  >
 *    Transactions
 *  </Tag>
 * )}
 * ```
 *
 * Checkout [Tags Documentation](https://blade.razorpay.com/?path=/story/components-tag--default) for more info.
 *
 */
var Tag = /*#__PURE__*/React__default.forwardRef(_Tag);

export { Tag };
