// @razorpay/blade/build/lib/web/development/components/Chip/AnimatedChip.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import styled from 'styled-components';
import { getAnimatedChipStyles } from './getAnimatedChipStyles.js';
import { chipMotionTokens } from './chipTokens.js';
import getIn from '../../utils/lodashButBetter/get.js';
import '../Box/BaseBox/index.js';
import '../../utils/makeMotionTime/index.web.js';
import '../../utils/index.js';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { castWebType } from '../../utils/platform/castUtils.js';
import { makeMotionTime } from '../../utils/makeMotionTime/makeMotionTime.web.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var AnimatedChip = /*#__PURE__*/styled(BaseBox).withConfig({
  displayName: "AnimatedChipweb__AnimatedChip",
  componentId: "sc-1xiynj1-0"
})(function (props) {
  var easing = getIn(props.theme.motion, chipMotionTokens.easing);
  var duration = castWebType(makeMotionTime(getIn(props.theme.motion, chipMotionTokens.duration)));
  return _objectSpread(_objectSpread({}, getAnimatedChipStyles(props)), {}, {
    width: '100%',
    transform: "scale(".concat(props.isPressed ? '0.92' : '1', ")"),
    transitionDuration: duration,
    transitionTimingFunction: easing
  });
});

export { AnimatedChip };
