// @razorpay/blade/build/lib/web/development/components/Chip/StyledChipWrapper.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import styled from 'styled-components';
import { chipMotionTokens, chipColorTokens } from './chipTokens.js';
import getIn from '../../utils/lodashButBetter/get.js';
import '../Box/BaseBox/index.js';
import '../../utils/makeMotionTime/index.web.js';
import '../../utils/index.js';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { castWebType } from '../../utils/platform/castUtils.js';
import { makeMotionTime } from '../../utils/makeMotionTime/makeMotionTime.web.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StyledChipWrapper = /*#__PURE__*/styled(BaseBox).withConfig({
  displayName: "StyledChipWrapperweb__StyledChipWrapper",
  componentId: "sc-1p3u1xx-0"
})(function (_ref) {
  var theme = _ref.theme,
    borderColor = _ref.borderColor,
    isChecked = _ref.isChecked,
    isDisabled = _ref.isDisabled,
    color = _ref.color;
  var easing = getIn(theme.motion, chipMotionTokens.easing);
  var duration = castWebType(makeMotionTime(getIn(theme.motion, chipMotionTokens.duration)));
  return _objectSpread({
    display: 'flex',
    borderColor: isChecked ? getIn(theme.colors, borderColor) : 'transparent'
  }, !isDisabled && {
    '&:hover': {
      backgroundColor: getIn(theme.colors, chipColorTokens.background[isChecked && color ? color : 'unchecked'].hover),
      transitionTimingFunction: easing,
      transitionDuration: duration
    }
  });
});

export { StyledChipWrapper };
