// @razorpay/blade/build/lib/web/production/components/Menu/MenuOverlay.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default from 'react';
import styled from 'styled-components';
import { MENU_MIN_WIDTH, overlayPaddingX, overlayPaddingY } from './tokens.js';
import '../Box/BaseBox/index.js';
import { componentZIndices } from '../../utils/componentZIndices.js';
import '../../utils/metaAttribute/index.js';
import { jsx } from 'react/jsx-runtime';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';

var _excluded = ["children", "zIndex", "_transitionStyle", "minWidth", "maxWidth", "width", "testID"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var UnfocussableOverlay = /*#__PURE__*/styled(BaseBox).withConfig({
  displayName: "MenuOverlayweb__UnfocussableOverlay",
  componentId: "sc-1nhzg4x-0"
})(function (_props) {
  return {
    '&:focus-visible': {
      outline: 'none'
    }
  };
});
var _MenuOverlay = function _MenuOverlay(_ref, ref) {
  var children = _ref.children,
    _ref$zIndex = _ref.zIndex,
    zIndex = _ref$zIndex === void 0 ? componentZIndices.dropdownOverlay : _ref$zIndex,
    _transitionStyle = _ref._transitionStyle,
    minWidth = _ref.minWidth,
    maxWidth = _ref.maxWidth,
    width = _ref.width,
    testID = _ref.testID,
    props = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/jsx(UnfocussableOverlay, _objectSpread(_objectSpread(_objectSpread({
    ref: ref
  }, props), {}, {
    zIndex: zIndex
  }, metaAttribute({
    name: MetaConstants.Menu,
    testID: testID
  })), {}, {
    minWidth: minWidth !== null && minWidth !== void 0 ? minWidth : MENU_MIN_WIDTH,
    width: width,
    maxWidth: maxWidth,
    children: /*#__PURE__*/jsx(BaseBox, {
      backgroundColor: "popup.background.subtle",
      paddingX: overlayPaddingX,
      paddingY: overlayPaddingY,
      elevation: "midRaised",
      borderWidth: "thin",
      borderColor: "surface.border.gray.muted",
      borderRadius: "medium",
      style: _transitionStyle,
      children: children
    })
  }));
};
var MenuOverlay = /*#__PURE__*/React__default.forwardRef(_MenuOverlay);

export { MenuOverlay };
