// @razorpay/blade/build/lib/web/production/components/QuickFilters/QuickFilterGroup/QuickFilterGroup.js
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import 'react';
import { useQuickFilterGroupContext, QuickFilterGroupProvider } from './QuickFilterContext.js';
import '../../Radio/index.js';
import '../../Box/BaseBox/index.js';
import '../../Checkbox/index.js';
import { useControllableState } from '../../../utils/useControllable.js';
import { useId } from '../../../utils/useId.js';
import { jsx } from 'react/jsx-runtime';
import { RadioGroup } from '../../Radio/RadioGroup/RadioGroup.js';
import { CheckboxGroup } from '../../Checkbox/CheckboxGroup/CheckboxGroup.js';
import { BaseBox } from '../../Box/BaseBox/BaseBox.web.js';

var _excluded = ["children", "onChange", "setSelectedQuickFilters"],
  _excluded2 = ["children", "testID", "value", "defaultValue", "onChange", "name", "selectionType"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var QuickFilterWrapper = function QuickFilterWrapper(_ref) {
  var children = _ref.children,
    onChange = _ref.onChange,
    setSelectedQuickFilters = _ref.setSelectedQuickFilters,
    rest = _objectWithoutProperties(_ref, _excluded);
  var _useQuickFilterGroupC = useQuickFilterGroupContext(),
    selectedQuickFilters = _useQuickFilterGroupC.selectedQuickFilters,
    selectionType = _useQuickFilterGroupC.selectionType;
  if (selectionType === 'single') {
    var _selectedQuickFilters;
    return /*#__PURE__*/jsx(RadioGroup, _objectSpread(_objectSpread({
      value: (_selectedQuickFilters = selectedQuickFilters[0]) !== null && _selectedQuickFilters !== void 0 ? _selectedQuickFilters : '',
      onChange: function onChange(_ref2) {
        var value = _ref2.value;
        setSelectedQuickFilters(function () {
          return [value];
        });
      },
      size: "small"
    }, rest), {}, {
      children: children
    }));
  }
  return /*#__PURE__*/jsx(CheckboxGroup, _objectSpread(_objectSpread({
    value: selectedQuickFilters,
    onChange: function onChange(_ref3) {
      var values = _ref3.values;
      setSelectedQuickFilters(function () {
        return values;
      });
    },
    size: "small"
  }, rest), {}, {
    children: children
  }));
};
var QuickFilterGroup = function QuickFilterGroup(_ref4) {
  var children = _ref4.children,
    testID = _ref4.testID,
    value = _ref4.value,
    defaultValue = _ref4.defaultValue,
    _onChange = _ref4.onChange,
    name = _ref4.name,
    selectionType = _ref4.selectionType,
    rest = _objectWithoutProperties(_ref4, _excluded2);
  var idBase = useId('quick-filter-group');
  var _useControllableState = useControllableState({
      value: value && selectionType === 'single' ? [value] : value,
      // If selectionType is single, we need to convert the value to an array
      defaultValue: defaultValue && selectionType === 'single' ? [defaultValue] : defaultValue !== null && defaultValue !== void 0 ? defaultValue : [],
      onChange: function onChange(values) {
        return _onChange === null || _onChange === void 0 ? void 0 : _onChange({
          values: values,
          name: name !== null && name !== void 0 ? name : idBase
        });
      }
    }),
    _useControllableState2 = _slicedToArray(_useControllableState, 2),
    selectedQuickFilters = _useControllableState2[0],
    setSelectedQuickFilters = _useControllableState2[1];
  return /*#__PURE__*/jsx(QuickFilterGroupProvider, {
    value: {
      selectionType: selectionType,
      selectedQuickFilters: selectedQuickFilters
    },
    children: /*#__PURE__*/jsx(QuickFilterWrapper, _objectSpread(_objectSpread({
      onChange: _onChange,
      setSelectedQuickFilters: setSelectedQuickFilters
    }, rest), {}, {
      children: /*#__PURE__*/jsx(BaseBox, {
        display: "flex",
        flexDirection: "row",
        gap: "spacing.3",
        children: children
      })
    }))
  });
};

export { QuickFilterGroup };
