// @razorpay/blade/build/lib/web/development/components/VisuallyHidden/VisuallyHidden.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import 'react';
import styled from 'styled-components';
import { screenReaderStyles } from './ScreenReaderStyles.js';
import '../../utils/metaAttribute/index.js';
import { jsx } from 'react/jsx-runtime';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StyledVisuallyHidden = /*#__PURE__*/styled.div.withConfig({
  displayName: "VisuallyHiddenweb__StyledVisuallyHidden",
  componentId: "g3hls3-0"
})(screenReaderStyles);
var VisuallyHidden = function VisuallyHidden(_ref) {
  var children = _ref.children,
    testID = _ref.testID;
  return /*#__PURE__*/jsx(StyledVisuallyHidden, _objectSpread(_objectSpread({}, metaAttribute({
    name: MetaConstants.VisuallyHidden,
    testID: testID
  })), {}, {
    children: children
  }));
};

export { VisuallyHidden };
