// @razorpay/blade/build/lib/web/production/components/Fade/Fade.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _typeof from '@babel/runtime/helpers/typeof';
import 'react';
import '../BaseMotion/index.js';
import '../../utils/index.js';
import { cssBezierToArray } from '../../utils/cssBezierToArray.js';
import { msToSeconds } from '../../utils/msToSeconds.js';
import { jsx } from 'react/jsx-runtime';
import useTheme from '../BladeProvider/useTheme.js';
import { castWebType } from '../../utils/platform/castUtils.js';
import { BaseMotionEntryExit } from '../BaseMotion/BaseMotion.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Fade = function Fade(_ref) {
  var children = _ref.children,
    isVisible = _ref.isVisible,
    _ref$type = _ref.type,
    type = _ref$type === void 0 ? 'inout' : _ref$type,
    _ref$motionTriggers = _ref.motionTriggers,
    motionTriggers = _ref$motionTriggers === void 0 ? ['mount'] : _ref$motionTriggers,
    shouldUnmountWhenHidden = _ref.shouldUnmountWhenHidden,
    delay = _ref.delay;
  var _useTheme = useTheme(),
    theme = _useTheme.theme;
  var enterDelay = _typeof(delay) === 'object' ? delay.enter : delay;
  var exitDelay = _typeof(delay) === 'object' ? delay.exit : delay;
  var fadeVariants = {
    initial: {
      opacity: 0
    },
    animate: {
      opacity: 1,
      transition: _objectSpread(_objectSpread({}, enterDelay ? {
        delay: msToSeconds(theme.motion.delay[enterDelay])
      } : {}), {}, {
        duration: msToSeconds(theme.motion.duration.xquick),
        ease: cssBezierToArray(castWebType(theme.motion.easing.entrance))
      })
    },
    exit: {
      opacity: 0,
      transition: _objectSpread(_objectSpread({}, exitDelay ? {
        delay: msToSeconds(theme.motion.delay[exitDelay])
      } : {}), {}, {
        duration: msToSeconds(theme.motion.duration.xquick),
        ease: cssBezierToArray(castWebType(theme.motion.easing.exit))
      })
    }
  };
  return /*#__PURE__*/jsx(BaseMotionEntryExit, {
    motionVariants: fadeVariants,
    type: type,
    isVisible: isVisible,
    motionTriggers: motionTriggers,
    shouldUnmountWhenHidden: shouldUnmountWhenHidden,
    children: children
  });
};

export { Fade };
