// @razorpay/blade/build/lib/web/production/components/Table/TableContext.js
import React__default from 'react';

/* eslint-disable @typescript-eslint/no-empty-function */
// eslint-disable-next-line @typescript-eslint/no-empty-function
var TableContext = /*#__PURE__*/React__default.createContext({
  selectionType: 'none',
  selectedRows: undefined,
  totalItems: 0,
  toggleRowSelectionById: function toggleRowSelectionById() {},
  toggleAllRowsSelection: function toggleAllRowsSelection() {},
  deselectAllRows: function deselectAllRows() {},
  rowDensity: 'normal',
  toggleSort: function toggleSort() {},
  currentSortedState: {
    sortKey: '',
    isSortReversed: false
  },
  setPaginationPage: function setPaginationPage() {},
  setPaginationRowSize: function setPaginationRowSize() {},
  disabledRows: [],
  setDisabledRows: function setDisabledRows() {},
  paginationType: 'client',
  setPaginationType: function setPaginationType() {},
  backgroundColor: 'surface.background.gray.intense',
  setHeaderRowDensity: function setHeaderRowDensity() {},
  showBorderedCells: false,
  hasHoverActions: false,
  setHasHoverActions: function setHasHoverActions() {},
  columnCount: 0,
  gridTemplateColumns: undefined,
  isVirtualized: false,
  tableData: [],
  isGrouped: false,
  tableToolbarPlacement: 'inline'
});
var useTableContext = function useTableContext() {
  var context = React__default.useContext(TableContext);
  return context;
};

export { TableContext, useTableContext };
