// @razorpay/blade/build/lib/web/production/components/Divider/Divider.js
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import React__default from 'react';
import styled from 'styled-components';
import '../Box/BaseBox/index.js';
import '../Box/styledProps/index.js';
import '../../utils/index.js';
import '../../utils/metaAttribute/index.js';
import '../../utils/makeAccessible/index.js';
import { jsx } from 'react/jsx-runtime';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { makeBorderSize } from '../../utils/makeBorderSize/makeBorderSize.js';
import { isReactNative } from '../../utils/platform/isReactNative.js';
import { makeAccessible } from '../../utils/makeAccessible/makeAccessible.web.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { getStyledProps } from '../Box/styledProps/getStyledProps.js';

var _excluded = ["orientation", "dividerStyle", "variant", "thickness", "height", "width", "testID"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StyledDivider = /*#__PURE__*/styled(BaseBox).withConfig({
  displayName: "Divider__StyledDivider",
  componentId: "sc-8k3avj-0"
})(function (_ref) {
  var theme = _ref.theme,
    borderPosition = _ref.borderPosition,
    dividerStyle = _ref.dividerStyle,
    thickness = _ref.thickness,
    isDividerHorizontal = _ref.isDividerHorizontal,
    width = _ref.width,
    height = _ref.height;
  return _objectSpread(_defineProperty(_defineProperty({
    borderWidth: 0
  }, "".concat(borderPosition, "Style"), dividerStyle), "".concat(borderPosition, "Width"), makeBorderSize(theme.border.width[thickness])), isDividerHorizontal ? {
    flexGrow: 1,
    width: width
  } : {
    alignSelf: 'stretch',
    height: height
  });
});
var _Divider = function _Divider(_ref2, ref) {
  var _ref2$orientation = _ref2.orientation,
    orientation = _ref2$orientation === void 0 ? 'horizontal' : _ref2$orientation,
    _ref2$dividerStyle = _ref2.dividerStyle,
    dividerStyle = _ref2$dividerStyle === void 0 ? 'solid' : _ref2$dividerStyle,
    _ref2$variant = _ref2.variant,
    variant = _ref2$variant === void 0 ? 'muted' : _ref2$variant,
    _ref2$thickness = _ref2.thickness,
    thickness = _ref2$thickness === void 0 ? 'thin' : _ref2$thickness,
    height = _ref2.height,
    width = _ref2.width,
    testID = _ref2.testID,
    styledProps = _objectWithoutProperties(_ref2, _excluded);
  var isDividerHorizontal = orientation === 'horizontal';
  var borderPosition = isDividerHorizontal ? 'borderBottom' : 'borderLeft';
  var borderColor = _defineProperty({}, "".concat(borderPosition, "Color"), "surface.border.gray.".concat(variant));
  var accessibilityProps = isReactNative() ? {} : makeAccessible({
    role: 'separator'
  });
  return /*#__PURE__*/jsx(StyledDivider, _objectSpread(_objectSpread(_objectSpread(_objectSpread({
    ref: ref,
    borderPosition: borderPosition,
    isDividerHorizontal: isDividerHorizontal,
    dividerStyle: dividerStyle,
    thickness: thickness,
    height: height,
    width: width
  }, borderColor), metaAttribute({
    name: MetaConstants.Divider,
    testID: testID
  })), getStyledProps(styledProps)), accessibilityProps));
};
var Divider = /*#__PURE__*/React__default.forwardRef(_Divider);

export { Divider };
