// @razorpay/blade/build/lib/web/development/components/TimePicker/TimePickerFooter.web.js
import 'react';
import '../Box/index.js';
import '../Button/index.js';
import '../Divider/index.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import { Box } from '../Box/Box.js';
import { Divider } from '../Divider/Divider.js';
import Button from '../Button/Button/Button.js';

var TimePickerFooter = function TimePickerFooter(_ref) {
  var onApply = _ref.onApply,
    onCancel = _ref.onCancel,
    _ref$isApplyDisabled = _ref.isApplyDisabled,
    isApplyDisabled = _ref$isApplyDisabled === void 0 ? false : _ref$isApplyDisabled;
  return /*#__PURE__*/jsxs(Box, {
    display: "flex",
    flexDirection: "column",
    children: [/*#__PURE__*/jsx(Divider, {}), /*#__PURE__*/jsxs(Box, {
      width: "100%",
      display: "flex",
      flexDirection: "row",
      gap: {
        base: 'spacing.5',
        m: 'spacing.3'
      },
      padding: {
        base: 'spacing.5',
        m: 'spacing.4'
      },
      children: [/*#__PURE__*/jsx(Button, {
        isFullWidth: true,
        variant: "tertiary",
        size: "xsmall",
        onClick: onCancel,
        children: "Cancel"
      }), /*#__PURE__*/jsx(Button, {
        isDisabled: isApplyDisabled,
        isFullWidth: true,
        variant: "primary",
        size: "xsmall",
        onClick: onApply,
        children: "Apply"
      })]
    })]
  });
};

export { TimePickerFooter };
