// @razorpay/blade/build/lib/web/development/components/Breadcrumb/Breadcrumb.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default from 'react';
import { BreadcrumbContext } from './BreadcrumbContext.js';
import '../Box/BaseBox/index.js';
import '../Typography/index.js';
import '../../utils/makeAccessible/index.js';
import '../Box/styledProps/index.js';
import '../../utils/metaAttribute/index.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import { Text } from '../Typography/Text/Text.js';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { getStyledProps } from '../Box/styledProps/getStyledProps.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { makeAccessible } from '../../utils/makeAccessible/makeAccessible.web.js';

var _excluded = ["size", "color", "showLastSeparator", "accessibilityLabel", "children"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Separator = function Separator(_ref) {
  var size = _ref.size,
    color = _ref.color;
  return /*#__PURE__*/jsx(Text, {
    as: "span",
    color: color === 'white' ? 'surface.text.staticWhite.muted' : 'surface.text.gray.muted',
    size: size,
    weight: "medium",
    children: "/"
  });
};
var listStyleNone = {
  listStyle: 'none'
};
var _Breadcrumb = function _Breadcrumb(_ref2, ref) {
  var _ref2$size = _ref2.size,
    size = _ref2$size === void 0 ? 'medium' : _ref2$size,
    _ref2$color = _ref2.color,
    color = _ref2$color === void 0 ? 'primary' : _ref2$color,
    _ref2$showLastSeparat = _ref2.showLastSeparator,
    showLastSeparator = _ref2$showLastSeparat === void 0 ? false : _ref2$showLastSeparat,
    _ref2$accessibilityLa = _ref2.accessibilityLabel,
    accessibilityLabel = _ref2$accessibilityLa === void 0 ? 'Breadcrumb' : _ref2$accessibilityLa,
    children = _ref2.children,
    rest = _objectWithoutProperties(_ref2, _excluded);
  var contextValue = React__default.useMemo(function () {
    return {
      size: size,
      color: color
    };
  }, [size, color]);
  return /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread(_objectSpread(_objectSpread({
    ref: ref,
    as: "nav"
  }, getStyledProps(rest)), metaAttribute({
    name: MetaConstants.Breadcrumb
  })), makeAnalyticsAttribute(rest)), {}, {
    children: /*#__PURE__*/jsx(BreadcrumbContext.Provider, {
      value: contextValue,
      children: /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread({
        as: "ol",
        margin: "spacing.0",
        padding: "spacing.0",
        display: "flex",
        flexWrap: "wrap",
        gap: "spacing.2",
        alignItems: "center",
        style: listStyleNone
      }, makeAccessible({
        label: accessibilityLabel
      })), {}, {
        children: React__default.Children.map(children, function (child, index) {
          var _props;
          var ariaCurrent = child !== null && child !== void 0 && (_props = child.props) !== null && _props !== void 0 && _props.isCurrentPage ? 'page' : undefined;
          return /*#__PURE__*/jsxs(BaseBox, _objectSpread(_objectSpread(_objectSpread({
            as: "li",
            display: "flex",
            alignItems: "center",
            gap: "spacing.2"
          }, makeAccessible({
            current: ariaCurrent
          })), metaAttribute({
            name: MetaConstants.BreadcrumbItem
          })), {}, {
            children: [child, /*#__PURE__*/jsxs(BaseBox, _objectSpread(_objectSpread({
              as: "span"
            }, makeAccessible({
              hidden: true
            })), {}, {
              children: [index !== React__default.Children.count(children) - 1 && /*#__PURE__*/jsx(Separator, {
                size: size
              }), index === React__default.Children.count(children) - 1 && showLastSeparator && /*#__PURE__*/jsx(Separator, {
                size: size,
                color: color
              })]
            }))]
          }), index);
        })
      }))
    })
  }));
};
var Breadcrumb = /*#__PURE__*/React__default.forwardRef(_Breadcrumb);

export { Breadcrumb };
