// @razorpay/blade/build/lib/web/production/components/Switch/Thumb.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import styled from 'styled-components';
import { switchSizes } from './switchTokens.js';
import getIn from '../../utils/lodashButBetter/get.js';
import isNumber from '../../utils/lodashButBetter/isNumber.js';
import '../Box/BaseBox/index.js';
import '../../utils/index.js';
import '../../utils/makeSpace/index.js';
import '../../utils/makeSize/index.js';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { makeSize } from '../../utils/makeSize/makeSize.js';
import { makeSpace } from '../../utils/makeSpace/makeSpace.js';
import { isReactNative } from '../../utils/platform/isReactNative.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Thumb = /*#__PURE__*/styled(BaseBox).withConfig({
  displayName: "Thumb",
  componentId: "jf4wbz-0"
})(function (_ref) {
  var theme = _ref.theme,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'medium' : _ref$size,
    deviceType = _ref.deviceType;
  var width = switchSizes.thumb[deviceType][size].width;
  var height = switchSizes.thumb[deviceType][size].height;
  var finalWidth = isNumber(width) ? makeSize(width) : makeSpace(getIn(theme, width));
  var finalHeight = isNumber(height) ? makeSize(height) : makeSpace(getIn(theme, height));
  var reactNativeStyles = {
    left: 0,
    margin: makeSpace(theme.spacing[1])
  };
  return _objectSpread({
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    width: finalWidth,
    height: finalHeight,
    position: isReactNative() ? 'absolute' : 'relative'
  }, isReactNative() && reactNativeStyles);
});

export { Thumb };
