// @razorpay/blade/build/lib/web/development/components/StepGroup/tokens.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import '../../tokens/global/index.js';
import { size } from '../../tokens/global/size.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var markerLineDotWidth = size['2'];
var markerLineDotSpacing = size['4'];
var markerLineWidth = size['2'];
var itemTopMargin = size['6'];

/**
 * Returns the tokens that are related to marker line alignment and spacing
 */
var getMarkerLineSpacings = function getMarkerLineSpacings(size$1) {
  var LINE_SPACINGS = {
    large: {
      markerBackgroundSize: size['24'],
      markerMargin: size['2'],
      indentationWidth: size['33']
      // eslint-disable-next-line @typescript-eslint/restrict-plus-operands
    },
    medium: {
      markerBackgroundSize: size['20'],
      markerMargin: size['2'],
      indentationWidth: size['31']
    }
  };
  var markerLeftAlignment =
  // eslint-disable-next-line @typescript-eslint/restrict-plus-operands
  (LINE_SPACINGS[size$1].markerBackgroundSize + LINE_SPACINGS[size$1].markerMargin) / 2;
  var markerTopAlignment = -1 * (
  // eslint-disable-next-line @typescript-eslint/restrict-plus-operands
  LINE_SPACINGS[size$1].markerBackgroundSize + LINE_SPACINGS[size$1].markerMargin * 2 + markerLineWidth) / 2;
  return _objectSpread(_objectSpread({}, LINE_SPACINGS[size$1]), {}, {
    markerLeftAlignment: markerLeftAlignment,
    markerTopAlignment: markerTopAlignment
  });
};
var stepItemHeaderTokens = {
  medium: {
    title: 'medium',
    description: 'small',
    timestamp: 'small'
  },
  large: {
    title: 'large',
    description: 'medium',
    timestamp: 'medium'
  }
};
var iconSizeTokens = {
  medium: 'small',
  large: 'medium'
};
var itemLineGap = {
  medium: 'spacing.2',
  large: 'spacing.3'
};

export { getMarkerLineSpacings, iconSizeTokens, itemLineGap, itemTopMargin, markerLineDotSpacing, markerLineDotWidth, markerLineWidth, stepItemHeaderTokens };
