// @razorpay/blade/build/lib/web/production/components/Skeleton/Skeleton.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default from 'react';
import { PulseAnimation } from './PulseAnimation.web.js';
import '../Box/styledProps/index.js';
import '../../utils/makeAccessible/index.js';
import '../../utils/metaAttribute/index.js';
import { jsx } from 'react/jsx-runtime';
import { getStyledProps } from '../Box/styledProps/getStyledProps.js';
import { makeAccessible } from '../../utils/makeAccessible/makeAccessible.web.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';

var _excluded = ["width", "maxWidth", "minWidth", "height", "maxHeight", "minHeight", "borderRadius", "flexWrap", "flexDirection", "flexGrow", "flexShrink", "flexBasis", "alignItems", "alignContent", "alignSelf", "justifyItems", "justifyContent", "justifySelf", "placeSelf", "placeItems", "order", "testID"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _Skeleton = function _Skeleton(_ref, ref) {
  var width = _ref.width,
    maxWidth = _ref.maxWidth,
    minWidth = _ref.minWidth,
    height = _ref.height,
    maxHeight = _ref.maxHeight,
    minHeight = _ref.minHeight,
    borderRadius = _ref.borderRadius,
    flexWrap = _ref.flexWrap,
    flexDirection = _ref.flexDirection,
    flexGrow = _ref.flexGrow,
    flexShrink = _ref.flexShrink,
    flexBasis = _ref.flexBasis,
    alignItems = _ref.alignItems,
    alignContent = _ref.alignContent,
    alignSelf = _ref.alignSelf,
    justifyItems = _ref.justifyItems,
    justifyContent = _ref.justifyContent,
    justifySelf = _ref.justifySelf,
    placeSelf = _ref.placeSelf,
    placeItems = _ref.placeItems,
    order = _ref.order,
    testID = _ref.testID,
    props = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/jsx(PulseAnimation, _objectSpread(_objectSpread(_objectSpread({
    ref: ref,
    width: width,
    minWidth: minWidth,
    maxWidth: maxWidth,
    height: height,
    minHeight: minHeight,
    maxHeight: maxHeight,
    borderRadius: borderRadius,
    flexWrap: flexWrap,
    flexDirection: flexDirection,
    flexGrow: flexGrow,
    flexShrink: flexShrink,
    flexBasis: flexBasis,
    alignItems: alignItems,
    alignContent: alignContent,
    alignSelf: alignSelf,
    justifyItems: justifyItems,
    justifyContent: justifyContent,
    justifySelf: justifySelf,
    placeSelf: placeSelf,
    placeItems: placeItems,
    order: order
  }, getStyledProps(props)), makeAccessible({
    hidden: true
  })), metaAttribute({
    name: MetaConstants.Skeleton,
    testID: testID
  })));
};
var Skeleton = /*#__PURE__*/React__default.forwardRef(_Skeleton);

export { Skeleton };
