// @razorpay/blade/build/lib/web/production/components/Tabs/TabItem.web.js
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import styled from 'styled-components';
import 'react';
import { CompositeItem } from '@floating-ui/react';
import { useTabsContext } from './TabsContext.js';
import { backgroundColor, paddingTop, paddingBottom, paddingX, iconColor, textColor, textSizeMap } from './tabTokens.js';
import { useTabsItemPropRestriction, iconSizeMap } from './utils.js';
import '../Typography/index.js';
import '../../utils/index.js';
import useInteraction from '../../utils/useInteraction.js';
import '../../utils/makeAccessible/index.js';
import '../../utils/metaAttribute/index.js';
import getIn from '../../utils/lodashButBetter/get.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import { makeSpace } from '../../utils/makeSpace/makeSpace.js';
import { makeBorderSize } from '../../utils/makeBorderSize/makeBorderSize.js';
import { castWebType } from '../../utils/platform/castUtils.js';
import { makeMotionTime } from '../../utils/makeMotionTime/makeMotionTime.web.js';
import { makeAccessible } from '../../utils/makeAccessible/makeAccessible.web.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { Text } from '../Typography/Text/Text.js';

var _excluded = ["children", "value", "leading", "trailing", "isDisabled", "href", "onClick"],
  _excluded2 = ["currentInteraction"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StyledTabButton = /*#__PURE__*/styled.button.withConfig({
  displayName: "TabItemweb__StyledTabButton",
  componentId: "ucns8k-0"
})(function (_ref) {
  var theme = _ref.theme,
    isSelected = _ref.isSelected,
    size = _ref.size,
    variant = _ref.variant,
    isFullWidthTabItem = _ref.isFullWidthTabItem,
    isVertical = _ref.isVertical;
  var isFilled = variant === 'filled';
  var orientation = isVertical ? 'vertical' : 'horizontal';
  var border = isVertical ? 'borderLeft' : 'borderBottom';
  var selectedState = isSelected ? 'selected' : 'unselected';
  var _variant = variant === 'borderless' ? 'bordered' : variant;
  var background = backgroundColor[selectedState][_variant];
  return _defineProperty(_defineProperty(_defineProperty(_defineProperty(_defineProperty(_defineProperty(_defineProperty(_defineProperty(_defineProperty(_defineProperty({
    appearance: 'none',
    textDecoration: 'none',
    border: 'none',
    outline: 'none',
    cursor: 'pointer',
    display: 'flex',
    alignItems: 'center',
    justifyContent: isVertical ? 'left' : 'center',
    gap: makeSpace(theme.spacing[3]),
    width: isFullWidthTabItem ? '100%' : undefined,
    // Padding
    paddingTop: makeSpace(getIn(theme, paddingTop[_variant][orientation][size])),
    paddingBottom: makeSpace(getIn(theme, paddingBottom[_variant][orientation][size])),
    paddingLeft: makeSpace(getIn(theme, paddingX[_variant][orientation][size])),
    paddingRight: makeSpace(getIn(theme, paddingX[_variant][orientation][size])),
    // colors
    backgroundColor: isSelected && isFilled && !isVertical ? 'transparent' : getIn(theme, background["default"]),
    borderRadius: isFilled && !isVertical ? theme.border.radius.small : 0
  }, "".concat(border, "Style"), 'solid'), "".concat(border, "Width"), isFilled ? 0 : makeBorderSize(isVertical ? theme.border.width.thick : theme.border.width.thin)), "".concat(border, "Color"), isVertical && isSelected ? theme.colors.interactive.border.primary["default"] : 'transparent'), '&:hover', _defineProperty(_defineProperty({}, "".concat(border, "Color"), isVertical && isSelected ? theme.colors.interactive.border.primary["default"] : theme.colors.interactive.border.gray.highlighted), "backgroundColor",
  // Don't want to show hover state on filled tabs when vertical because
  // The hover color needs to be on the TabIndicator instead.
  isSelected && isFilled && !isVertical ? 'transparent' : getIn(theme, background.highlighted))), '&:disabled', {
    cursor: 'not-allowed',
    backgroundColor: getIn(theme, background.disabled)
  }), '&:focus-visible', {
    borderRadius: makeSpace(theme.border.radius.medium),
    boxShadow: "0px 0px 0px 4px ".concat(theme.colors.surface.border.primary.muted),
    backgroundColor: getIn(theme, background.highlighted)
  }), "transitionProperty", 'all'), "transitionTimingFunction", castWebType(theme.motion.easing.standard)), "transitionDuration", castWebType(makeMotionTime(theme.motion.duration.gentle))), '*', {
    transitionProperty: 'color, fill',
    transitionTimingFunction: castWebType(theme.motion.easing.standard),
    transitionDuration: castWebType(makeMotionTime(theme.motion.duration.xquick))
  });
});
var TabItem = function TabItem(_ref3) {
  var children = _ref3.children,
    value = _ref3.value,
    Leading = _ref3.leading,
    trailing = _ref3.trailing,
    _ref3$isDisabled = _ref3.isDisabled,
    isDisabled = _ref3$isDisabled === void 0 ? false : _ref3$isDisabled,
    href = _ref3.href,
    _onClick = _ref3.onClick,
    rest = _objectWithoutProperties(_ref3, _excluded);
  var _useTabsContext = useTabsContext(),
    size = _useTabsContext.size,
    isFullWidthTabItem = _useTabsContext.isFullWidthTabItem,
    selectedValue = _useTabsContext.selectedValue,
    setSelectedValue = _useTabsContext.setSelectedValue,
    baseId = _useTabsContext.baseId,
    variant = _useTabsContext.variant,
    isVertical = _useTabsContext.isVertical;
  var _useInteraction = useInteraction(),
    currentInteraction = _useInteraction.currentInteraction,
    interactionProps = _objectWithoutProperties(_useInteraction, _excluded2);
  var validatedTrailingComponent = useTabsItemPropRestriction(trailing, size);
  var isSelected = selectedValue === value;
  var selectedState = isSelected ? 'selected' : 'unselected';
  var panelId = "".concat(baseId, "-").concat(value, "-tabpanel");
  var tabItemId = "".concat(baseId, "-").concat(value, "-tabitem");
  var isFilled = variant === 'filled';
  var interactionMap = {
    "default": 'default',
    hover: 'highlighted',
    focus: 'highlighted',
    disabled: 'disabled'
  };
  var interaction = isDisabled ? 'disabled' : interactionMap[currentInteraction];
  return /*#__PURE__*/jsx(CompositeItem, {
    render: /*#__PURE__*/jsxs(StyledTabButton, _objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread({
      type: "button",
      as: href ? 'a' : 'button',
      href: href,
      isVertical: isVertical,
      isSelected: isSelected,
      variant: variant,
      isFullWidthTabItem: isFullWidthTabItem || isFilled,
      id: tabItemId,
      size: size,
      disabled: isDisabled
    }, interactionProps), {}, {
      onClick: function onClick(e) {
        setSelectedValue === null || setSelectedValue === void 0 || setSelectedValue(function () {
          return value;
        });
        _onClick === null || _onClick === void 0 || _onClick(e);
      }
    }, makeAccessible({
      role: 'tab',
      selected: isSelected,
      controls: panelId
    })), metaAttribute({
      name: MetaConstants.TabItem
    })), makeAnalyticsAttribute(rest)), {}, {
      children: [Leading ? /*#__PURE__*/jsx(Leading, {
        size: iconSizeMap[size],
        color: iconColor[selectedState][interaction]
      }) : null, children ? /*#__PURE__*/jsx(Text, {
        color: textColor[selectedState][interaction],
        size: textSizeMap[size],
        weight: "semibold",
        children: children
      }) : null, validatedTrailingComponent]
    }))
  });
};

export { TabItem };
