// @razorpay/blade/build/lib/web/development/components/EmptyState/EmptyState.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default from 'react';
import { emptyStateSizeTokens } from './emptyStateTokens.js';
import '../Box/BaseBox/index.js';
import '../Typography/index.js';
import '../../utils/metaAttribute/index.js';
import '../Box/styledProps/index.js';
import '../../utils/assignWithoutSideEffects/index.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import { jsxs, jsx } from 'react/jsx-runtime';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { getStyledProps } from '../Box/styledProps/getStyledProps.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { Heading } from '../Typography/Heading/Heading.js';
import { Text } from '../Typography/Text/Text.js';
import { assignWithoutSideEffects } from '../../utils/assignWithoutSideEffects/assignWithoutSideEffects.js';

var _excluded = ["asset", "title", "description", "children", "size", "testID"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _EmptyState = function _EmptyState(_ref, ref) {
  var asset = _ref.asset,
    title = _ref.title,
    description = _ref.description,
    children = _ref.children,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'medium' : _ref$size,
    testID = _ref.testID,
    props = _objectWithoutProperties(_ref, _excluded);
  var tokens = emptyStateSizeTokens[size];
  return /*#__PURE__*/jsxs(BaseBox, _objectSpread(_objectSpread(_objectSpread(_objectSpread({
    ref: ref,
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
    justifyContent: "center",
    gap: tokens.gapBetweenSections
  }, metaAttribute({
    name: MetaConstants.EmptyState,
    testID: testID
  })), getStyledProps(props)), makeAnalyticsAttribute(props)), {}, {
    children: [asset ? /*#__PURE__*/jsx(BaseBox, {
      maxWidth: tokens.assetMaxWidth,
      maxHeight: tokens.assetMaxHeight,
      children: asset
    }) : null, /*#__PURE__*/jsxs(BaseBox, {
      display: "flex",
      flexDirection: "column",
      gap: "spacing.1",
      alignItems: "center",
      children: [title ? /*#__PURE__*/jsx(Heading, {
        textAlign: "center",
        size: tokens.titleSize,
        weight: "semibold",
        color: "surface.text.gray.subtle",
        children: title
      }) : null, description ? /*#__PURE__*/jsx(Text, {
        textAlign: "center",
        size: tokens.descriptionSize,
        color: "surface.text.gray.muted",
        weight: "regular",
        variant: "body",
        children: description
      }) : null]
    }), children]
  }));
};
var EmptyState = /*#__PURE__*/assignWithoutSideEffects(/*#__PURE__*/React__default.forwardRef(_EmptyState), {
  displayName: 'EmptyState',
  componentId: 'EmptyState'
});

export { EmptyState };
