// @razorpay/blade/build/lib/web/development/components/Switch/Switch.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default from 'react';
import { ThumbIcon } from './ThumbIcon.js';
import { switchHoverTokens } from './switchTokens.js';
import { Thumb } from './Thumb.js';
import { AnimatedThumb } from './AnimatedThumb.web.js';
import { SwitchTrack } from './SwitchTrack.web.js';
import '../BladeProvider/index.js';
import '../Box/BaseBox/index.js';
import { useCheckbox } from '../Checkbox/useCheckbox.js';
import { SelectorInput } from '../Form/Selector/SelectorInput.web.js';
import '../../utils/index.js';
import '../../utils/metaAttribute/index.js';
import { SelectorLabel } from '../Form/Selector/SelectorLabel.web.js';
import '../../utils/makeAccessible/index.js';
import '../../utils/assignWithoutSideEffects/index.js';
import '../Box/styledProps/index.js';
import { getOuterMotionRef, getInnerMotionRef } from '../../utils/getMotionRefs.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import useTheme from '../BladeProvider/useTheme.js';
import { useBreakpoint } from '../../utils/useBreakpoint/useBreakpoint.js';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { getStyledProps } from '../Box/styledProps/getStyledProps.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { makeAccessible } from '../../utils/makeAccessible/makeAccessible.web.js';
import { assignWithoutSideEffects } from '../../utils/assignWithoutSideEffects/assignWithoutSideEffects.js';

var _excluded = ["defaultChecked", "isChecked", "isDisabled", "name", "onChange", "size", "value", "accessibilityLabel", "testID", "_motionMeta"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _Switch = function _Switch(_ref, ref) {
  var defaultChecked = _ref.defaultChecked,
    isChecked = _ref.isChecked,
    isDisabled = _ref.isDisabled,
    name = _ref.name,
    onChange = _ref.onChange,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'medium' : _ref$size,
    value = _ref.value,
    accessibilityLabel = _ref.accessibilityLabel,
    testID = _ref.testID,
    _motionMeta = _ref._motionMeta,
    rest = _objectWithoutProperties(_ref, _excluded);
  var _useTheme = useTheme(),
    theme = _useTheme.theme;
  var _useBreakpoint = useBreakpoint({
      breakpoints: theme.breakpoints
    }),
    matchedDeviceType = _useBreakpoint.matchedDeviceType;
  var _React$useState = React__default.useState(false),
    _React$useState2 = _slicedToArray(_React$useState, 2),
    isPressed = _React$useState2[0],
    setIsPressed = _React$useState2[1];
  var _useCheckbox = useCheckbox({
      role: 'switch',
      defaultChecked: defaultChecked,
      isChecked: isChecked,
      isIndeterminate: false,
      hasError: false,
      hasHelperText: false,
      isDisabled: isDisabled,
      isRequired: false,
      name: name,
      value: value,
      onChange: onChange
    }),
    state = _useCheckbox.state,
    inputProps = _useCheckbox.inputProps;
  var handlePointerPressedIn = React__default.useCallback(function () {
    if (isDisabled) return;
    setIsPressed(true);
  }, [isDisabled]);
  var handlePointerPressedOut = React__default.useCallback(function () {
    if (isDisabled) return;
    setIsPressed(false);
  }, [isDisabled]);
  var handleKeyboardPressedIn = React__default.useCallback(function (e) {
    if (isDisabled) return;
    if (e.key === ' ') {
      setIsPressed(true);
    }
  }, [isDisabled]);
  var handleKeyboardPressedOut = React__default.useCallback(function (e) {
    if (isDisabled) return;
    if (e.key === ' ') {
      setIsPressed(false);
    }
  }, [isDisabled]);
  return /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread(_objectSpread(_objectSpread({
    ref: getOuterMotionRef({
      _motionMeta: _motionMeta,
      ref: ref
    })
  }, metaAttribute({
    testID: testID,
    name: MetaConstants.Switch
  })), {}, {
    // @ts-ignore on rn inline-block is not valid but type here will be `flex | inline-block`
    display: state.isReactNative ? 'flex' : 'inline-block'
  }, getStyledProps(rest)), makeAnalyticsAttribute(rest)), {}, {
    children: /*#__PURE__*/jsxs(SelectorLabel, {
      componentName: MetaConstants.SwitchLabel,
      onTouchStart: handlePointerPressedIn,
      onTouchEnd: handlePointerPressedOut,
      onMouseDown: handlePointerPressedIn,
      onMouseUp: handlePointerPressedOut,
      onMouseOut: handlePointerPressedOut,
      onKeyDown: handleKeyboardPressedIn,
      onKeyUp: handleKeyboardPressedOut,
      style: {
        cursor: isDisabled ? 'not-allowed' : 'pointer'
      },
      inputProps: state.isReactNative ? // accessibility label for react-native needs to be added
      // since there is no text children inside Switch
      _objectSpread(_objectSpread({}, inputProps), makeAccessible({
        label: accessibilityLabel
      })) : {},
      children: [/*#__PURE__*/jsx(SelectorInput, {
        hoverTokens: switchHoverTokens,
        ref: getInnerMotionRef({
          _motionMeta: _motionMeta,
          ref: ref
        }),
        isChecked: state.isChecked,
        isDisabled: isDisabled,
        hasError: false,
        inputProps: inputProps,
        accessibilityLabel: accessibilityLabel
      }), /*#__PURE__*/jsx(SwitchTrack, _objectSpread(_objectSpread({
        size: size,
        deviceType: matchedDeviceType,
        isDisabled: Boolean(isDisabled),
        isChecked: state.isChecked
      }, makeAccessible({
        hidden: true
      })), {}, {
        children: /*#__PURE__*/jsx(Thumb, {
          size: size,
          deviceType: matchedDeviceType,
          isChecked: state.isChecked,
          children: /*#__PURE__*/jsx(AnimatedThumb, {
            isPressed: isPressed,
            isChecked: state.isChecked,
            isDisabled: isDisabled,
            size: size,
            children: /*#__PURE__*/jsx(ThumbIcon, {
              isChecked: state.isChecked,
              isDisabled: isDisabled,
              size: size
            })
          })
        })
      }))]
    })
  }));
};
var Switch = /*#__PURE__*/assignWithoutSideEffects(/*#__PURE__*/React__default.forwardRef(_Switch), {
  displayName: 'Switch'
});

export { Switch };
