// @razorpay/blade/build/lib/web/development/components/Switch/SwitchTrack.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import styled from 'styled-components';
import { getTrackStyles } from './getTrackStyles.js';
import '../Box/BaseBox/index.js';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var SwitchTrack = /*#__PURE__*/styled(BaseBox).withConfig({
  displayName: "SwitchTrackweb__SwitchTrack",
  componentId: "t4l0ax-0"
})(function (_ref) {
  var theme = _ref.theme,
    size = _ref.size,
    deviceType = _ref.deviceType,
    isDisabled = _ref.isDisabled,
    isChecked = _ref.isChecked;
  return _objectSpread({}, getTrackStyles({
    theme: theme,
    size: size,
    deviceType: deviceType,
    isDisabled: isDisabled,
    isChecked: isChecked
  }));
});

export { SwitchTrack };
