// @razorpay/blade/build/lib/web/production/components/Box/BaseBox/BaseBox.web.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import styled from 'styled-components';
import { useMemoizedStyles } from './useMemoizedStyles.web.js';
import { omitPropsFromHTML } from '../../../utils/omitPropsFromHTML/index.js';
import '../../../utils/metaAttribute/index.js';
import '../../../utils/assignWithoutSideEffects/index.js';
import '../../../utils/makeAnalyticsAttribute/index.js';
import { metaAttribute } from '../../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../../utils/metaAttribute/metaConstants.js';
import { makeAnalyticsAttribute } from '../../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { assignWithoutSideEffects } from '../../../utils/assignWithoutSideEffects/assignWithoutSideEffects.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var _BaseBox = /*#__PURE__*/styled.div.attrs(function (props) {
  return _objectSpread(_objectSpread({}, metaAttribute({
    name: props['data-blade-component'] || MetaConstants.BaseBox,
    testID: props['data-testid'] || props.testID
  })), makeAnalyticsAttribute(props));
}).withConfig({
  shouldForwardProp: omitPropsFromHTML,
  displayName: 'BaseBox'
})(function (props) {
  var cssObject = useMemoizedStyles(props);
  return cssObject;
});
var BaseBox = /*#__PURE__*/assignWithoutSideEffects(_BaseBox, {
  componentId: 'BaseBox'
});

export { BaseBox };
