// @razorpay/blade/build/lib/web/production/components/Charts/utils/getHighestColorInRange.js
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';

/**
 *
 * @param colorToken - The color token to get the highest color for
 * @param followIntensityMapping - Whether to follow the intensity mapping for the color token (only for categorical colors)
 * if enabled, the color token will be mapped to the intensity based on the intensity mapping
 * if disabled, the color token will be mapped to the strong intensity
 * @returns The highest color token in the range
 */

var getHighestColorInRange = function getHighestColorInRange(_ref) {
  var colorToken = _ref.colorToken,
    _ref$followIntensityM = _ref.followIntensityMapping,
    followIntensityMapping = _ref$followIntensityM === void 0 ? false : _ref$followIntensityM;
  // Check if it's a sequential color token
  var sequentialMatch = colorToken.match(/^data\.background\.sequential\.([^.]+)\.(\d+)$/);
  if (sequentialMatch) {
    var _sequentialMatch = _slicedToArray(sequentialMatch, 2),
      colorName = _sequentialMatch[1];
    return "data.background.sequential.".concat(colorName, ".1000");
  }

  // Check if it's a categorical color token
  var categoricalMatch = colorToken.match(/^data\.background\.categorical\.([^.]+)\.([^.]+)$/);
  if (categoricalMatch) {
    var _categoricalMatch = _slicedToArray(categoricalMatch, 3),
      _colorName = _categoricalMatch[1],
      intensity = _categoricalMatch[2];
    if (followIntensityMapping) {
      // Dynamic color intensity mapping
      var intensityMapping = {
        faint: 'strong',
        subtle: 'intense',
        moderate: 'moderate',
        intense: 'subtle',
        strong: 'faint'
      };
      var mappedIntensity = intensityMapping[intensity] || intensity;
      return "data.background.categorical.".concat(_colorName, ".").concat(mappedIntensity);
    } else {
      // Default behavior: always return strong
      return "data.background.categorical.".concat(_colorName, ".strong");
    }
  }

  // This should never happen due to TypeScript typing, but handle gracefully
  console.warn("Invalid color token format: ".concat(colorToken, ". Expected format: data.background.{categorical|sequential}.{color}.{intensity|value}"));
  return colorToken;
};

export { getHighestColorInRange };
