// @razorpay/blade/build/lib/web/development/components/Accordion/AccordionItem.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default from 'react';
import { AccordionButton } from './AccordionButton.web.js';
import { useAccordion, AccordionItemContext } from './AccordionContext.js';
import { AccordionItemBody } from './AccordionItemBody.js';
import { componentIds } from './componentIds.js';
import '../Divider/index.js';
import '../Box/BaseBox/index.js';
import '../../utils/metaAttribute/index.js';
import '../../utils/index.js';
import { Collapsible } from '../Collapsible/Collapsible.js';
import '../Collapsible/index.js';
import '../../utils/isValidAllowedChildren/index.js';
import '../../utils/logger/index.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import { jsx, jsxs } from 'react/jsx-runtime';
import { getComponentId } from '../../utils/isValidAllowedChildren/isValidAllowedChildren.js';
import { throwBladeError } from '../../utils/logger/logger.js';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';
import { CollapsibleBody } from '../Collapsible/CollapsibleBody.js';
import { isReactNative } from '../../utils/platform/isReactNative.js';
import { Divider } from '../Divider/Divider.js';

var _excluded = ["title", "description", "icon", "children", "isDisabled", "_index", "testID"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var AccordionItem = function AccordionItem(_ref) {
  var title = _ref.title,
    description = _ref.description,
    icon = _ref.icon,
    children = _ref.children,
    isDisabled = _ref.isDisabled,
    _index = _ref._index,
    testID = _ref.testID,
    rest = _objectWithoutProperties(_ref, _excluded);
  var _useAccordion = useAccordion(),
    expandedIndex = _useAccordion.expandedIndex,
    onExpandChange = _useAccordion.onExpandChange,
    defaultExpandedIndex = _useAccordion.defaultExpandedIndex,
    variant = _useAccordion.variant,
    numberOfItems = _useAccordion.numberOfItems;
  var isExpanded = expandedIndex === _index;
  var isDefaultExpanded = defaultExpandedIndex === _index;
  var isDeprecatedAPI = Boolean(title) || Boolean(description) || Boolean(icon);
  var _React$Children$toArr = React__default.Children.toArray(children),
    _React$Children$toArr2 = _slicedToArray(_React$Children$toArr, 2),
    header = _React$Children$toArr2[0],
    body = _React$Children$toArr2[1];
  if (!isDeprecatedAPI) {
    // Only doing validation in new API. Deprecated API allows everything as AccordionItem children
    var headerComponentId = getComponentId(header);
    var bodyComponentId = getComponentId(body);
    if (headerComponentId !== componentIds.AccordionItemHeader && bodyComponentId !== componentIds.AccordionItemBody) {
      throwBladeError({
        message: 'AccordionItem only allows AccordionItemHeader as first component and AccordionItemBody as second. Check Accordion documentation',
        moduleName: 'AccordionItem'
      });
    }
  }
  var isLastItem = _index !== undefined && _index < numberOfItems - 1;
  var handleExpandChange = function handleExpandChange(_ref2) {
    var isExpanded = _ref2.isExpanded;
    if (isExpanded && typeof _index !== 'undefined') {
      onExpandChange(_index);
    } else {
      onExpandChange(-1);
    }
  };
  return /*#__PURE__*/jsx(AccordionItemContext.Provider, {
    value: {
      index: _index,
      isDisabled: isDisabled
    },
    children: /*#__PURE__*/jsxs(BaseBox, _objectSpread(_objectSpread({}, metaAttribute({
      name: MetaConstants.AccordionItem,
      testID: testID
    })), {}, {
      children: [/*#__PURE__*/jsxs(Collapsible, _objectSpread(_objectSpread({
        isExpanded: isExpanded,
        defaultIsExpanded: isDefaultExpanded,
        onExpandChange: handleExpandChange
        // Accordion has its own width restrictions
        ,
        _shouldApplyWidthRestrictions: false,
        _dangerouslyDisableValidations: true
        // adding analytics attributes to Collapsible section
      }, makeAnalyticsAttribute(rest)), {}, {
        children: [/*#__PURE__*/jsx(AccordionButton, {
          index: _index,
          icon: icon,
          title: title,
          header: header,
          isDisabled: isDisabled,
          isDeprecatedAPI: isDeprecatedAPI
        }), /*#__PURE__*/jsx(CollapsibleBody
        // Just React Native things, need this 100% so collapsed content flows correctly inside Accordion
        // In new API, AccordionItemBody takes 100% width to avoid issues like this - https://github.com/razorpay/blade/pull/1814
        , {
          width: isReactNative() || !isDeprecatedAPI ? '100%' : undefined,
          children: isDeprecatedAPI ? /*#__PURE__*/jsx(AccordionItemBody, {
            _description: description,
            children: children
          }) : body
        })]
      })), isLastItem || variant === 'transparent' ? /*#__PURE__*/jsx(Divider, {}) : null]
    }))
  });
};

export { AccordionItem };
