// @razorpay/blade/build/lib/web/development/components/Accordion/Accordion.js
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import _objectWithoutProperties from '@babel/runtime/helpers/objectWithoutProperties';
import React__default, { useState, Children, useCallback, useMemo, cloneElement } from 'react';
import { AccordionContext } from './AccordionContext.js';
import { MAX_WIDTH } from './styles.web.js';
import '../Box/BaseBox/index.js';
import '../Box/styledProps/index.js';
import '../../tokens/global/index.js';
import '../../utils/index.js';
import '../../utils/metaAttribute/index.js';
import '../../utils/makeAnalyticsAttribute/index.js';
import { jsx } from 'react/jsx-runtime';
import { makeSize } from '../../utils/makeSize/makeSize.js';
import { size } from '../../tokens/global/size.js';
import { BaseBox } from '../Box/BaseBox/BaseBox.web.js';
import { metaAttribute } from '../../utils/metaAttribute/metaAttribute.web.js';
import { MetaConstants } from '../../utils/metaAttribute/metaConstants.js';
import { getStyledProps } from '../Box/styledProps/getStyledProps.js';
import { makeAnalyticsAttribute } from '../../utils/makeAnalyticsAttribute/makeAnalyticsAttribute.js';

var _excluded = ["defaultExpandedIndex", "expandedIndex", "onExpandChange", "showNumberPrefix", "children", "variant", "size", "maxWidth", "minWidth", "testID"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var MIN_WIDTH = {
  s: makeSize(size[200]),
  m: makeSize(size[360]),
  l: makeSize(size[360])
};
var getVariantStyles = function getVariantStyles(variant) {
  if (variant === 'transparent') {
    return {};
  }
  return {
    backgroundColor: 'surface.background.gray.intense',
    borderRadius: 'medium',
    borderWidth: 'thinner',
    borderColor: 'surface.border.gray.subtle'
  };
};

/**
 * # Accordion
 *
 * An accordion is used to allow users to toggle between different content sections in a compact vertical stack.
 *
 * ## Usage
 *
 * ```jsx
 * <Accordion>
 *  <AccordionItem>
 *    <AccordionItemHeader title="Title" />
 *    <AccordionItemBody>
 *      <Text color="surface.text.gray.subtle">
 *        Hello this is accordion body content
 *        </Text>
 *     </AccordionItemBody>
 *  </AccordionItem>
 *  <AccordionItem>
 *    <AccordionItemHeader title="Title" />
 *    <AccordionItemBody>
 *      <Text color="surface.text.gray.subtle">
 *         Hello this is accordion body content
 *      </Text>
 *    </AccordionItemBody>
 *  </AccordionItem>
 * </Accordion>
 * ```
 *
 * Checkout https://blade.razorpay.com/?path=/docs/components-accordion--docs
 *
 */
var _Accordion = function _Accordion(_ref, ref) {
  var defaultExpandedIndex = _ref.defaultExpandedIndex,
    expandedIndex = _ref.expandedIndex,
    onExpandChange = _ref.onExpandChange,
    _ref$showNumberPrefix = _ref.showNumberPrefix,
    showNumberPrefix = _ref$showNumberPrefix === void 0 ? false : _ref$showNumberPrefix,
    children = _ref.children,
    _ref$variant = _ref.variant,
    variant = _ref$variant === void 0 ? 'transparent' : _ref$variant,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'large' : _ref$size,
    maxWidth = _ref.maxWidth,
    minWidth = _ref.minWidth,
    testID = _ref.testID,
    rest = _objectWithoutProperties(_ref, _excluded);
  var _useState = useState(defaultExpandedIndex),
    _useState2 = _slicedToArray(_useState, 2),
    expandedAccordionItemIndex = _useState2[0],
    setExpandedAccordionItemIndex = _useState2[1];
  var numberOfItems = Children.count(children);
  var handleExpandChange = useCallback(function (nextExpandedIndex) {
    if (typeof expandedIndex !== 'undefined') {
      // controlled
      onExpandChange === null || onExpandChange === void 0 || onExpandChange({
        expandedIndex: nextExpandedIndex
      });
    } else {
      // uncontrolled
      setExpandedAccordionItemIndex(nextExpandedIndex);
      onExpandChange === null || onExpandChange === void 0 || onExpandChange({
        expandedIndex: nextExpandedIndex
      });
    }
  }, [onExpandChange, expandedIndex]);
  var accordionContext = useMemo(function () {
    return {
      expandedIndex: expandedIndex !== null && expandedIndex !== void 0 ? expandedIndex : expandedAccordionItemIndex,
      defaultExpandedIndex: defaultExpandedIndex,
      onExpandChange: handleExpandChange,
      showNumberPrefix: showNumberPrefix,
      variant: variant,
      numberOfItems: numberOfItems,
      size: size
    };
  }, [expandedAccordionItemIndex, handleExpandChange, expandedIndex, showNumberPrefix, defaultExpandedIndex, variant, numberOfItems, size]);
  return /*#__PURE__*/jsx(AccordionContext.Provider, {
    value: accordionContext,
    children: /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread(_objectSpread(_objectSpread({
      ref: ref,
      minWidth: minWidth !== null && minWidth !== void 0 ? minWidth : MIN_WIDTH,
      maxWidth: maxWidth !== null && maxWidth !== void 0 ? maxWidth : MAX_WIDTH
    }, metaAttribute({
      name: MetaConstants.Accordion,
      testID: testID
    })), getStyledProps(rest)), makeAnalyticsAttribute(rest)), {}, {
      children: /*#__PURE__*/jsx(BaseBox, _objectSpread(_objectSpread({}, getVariantStyles(variant)), {}, {
        width: "100%",
        children: Children.map(children, function (child, index) {
          return /*#__PURE__*/cloneElement(child, {
            _index: index,
            key: index
          });
        })
      }))
    }))
  });
};
var Accordion = /*#__PURE__*/React__default.forwardRef(_Accordion);

export { Accordion };
